# SPATA2

A Toolbox for Spatial Expression Analysis

For detailed tutorials and examples click on the link below:

https://themilolab.github.io/SPATA2/

If you have used SPATA2 for your analysis please consider citing: 

Kueckelhaus, J., Frerich, S., Kada-Benotmane, J. et al. Inferring histology-associated gene expression gradients in spatial transcriptomic studies. Nat Commun 15, 7280 (2024). https://doi.org/10.1038/s41467-024-50904-x

Furthermore, since SPATA2 uses a lot of Seurat's processing function, please also consider citing: 

Hao, Y., Stuart, T., Kowalski, M.H. et al. Dictionary learning for integrative, multimodal and scalable single-cell analysis. Nat Biotechnol 42, 293–304 (2024). https://doi.org/10.1038/s41587-023-01767-y

Additional references you could cite depend on the functions, used. E.g. `runBayesSpaceClustering()` implements the 
BayesSpace cluster algorithm of 

Zhao, E., Stone, M.R., Ren, X. et al. Spatial transcriptomics at subspot resolution with BayesSpace. Nat Biotechnol 39, 1375–1384 (2021). https://doi.org/10.1038/s41587-021-00935-2

You always find the required reference under the documentation `?runBayesSpaceClustering`, section reference.

## Licences Information
This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>.


