---
title: "Differential Expression Analysis (DEA)"
output:
  html_document:
    theme: flatly
    df_print: paged
    css: "markdown-styles.css"
    toc: true
    toc_depth: 3

---

```{r setup, include=FALSE}

knitr::opts_chunk$set(echo = TRUE, message = FALSE, fig.show = "hold", warning = FALSE, echo = TRUE, eval = TRUE, class.source = "code-chunk", out.width = "100%")

```

```{r setup-invis, echo = F, eval = T}

# load packageas
devtools::load_all()
library(tidyverse)

# load object
object_t269 <- loadSpataObject("data/T269.RDS")

```

## 1. Prerequisites

None.

## 2. Introduction

Differential expression analysis (DEA) aims to discover quantitative changes in gene expression levels between defined experimental groups. Grouping information is stored in form of grouping variables in the meta data of `SPATA2` object. This includes all SPATA2 intern generated grouping options such as [spatial segmentation](spata-v2-spatial-segmentation.html) and [clustering](spata-v2-clustering.html) as well as any other grouping variable that has been added via `addFeatures()`.  

```{r echo = T, eval = F}

library(SPATA2)
library(SPATAData)
library(tidyverse)

# download sample UKF269T
object_t269 <- downloadFromPublication(pub = "kueckelhaus_et_al_2024", id = "UKF269T")

# add the grouping variables if you created object_t269 yourself from raw data 
data("spatial_segmentations")

object_t269 <- 
  addFeatures(
    object = object_t269, 
    feature_df = spatial_segmentations$UKF269T, 
    overwrite = TRUE
  )

data("clustering")

object_t269 <- 
  addFeatures(
    object = object_t269, 
    feature_df = clustering$UKF269T, 
    overwrite = TRUE
  )

# plot histology images 
plotSurface(object_t269, color_by = "histology", pt_clrp = "npg")

plotSurface(object_t269, color_by = "bayes_space", pt_clrp = "uc")

```

The function `getGroupingOptions()` returns the variable names based on which DEA can be conducted. This vignette conducts differential expression analysis based on the histological grouping. You can exchange *'histology'* with *'bayes_space'* whenever the grouping is specified to use the clustering results. 

```{r}

getGroupingOptions(object_t269)

```


## 3. Running the analysis

DEA in **SPATA2** uses the function `Seurat::FindAllMarkers()`. It's output is a data.frame in which each row corresponds to a gene that turned out to be a marker gene for one of the identity groups. Additional variables provide information about it's p-value, adjusted p-value, logFold change etc. `runDEA()` does not alter the output but stores it in the `SPATA2` object.

```{r echo = T, eval = F}

object_t269 <- runDEA(object = object_t269, across = "histology")

```

## 3. Extracting results

There are two main functions with which you can manually extract the DEA results desired. First, `getDeaResultsDf()` returns the original resulting data.frame of `Seurat::FindAllMarkers()`. `getDeaGenes()` returns a vector of gene names. 

```{r}

# extract the complete data.frame
dea_df <- 
  getDeaResultsDf(
    object = object_t269, 
    across = "histology"
  )

nrow(dea_df)

head(dea_df)

```

Using the arguments `across_subset`, `min_lfc`, `n_highest_lfc`, `max_adj_pval`, `n_lowest_pval` the output of the function can be adjusted to specific questions. 

```{r}

# e.g. top 10 genes for histology area 'tumor' 
getDeaResultsDf(
  object = object_t269, 
  across = "histology",
  across_subset = "transition", # the group name(s) of interest,
  n_highest_lfc = 10, # top ten genes
  max_adj_pval = 0.01 # pval must be lower or equal than 0.01
)

```

## 4. Visualize results

### 4.1 Heatmaps

`plotDeaHeatmap()` visualizes the results of DEA by using to the results you would extract with `getDeaResultsDf()`. 

```{r fig.cap = "Fig.3 DEA-Heatmap."}

hm <- 
  plotDeaHeatmap(
    object = object_t269, 
    across = "histology",
    clrp = "npg",
    n_highest_lfc = 10, # subset genes
    n_bcs = 100
  )

hm


```


### 4.2 Dotplots

`plotDeaDotPlot()` visualizes the results of DEA either by group...

```{r fig.cap = "Fig.4 DEA-Dotplot by group."}

plotDeaDotPlot(
  object = object_t269, 
  across = "histology",
  across_subset = c("tumor", "infiltrated"), # specify single groups if desired
  n_highest_lfc = 5,
  by_group = TRUE, 
  scales = "free_y", 
  nrow = 1
)

```

... or with all groups together. 

```{r fig.cap = "Fig.5 DEA-Dotplot for all groups."}

plotDeaDotPlot(
  object = object_t269, 
  across = "histology",
  color_by = "histology",
  pt_clrp = "npg",
  n_highest_lfc = 5, 
  by_group = FALSE
)

```

### 4.3 Box- and Violinplots

There are additional ways to visualize the results of your DEA. As with almost all plotting functions in **SPATA2** a vector of gene names is necessary for the function to know which genes to plot. `getDeaGenes()` is the second function to extract DEA results and a wrapper around `getDeaResultsDf()` that returns a vector gene names.

```{r}

genes_of_interest <-
  getDeaGenes(
    object = object_t269,
    across = "histology", # the grouping variable
    method_de = "wilcox", # the method with which the results were computed
    n_highest_lfc = 10, 
    max_adj_pval = 0.01
  )

head(genes_of_interest) # first six
tail(genes_of_interest) # last six

```

A vector of gene names as returned by `getDeaGenes()` is a perfectly valid input for other plotting functions.

```{r out.width = "100%", fig.cap = "Fig.6 DEA-Boxplots" }

top_5_markers_269 <- 
  getDeaGenes(
    object = object_t269,
    across = "histology",
    n_lowest_pval = 5, 
    min_lfc = 0.1 # set min_lfc! else downregulated genes are included
    ) 

top_5_markers_269

# plot results for t269
plotBoxplot(
  object = object_t269,
  variables = top_5_markers_269,
  across = "histology", 
  clrp = "npg",
  nrow = 3 
  ) + 
  theme(axis.text.x = element_blank(), axis.ticks.x = element_blank()) +
  legendTop()

```

### 4.3 Surface plots

`getDeaGenes()` complements `plotSurfaceComparison()`. 

```{r out.width = "100%", fig.show = "hold", fig.cap = "Fig.8 Surface plots with DEA results for sample 269_T."}

transition_markers <- 
  getDeaGenes(object_t269, across = "histology", across_subset = "transition")

plotSurfaceComparison(
  object = object_t269, 
  color_by = transition_markers[1:9],
  pt_clrsp = color_vector("npg")[2], # plot cluster color against white
  outline = TRUE,
  nrow = 3
  ) + 
  legendNone()

```
