---
title: "Spatial Annotation Screening"
output:
  html_document:
    theme: flatly
    css: "markdown-styles.css"
    toc: true
    toc_depth: 3

---

```{r setup, include=FALSE}

knitr::opts_chunk$set(echo = TRUE, eval = TRUE, message = FALSE, class.source = "code-chunk", out.width = "50%", rows.print = 6, warning = FALSE, cache = FALSE, fig.show = "hold")

if(TRUE){

  object_t313 <- readRDS(file = "data/object_UKF313T.RDS") 
  sas_out <- readRDS(file = "data/sas_out.RDS")

}

```

This vignette exemplifies how to use Spatial Annotation Screening in SPATA2. It builds on the spatial annotations created in the vignette [Creating Spatial Annotations]() for the glioblastoma sample [UKF313T]().

```{r echo = T, eval = F}

library(SPATA2)
library(tidyverse)

# exchange my/path/to with your directory
object_t313 <- loadSpataObject("my/path/to/object_UKF313T.RDS")

```

Spatial Annotation Screening (SAS) pursues the hypothesis that specific genes - or other numeric features 343 for that matter - display non-random expression patterns in relation to spatial reference features, such as spatial annotations. SAS utilizes these reference features to incorporate the integration of potential biological forces in the identification of spatially variable genes, such as the presence of necrosis within a tumor. This allows for a supervised, hypothesis-driven screening for spatial patterns, which, unlike differential expression analysis (DEA), acknowledges the continuous nature of gene expression and avoids the limitations of group-based testing. The algorithm is wrapped up in the function `spatialAnnotationScreening()`. Note that parameters `core` and `distance` are essential for the outcome. Make sure to align either input to your questioning. See the vignette on [Using Spatial Annotations]() for more information. 

```{r echo = T, eval = F}

# prefiltering genes for spatial variability with SPARKX is recommended
object_t313 <- runSparkx(object_t313)

sparkx_genes <- getSparkxGenes(object_t313, threshold_pval = 0.05)

necrotic_ids <- c("necrotic_area", "necrotic_edge", "necrotic_edge2")

sas_out <- 
  spatialAnnotationScreening(
    object = object_t313, 
    ids = necrotic_ids, 
    variables = sparkx_genes, 
    core = FALSE, # do not include the core of the annotations
    distance = "dte" # distance to edge 
  )

```

The output of `spatialAnnotationScreening()` is an S4 object of class `SpatialAnnotationScreening` which contains the set up as well as the results from the algorithm. 

```{r fig.cap = "Fig.1 Recreate the screening set up."}

sas_areas <- color_vector(clrp = "npg", names = c("core", "environment", "periphery"))


# use the coordinates data.frame to recreate the set up 
plotSurface(sas_out@coords, color_by = "rel_loc", clrp_adjust = sas_areas)
plotSurface(sas_out@coords, color_by = "dist") + labs(color = "Dist [mm]")

```

Slot @significance contains a data.frame with one row for each screened variable which provides information regarding the degree of randomness the inferred pattern contains as quantified by the total variation (*tot_var*). The p-value gives the probability to obtain such a total variation under complete randomness and indicates the degree of significance. Column *fdr* contains the adjusted p-value according to the False Discovery Rate. 

```{r}

filter(sas_out@significance, p_value < 0.05)  

```

Slot @results contains the model fitting results. It is a data.frame where each row corresponds to a variable ~ model pair. The columns *mae* (mean absolute error) and *rmse* (root mean squared error) indicate the quality of the fit. The lower the value the better. 

```{r echo = T, eval = F}

filter(sas_out@results, rmse < 0.2)

```
```{r echo = F, eval = T}

filter(sas_out@results, rmse < 0.2) 

```

You can filter the results for significant variables below a certain threshold for a specific model to obtain genes of potential interest. 

```{r fig.cap = "Fig.2 Extraction and visualization of screening results.", fig.dim = c(9,9)}

sign_genes <- filter(sas_out@significance, fdr < 0.05)[["variables"]]

desc_genes <- 
  # filter for significance
  filter(.data = sas_out@results, variables %in% {{sign_genes}}) %>% 
  # filter for specific models
  filter(str_detect(models, pattern = "descending") & rmse < 0.2) %>% 
  pull(variables) %>% 
  unique()

asc_genes <- 
  # filter for significance
  filter(.data = sas_out@results, variables %in% {{sign_genes}}) %>% 
  # filter for specific models
  filter(str_detect(models, pattern = "ascending") & rmse < 0.2) %>% 
  pull(variables) %>% 
  unique()

length(desc_genes)

length(asc_genes)

plotSurfaceComparison(object_t313, color_by = desc_genes[1:9], normalize = T, outline = T, pt_clrsp = "BuPu") + 
  ggpLayerSpatAnnOutline(object_t313, ids = necrotic_ids, fill = "grey", incl_edge = T)

plotSurfaceComparison(object_t313, color_by = asc_genes[1:9], normalize = T, outline = T, pt_clrsp = "Reds 3") + 
  ggpLayerSpatAnnOutline(object_t313, ids = necrotic_ids, fill = "grey", incl_edge = T)

```





