<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Add histology image — ggpLayerImage • SPATA2</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Add histology image — ggpLayerImage"><meta name="description" content="Creates ggplot2 layer with the histology image
as a raster."><meta property="og:description" content="Creates ggplot2 layer with the histology image
as a raster."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">SPATA2</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">3.0.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item"><a class="nav-link" href="../articles/installation.html">Installation</a></li>
<li class="nav-item"><a class="nav-link" href="../articles/spata-data.html">SPATAData</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><h6 class="dropdown-header" data-toc-skip>Initiating a SPATA2 object</h6></li>
    <li><a class="dropdown-item" href="../articles/initiation-and-preprocessing-customized.html">Initiation &amp; Preprocessing with individual data</a></li>
    <li><a class="dropdown-item" href="../articles/initiation-and-preprocessing-merfish.html">Initiation &amp; Preprocessing with MERFISH data</a></li>
    <li><a class="dropdown-item" href="../articles/initiation-and-preprocessing-visium.html">Initiation &amp; Preprocessing with Visium data</a></li>
    <li><a class="dropdown-item" href="../articles/initiation-and-preprocessing-visium-hd.html">Initiation &amp; Preprocessing with VisiumHD data</a></li>
    <li><h6 class="dropdown-header" data-toc-skip>Basic Manipulation of a SPATA2 object</h6></li>
    <li><a class="dropdown-item" href="../articles/adding-data.html">Adding Data to SPATA2 objects</a></li>
    <li><a class="dropdown-item" href="../articles/package-compatibility.html">Platform Compatibility</a></li>
    <li><a class="dropdown-item" href="../articles/molecular-variables.html">Molecular Variables</a></li>
    <li><a class="dropdown-item" href="../articles/image-handling.html">Image Handling</a></li>
    <li><a class="dropdown-item" href="../articles/spatial-segmentation.html">Spatial Segmentation</a></li>
    <li><h6 class="dropdown-header" data-toc-skip>Implemented Algorithms &amp; Concepts</h6></li>
    <li><a class="dropdown-item" href="../articles/cnv.html">Copy Number Variations</a></li>
    <li><a class="dropdown-item" href="../articles/clustering.html">Clustering</a></li>
    <li><a class="dropdown-item" href="../articles/dimensional-reduction.html">Dimensional Reduction</a></li>
    <li><a class="dropdown-item" href="../articles/dea.html">Differential Expression</a></li>
    <li><a class="dropdown-item" href="../articles/gsea.html">Gene Set Enrichment</a></li>
    <li><a class="dropdown-item" href="../articles/spatial-measures.html">Spatial Measures</a></li>
    <li><h6 class="dropdown-header" data-toc-skip>Spatial Gradient Screening</h6></li>
    <li><a class="dropdown-item" href="../articles/creating-spatial-annotations.html">Creating Spatial Annotations</a></li>
    <li><a class="dropdown-item" href="../articles/using-spatial-annotations.html">Using Spatial Annotations</a></li>
    <li><a class="dropdown-item" href="../articles/spatial-annotation-screening.html">Spatial Annotation Screening</a></li>
    <li><a class="dropdown-item" href="../articles/creating-spatial-trajectories.html">Creating Spatial Trajectories</a></li>
    <li><a class="dropdown-item" href="../articles/spatial-trajectory-screening.html">Spatial Trajectory Screening</a></li>
    <li><h6 class="dropdown-header" data-toc-skip>Visualization</h6></li>
    <li><a class="dropdown-item" href="../articles/plotting-surface.html">Surface Plotting</a></li>
  </ul></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Add histology image</h1>

      <div class="d-none name"><code>ggpLayerImage.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Creates ggplot2 layer with the histology image
as a raster.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">ggpLayerImage</span><span class="op">(</span><span class="va">object</span>, <span class="va">...</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for class 'SPATA2'</span></span>
<span><span class="fu">ggpLayerImage</span><span class="op">(</span></span>
<span>  <span class="va">object</span>,</span>
<span>  img_name <span class="op">=</span> <span class="fu"><a href="activateImage.html">activeImage</a></span><span class="op">(</span><span class="va">object</span><span class="op">)</span>,</span>
<span>  transform <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  img_alpha <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  scale_fct <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for class 'SpatialData'</span></span>
<span><span class="fu">ggpLayerImage</span><span class="op">(</span></span>
<span>  <span class="va">object</span>,</span>
<span>  img_name <span class="op">=</span> <span class="fu"><a href="activateImage.html">activeImage</a></span><span class="op">(</span><span class="va">object</span><span class="op">)</span>,</span>
<span>  transform <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  img_alpha <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  scale_fct <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for class 'HistoImage'</span></span>
<span><span class="fu">ggpLayerImage</span><span class="op">(</span><span class="va">object</span>, transform <span class="op">=</span> <span class="cn">TRUE</span>, scale_fct <span class="op">=</span> <span class="fl">1</span>, img_alpha <span class="op">=</span> <span class="fl">1</span>, <span class="va">...</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for class 'SpatialAnnotation'</span></span>
<span><span class="fu">ggpLayerImage</span><span class="op">(</span><span class="va">object</span>, img_alpha <span class="op">=</span> <span class="fl">1</span>, rescale_axes <span class="op">=</span> <span class="cn">TRUE</span>, scale_fct <span class="op">=</span> <span class="fl">1</span>, <span class="va">...</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for class 'Image'</span></span>
<span><span class="fu">ggpLayerImage</span><span class="op">(</span><span class="va">object</span>, scale_fct <span class="op">=</span> <span class="fl">1</span>, img_alpha <span class="op">=</span> <span class="fl">1</span>, <span class="va">...</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for class 'data.frame'</span></span>
<span><span class="fu">ggpLayerImage</span><span class="op">(</span><span class="va">object</span>, <span class="va">fill_by</span>, img_alpha <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-object">object<a class="anchor" aria-label="anchor" href="#arg-object"></a></dt>
<dd><p>An object of class <code><a href="SPATA2.html">SPATA2</a></code> or, in case of S4 generics,
objects of classes for which a method has been defined.</p></dd>


<dt id="arg--">...<a class="anchor" aria-label="anchor" href="#arg--"></a></dt>
<dd><p>Used to absorb deprecated arguments or functions.</p></dd>


<dt id="arg-img-name">img_name<a class="anchor" aria-label="anchor" href="#arg-img-name"></a></dt>
<dd><p>Character value. The name of the image of interest.
If <code>NULL</code>, the <a href="concept_active.html">active</a> image is chosen by default. Either way, must
be one of <code><a href="getImageNames.html">getImageNames()</a></code>.</p></dd>


<dt id="arg-transform">transform<a class="anchor" aria-label="anchor" href="#arg-transform"></a></dt>
<dd><p>Logical value. Ignored if no images exist in the object. If <code>TRUE</code>,
spatial transformation instructions saved during image alignment of
the image <code>img_name</code> are applied. Only relevant if there are any images registered
in the <code>SPATA2</code> object.</p></dd>


<dt id="arg-img-alpha">img_alpha<a class="anchor" aria-label="anchor" href="#arg-img-alpha"></a></dt>
<dd><p>Numeric value. Sets the transparency for the image.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p><code>ggpLayer*()</code>-functions return lists of <code>ggproto</code> objects
that can be added to ggplots via the <code>+</code> operator. In most of the cases
they are supposed to be added to plots created with the <code>plotSurface*()</code>
family.</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>The image is plotted via <code><a href="https://ggplot2.tidyverse.org/reference/geom_tile.html" class="external-link">ggplot2::geom_raster()</a></code> by mapping the pixel position
to the x-axis and the y-axis. See section Image visualization
with <code>ggplot2</code> for more details.</p>
    </div>
    <div class="section level2">
    <h2 id="image-visualization-with-ggplot-">Image visualization with ggplot2<a class="anchor" aria-label="anchor" href="#image-visualization-with-ggplot-"></a></h2>



<p>When comparing the output of <code>ggplot() + ggpLayerImage()</code> with other image plotting functions,
you may notice that the image appears horizontally flipped when plotted using <code>ggpLayerImage()</code>.
This behavior is due to the use of a Cartesian coordinate system in <code>SPATA2</code>, where a pixel
with coordinates c(width = 1, height = 1) is plotted on the left side at the bottom.
In contrast, functions like <code><a href="https://rdrr.io/pkg/EBImage/man/display.html" class="external-link">EBImage::display()</a></code> or <code><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">graphics::plot()</a></code> use an <em>image space</em> coordinate system,
where pixel heights start from the top. Consequently, in <em>image space</em>, pixel c(width = 1, height = 1)
is displayed on the top resulting in mirror inverted visualization of the image.</p>
<p>We chose to use a Cartesian coordinate system in <code>SPATA2</code> because we believe it provides a more intuitive
framework for the spatial alignment of tissue, spatial annotations, spatial trajectories,
barcoded sots, single cells, etc. where coordinates in the corresponding data.frames are provided
in form of <em>x</em>- and <em>y</em>-variables. See <code><a href="getCoordsDf.html">getCoordsDf()</a></code>, <code><a href="getImgAnnOutlineDf.html">getImgAnnOutlineDf()</a></code>, <code><a href="getTissueOutlineDf.html">getTissueOutlineDf()</a></code> etc.</p>
<p>If you prefer to view your image in the regular orientation, you can use the <code><a href="flipAll.html">flipAll()</a></code> function on your object,
specifying <code>axis = "horizontal"</code>, to reverse this effect.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://themilolab.com/" class="external-link">SPATA2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyverse.tidyverse.org" class="external-link">tidyverse</a></span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"example_data"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">object</span> <span class="op">&lt;-</span> <span class="fu"><a href="loadExampleObject.html">loadExampleObject</a></span><span class="op">(</span><span class="st">"UKF275T"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span> <span class="fu">ggpLayerImage</span><span class="op">(</span><span class="va">object</span><span class="op">)</span></span>
<span></span>
<span></span></code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Jan Kueckelhaus, Dieter-Henrik Heiland.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.0.</p>
</div>

    </footer></div>





  </body></html>

