<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Package index â€¢ SPATA2</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Package index"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">SPATA2</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">3.0.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item"><a class="nav-link" href="../articles/installation.html">Installation</a></li>
<li class="nav-item"><a class="nav-link" href="../articles/spata-data.html">SPATAData</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><h6 class="dropdown-header" data-toc-skip>Initiating a SPATA2 object</h6></li>
    <li><a class="dropdown-item" href="../articles/initiation-and-preprocessing-customized.html">Initiation &amp; Preprocessing with individual data</a></li>
    <li><a class="dropdown-item" href="../articles/initiation-and-preprocessing-merfish.html">Initiation &amp; Preprocessing with MERFISH data</a></li>
    <li><a class="dropdown-item" href="../articles/initiation-and-preprocessing-visium.html">Initiation &amp; Preprocessing with Visium data</a></li>
    <li><a class="dropdown-item" href="../articles/initiation-and-preprocessing-visium-hd.html">Initiation &amp; Preprocessing with VisiumHD data</a></li>
    <li><h6 class="dropdown-header" data-toc-skip>Basic Manipulation of a SPATA2 object</h6></li>
    <li><a class="dropdown-item" href="../articles/adding-data.html">Adding Data to SPATA2 objects</a></li>
    <li><a class="dropdown-item" href="../articles/package-compatibility.html">Platform Compatibility</a></li>
    <li><a class="dropdown-item" href="../articles/molecular-variables.html">Molecular Variables</a></li>
    <li><a class="dropdown-item" href="../articles/image-handling.html">Image Handling</a></li>
    <li><a class="dropdown-item" href="../articles/spatial-segmentation.html">Spatial Segmentation</a></li>
    <li><h6 class="dropdown-header" data-toc-skip>Implemented Algorithms &amp; Concepts</h6></li>
    <li><a class="dropdown-item" href="../articles/cnv.html">Copy Number Variations</a></li>
    <li><a class="dropdown-item" href="../articles/clustering.html">Clustering</a></li>
    <li><a class="dropdown-item" href="../articles/dimensional-reduction.html">Dimensional Reduction</a></li>
    <li><a class="dropdown-item" href="../articles/dea.html">Differential Expression</a></li>
    <li><a class="dropdown-item" href="../articles/gsea.html">Gene Set Enrichment</a></li>
    <li><a class="dropdown-item" href="../articles/spatial-measures.html">Spatial Measures</a></li>
    <li><h6 class="dropdown-header" data-toc-skip>Spatial Gradient Screening</h6></li>
    <li><a class="dropdown-item" href="../articles/creating-spatial-annotations.html">Creating Spatial Annotations</a></li>
    <li><a class="dropdown-item" href="../articles/using-spatial-annotations.html">Using Spatial Annotations</a></li>
    <li><a class="dropdown-item" href="../articles/spatial-annotation-screening.html">Spatial Annotation Screening</a></li>
    <li><a class="dropdown-item" href="../articles/creating-spatial-trajectories.html">Creating Spatial Trajectories</a></li>
    <li><a class="dropdown-item" href="../articles/spatial-trajectory-screening.html">Spatial Trajectory Screening</a></li>
    <li><h6 class="dropdown-header" data-toc-skip>Visualization</h6></li>
    <li><a class="dropdown-item" href="../articles/plotting-surface.html">Surface Plotting</a></li>
  </ul></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-index">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Package index</h1>
    </div>

    <div class="section level2">
      <h2 id="around-the-spata-object">Around the SPATA2 object<a class="anchor" aria-label="anchor" href="#around-the-spata-object"></a></h2>

      <div class="section-desc"><p>These functions generate the SPATA2 object you will work with throughot this and other platforms.</p></div>


    </div><div class="section level2">

      <h3 id="initiation">Initiation<a class="anchor" aria-label="anchor" href="#initiation"></a></h3>
      <div class="section-desc"><p>The initiation functions create a SPATA2 object from scratch or with standardized output from different platforms.</p></div>


    </div><div class="section level2">




      <dl><dt>

          <code><a href="initiateSpataObject.html">initiateSpataObject()</a></code>

        </dt>
        <dd>Initiate an object of class <code>SPATA2</code></dd>
      </dl><dl><dt>

          <code><a href="initiateSpataObjectMERFISH.html">initiateSpataObjectMERFISH()</a></code>

        </dt>
        <dd>Initiate an object of class <code>SPATA2</code> from platform MERFISH</dd>
      </dl><dl><dt>

          <code><a href="initiateSpataObjectSlideSeqV1.html">initiateSpataObjectSlideSeqV1()</a></code>

        </dt>
        <dd>Initiate an object of class <code>SPATA2</code> from platform SlideSeq</dd>
      </dl><dl><dt>

          <code><a href="initiateSpataObjectVisium.html">initiateSpataObjectVisium()</a></code>

        </dt>
        <dd>Initiate an object of class <code>SPATA2</code> from platform Visium</dd>
      </dl><dl><dt>

          <code><a href="initiateSpataObjectVisiumHD.html">initiateSpataObjectVisiumHD()</a></code>

        </dt>
        <dd>Initiate an object of class <code>SPATA2</code> from platform VisiumHD</dd>
      </dl><dl><dt>

          <code><a href="initiateSpataObjectXenium.html">initiateSpataObjectXenium()</a></code>

        </dt>
        <dd>Initiate an object of class <code>SPATA2</code> from platform Xenium</dd>
      </dl></div><div class="section level2">

      <h3 id="transformation">Transformation<a class="anchor" aria-label="anchor" href="#transformation"></a></h3>
      <div class="section-desc"><p>The transform-functions ensure compatibility between different platforms.</p></div>


    </div><div class="section level2">




      <dl><dt>

          <code><a href="asGiotto.html">asGiotto()</a></code>

        </dt>
        <dd>Transform SPATA2 object to Giotto object</dd>
      </dl><dl><dt>

          <code><a href="asSPATA2.html">asSPATA2()</a></code>

        </dt>
        <dd>Transform miscellaneous objects to SPATA2 objects</dd>
      </dl><dl><dt>

          <code><a href="asSeurat.html">asSeurat()</a></code>

        </dt>
        <dd>Transform SPATA2 object to Seurat object</dd>
      </dl><dl><dt>

          <code><a href="asSingleCellExperiment.html">asSingleCellExperiment()</a></code>

        </dt>
        <dd>Transform SPATA2 object to SingleCellExperiment object</dd>
      </dl><dl><dt>

          <code><a href="asSummarizedExperiment.html">asSummarizedExperiment()</a></code>

        </dt>
        <dd>Transform SPATA2 object to SummarizedExperiment object</dd>
      </dl></div><div class="section level2">
      <h2 id="loading-downloading-and-saving">Loading, Downloading and Saving<a class="anchor" aria-label="anchor" href="#loading-downloading-and-saving"></a></h2>

      <div class="section-desc"><p>Some handy functions that make saving and loading corresponding objects more convenient.</p></div>


    </div><div class="section level2">




      <dl><dt>

          <code><a href="loadExampleObject.html">loadExampleObject()</a></code>

        </dt>
        <dd>Load example SPATA2 object</dd>
      </dl><dl><dt>

          <code><a href="loadSpataObject.html">loadSpataObject()</a></code>

        </dt>
        <dd>Load SPATA2 object</dd>
      </dl><dl><dt>

          <code><a href="saveSpataObject.html">saveSpataObject()</a></code>

        </dt>
        <dd>Save a SPATA2 object with a default directory</dd>
      </dl><dl><dt>

          <code><a href="downloadSpataObject.html">downloadSpataObject()</a></code>

        </dt>
        <dd>Download a SPATA2 object</dd>
      </dl><dl><dt>

          <code><a href="downloadSpataObjects.html">downloadSpataObjects()</a></code>

        </dt>
        <dd>Download and save several SPATA2 objects</dd>
      </dl></div><div class="section level2">
      <h2 id="object-subsetting-manipulation">Object Subsetting &amp; Manipulation<a class="anchor" aria-label="anchor" href="#object-subsetting-manipulation"></a></h2>

      <div class="section-desc"><p>Miscellaneous subsetting and renaming options.</p></div>


    </div><div class="section level2">




      <dl><dt>

          <code><a href="cropSpataObject.html">cropSpataObject()</a></code>

        </dt>
        <dd>Crop SPATA2 object</dd>
      </dl><dl><dt>

          <code><a href="filterSpataObject.html">filterSpataObject()</a></code>

        </dt>
        <dd>Subset SPATA2 object with logical expressions</dd>
      </dl><dl><dt>

          <code><a href="reduceSpataObject.html">reduceSpataObject()</a></code>

        </dt>
        <dd>Reduce SPATA2 object</dd>
      </dl><dl><dt>

          <code><a href="renameSpataObject.html">renameSpataObject()</a></code>

        </dt>
        <dd>Rename SPATA2 object</dd>
      </dl><dl><dt>

          <code><a href="splitSpataObject.html">splitSpataObject()</a></code>

        </dt>
        <dd>Split SPATA2 object</dd>
      </dl><dl><dt>

          <code><a href="subsetSpataObject.html">subsetSpataObject()</a></code>

        </dt>
        <dd>Subset SPATA2 object with barcodes</dd>
      </dl><dl><dt>

          <code><a href="updateSpataObject.html">updateSpataObject()</a></code>

        </dt>
        <dd>Update <code>SPATA2</code> object</dd>
      </dl></div><div class="section level2">
      <h2 id="object-manipulation">Object Manipulation<a class="anchor" aria-label="anchor" href="#object-manipulation"></a></h2>




    </div><div class="section level2">

      <h3 id="add-content-to-spata-object">Add Content to SPATA2 object<a class="anchor" aria-label="anchor" href="#add-content-to-spata-object"></a></h3>
      <div class="section-desc"><p>The add-functions let you add content to the <code>SPATA2</code> object savely.</p></div>


    </div><div class="section level2">




      <dl><dt>

          <code><a href="addFeatures.html">addFeatures()</a></code>

        </dt>
        <dd>Add meta features</dd>
      </dl><dl><dt>

          <code><a href="addInnerBorder.html">addInnerBorder()</a></code>

        </dt>
        <dd>Add holes to spatial annotations</dd>
      </dl><dl><dt>

          <code><a href="addMetaDataMol.html">addMetaDataMol()</a></code>

        </dt>
        <dd>Add meta variables for molecular data</dd>
      </dl><dl><dt>

          <code><a href="addMetaDataObs.html">addMetaDataObs()</a></code>

        </dt>
        <dd>Add meta variables for observations</dd>
      </dl><dl><dt>

          <code><a href="addMoleculeCoordinates.html">addMoleculeCoordinates()</a></code>

        </dt>
        <dd>Add molecule coordinates</dd>
      </dl><dl><dt>

          <code><a href="addProcessedMatrix.html">addProcessedMatrix()</a></code>

        </dt>
        <dd>Add a processed matrix</dd>
      </dl><dl><dt>

          <code><a href="addSignature.html">addGeneSet()</a></code> <code><a href="addSignature.html">addMetaboliteSet()</a></code> <code><a href="addSignature.html">addProteinSet()</a></code> <code><a href="addSignature.html">addSignature()</a></code>

        </dt>
        <dd>Add molecular signature</dd>
      </dl><dl><dt>

          <code><a href="addSpatialAnnotation.html">addSpatialAnnotation()</a></code>

        </dt>
        <dd>Create and add a spatial annotation manually</dd>
      </dl><dl><dt>

          <code><a href="addSpatialTrajectory.html">addSpatialTrajectory()</a></code>

        </dt>
        <dd>Create and add a spatial trajectory</dd>
      </dl></div><div class="section level2">

      <h3 id="create-content">Create Content<a class="anchor" aria-label="anchor" href="#create-content"></a></h3>
      <div class="section-desc"><p>Create miscellaneous content and/or data containers.</p></div>


    </div><div class="section level2">




      <dl><dt>

          <code><a href="createHistoImage.html">createHistoImage()</a></code>

        </dt>
        <dd>Create an object of class <code>HistoImage</code></dd>
      </dl><dl><dt>

          <code><a href="createMolecularAssay.html">createMolecularAssay()</a></code>

        </dt>
        <dd>Create and add an object of class <code>MolecularAssay</code></dd>
      </dl><dl><dt>

          <code><a href="createSpatialData.html">createSpatialData()</a></code> <code><a href="createSpatialData.html">createSpatialDataMERFISH()</a></code> <code><a href="createSpatialData.html">createSpatialDataSlideSeqV1()</a></code> <code><a href="createSpatialData.html">createSpatialDataVisium()</a></code> <code><a href="createSpatialData.html">createSpatialDataVisiumHD()</a></code> <code><a href="createSpatialData.html">createSpatialDataXenium()</a></code>

        </dt>
        <dd>Create an object of class <code>SpatialData</code> from raw output</dd>
      </dl><dl><dt>

          <code><a href="createSpatialMethod.html">createSpatialMethod()</a></code>

        </dt>
        <dd>Create an object of class <code>SpatialMethod</code></dd>
      </dl><dl><dt>

          <code><a href="createSpatialSegmentation.html">createSpatialSegmentation()</a></code>

        </dt>
        <dd>Interactive sample segmentation</dd>
      </dl><dl><dt>

          <code><a href="createSpatialTrajectories.html">createSpatialTrajectories()</a></code>

        </dt>
        <dd>Create and add spatial trajectories interactively</dd>
      </dl></div><div class="section level2">

      <h3 id="process-miscellaneous-content">Process Miscellaneous Content<a class="anchor" aria-label="anchor" href="#process-miscellaneous-content"></a></h3>
      <div class="section-desc"><p>A collection of miscellaneous processing functions.</p></div>


    </div><div class="section level2">




      <dl><dt>

          <code><a href="identifyVariableMolecules.html">identifyVariableMolecules()</a></code>

        </dt>
        <dd>Identify molecules of high variability</dd>
      </dl><dl><dt>

          <code><a href="identifyPixelContent.html">identifyPixelContent()</a></code>

        </dt>
        <dd>Identify pixel content</dd>
      </dl><dl><dt>

          <code><a href="normalizeCounts.html">normalizeCounts()</a></code>

        </dt>
        <dd>Normalize raw counts</dd>
      </dl></div><div class="section level2">

      <h3 id="rename-content">Rename Content<a class="anchor" aria-label="anchor" href="#rename-content"></a></h3>
      <div class="section-desc"><p>The rename-functions let you adjust the name of certain content with which you refer to it via arguments.</p></div>


    </div><div class="section level2">




      <dl><dt>

          <code><a href="renameGroups.html">renameGroups()</a></code> <code><a href="renameGroups.html">renameTissueSection()</a></code>

        </dt>
        <dd>Rename cluster/group names</dd>
      </dl><dl><dt>

          <code><a href="renameImage.html">renameImage()</a></code>

        </dt>
        <dd>Rename an image</dd>
      </dl><dl><dt>

          <code><a href="renameMetaFeatures.html">renameMetaFeatures()</a></code>

        </dt>
        <dd>Rename features</dd>
      </dl><dl><dt>

          <code><a href="renameSpataObject.html">renameSpataObject()</a></code>

        </dt>
        <dd>Rename SPATA2 object</dd>
      </dl><dl><dt>

          <code><a href="renameSpatialAnnotation.html">renameSpatialAnnotation()</a></code>

        </dt>
        <dd>Rename a spatial annotation</dd>
      </dl><dl><dt>

          <code><a href="renameSpatialTrajectory.html">renameSpatialTrajectory()</a></code>

        </dt>
        <dd>Rename a spatial trajectory.</dd>
      </dl></div><div class="section level2">

      <h3 id="set-content">Set Content<a class="anchor" aria-label="anchor" href="#set-content"></a></h3>
      <div class="section-desc"><p>The set-functions let you set the content of specific slots.</p></div>


    </div><div class="section level2">




      <dl><dt>

          <code><a href="setDefault.html">setDefault()</a></code>

        </dt>
        <dd>Set object specific default</dd>
      </dl><dl><dt>

          <code><a href="setImageDir.html">setImageDir()</a></code>

        </dt>
        <dd>Set image directory</dd>
      </dl><dl><dt>

          <code><a href="setMetaVarDf.html">setMetaVarDf()</a></code>

        </dt>
        <dd>Set molecular meta data.frame</dd>
      </dl><dl><dt>

          <code><a href="setScaleFactor.html">setScaleFactor()</a></code>

        </dt>
        <dd>Set scale factors</dd>
      </dl><dl><dt>

          <code><a href="setSpataDir.html">setSpataDir()</a></code>

        </dt>
        <dd>Set SPATA2 directory</dd>
      </dl></div><div class="section level2">
      <h2 id="availability-checks">Availability Checks<a class="anchor" aria-label="anchor" href="#availability-checks"></a></h2>

      <div class="section-desc"><p>A collection of functions that check availability of processing steps and return logical values. (This is a subset of all related functions.)</p></div>


    </div><div class="section level2">




      <dl><dt>

          <code><a href="containsAssay.html">containsAssay()</a></code>

        </dt>
        <dd>Check availability of an assay</dd>
      </dl><dl><dt>

          <code><a href="containsCCD.html">containsCCD()</a></code>

        </dt>
        <dd>Check availability of center to center distance</dd>
      </dl><dl><dt>

          <code><a href="containsCNV.html">containsCNV()</a></code>

        </dt>
        <dd>Check availability CNV results</dd>
      </dl><dl><dt>

          <code><a href="containsCells.html">containsCells()</a></code>

        </dt>
        <dd>Check availability of cells</dd>
      </dl><dl><dt>

          <code><a href="containsHistoImages.html">containsHistoImages()</a></code>

        </dt>
        <dd>Check availability of image containers</dd>
      </dl><dl><dt>

          <code><a href="containsImage.html">containsImage()</a></code>

        </dt>
        <dd>Check availability of an image</dd>
      </dl><dl><dt>

          <code><a href="containsInnerBorders.html">containsInnerBorders()</a></code>

        </dt>
        <dd>Check for inner borders in a spatial annotation</dd>
      </dl><dl><dt>

          <code><a href="containsMethod.html">containsMethod()</a></code>

        </dt>
        <dd>Check availability of specific methods</dd>
      </dl><dl><dt>

          <code><a href="containsModality.html">containsModality()</a></code>

        </dt>
        <dd>Check availability of molecular modality</dd>
      </dl><dl><dt>

          <code><a href="containsMoleculeCoordinates.html">containsMoleculeCoordinates()</a></code>

        </dt>
        <dd>Check availability molecule coordinates</dd>
      </dl><dl><dt>

          <code><a href="containsPixelContent.html">containsPixelContent()</a></code>

        </dt>
        <dd>Check availability pixel content</dd>
      </dl><dl><dt>

          <code><a href="containsScaleFactor.html">containsScaleFactor()</a></code>

        </dt>
        <dd>Check availability of specific scale factors</dd>
      </dl><dl><dt>

          <code><a href="containsSpatialAnnotations.html">containsSpatialAnnotations()</a></code>

        </dt>
        <dd>Check availability of spatial annotations</dd>
      </dl><dl><dt>

          <code><a href="containsSpatialData.html">containsSpatialData()</a></code>

        </dt>
        <dd>Checks availability of a SpatialData object</dd>
      </dl><dl><dt>

          <code><a href="containsSpatialOutliers.html">containsSpatialOutliers()</a></code>

        </dt>
        <dd>Check if spatial outliers exist</dd>
      </dl><dl><dt>

          <code><a href="containsSpots.html">containsSpots()</a></code>

        </dt>
        <dd>Check availability of spots</dd>
      </dl><dl><dt>

          <code><a href="containsTissueOutline.html">containsTissueOutline()</a></code>

        </dt>
        <dd>Check availability of tissue outline</dd>
      </dl></div><div class="section level2">
      <h2 id="built-in-data">Built-in Data<a class="anchor" aria-label="anchor" href="#built-in-data"></a></h2>

      <div class="section-desc"><p>A list of data built into SPATA2.</p></div>


    </div><div class="section level2">




      <dl><dt>

          <code><a href="cnv_ref.html">cnv_ref</a></code>

        </dt>
        <dd>A list of reference data for copy number variation analysis (CNV)</dd>
      </dl><dl><dt>

          <code><a href="cnv_regions_df.html">cnv_regions_df</a></code>

        </dt>
        <dd>A data.frame necessary for cnv-analysis. Contains information about start and end points of chromosomes.</dd>
      </dl><dl><dt>

          <code><a href="example_data.html">example_data</a></code>

        </dt>
        <dd>Data required for reproducible examples in SPATA2 tutorials.</dd>
      </dl><dl><dt>

          <code><a href="gene_pos_df.html">gene_pos_df</a></code>

        </dt>
        <dd>A data.frame necessary for cnv-analysis. Contains information about the gene positions on chromosomes. Contains the following variables:</dd>
      </dl><dl><dt>

          <code><a href="signatures.html">signatures</a></code>

        </dt>
        <dd>The default collection of frequently used signatures</dd>
      </dl><dl><dt>

          <code><a href="visium_spots.html">visium_spots</a></code>

        </dt>
        <dd>A list of a slot named <em>VisiumSmall</em>, which contains the data.frame of all barcoded spots of 6.5mm capture area and a slot named <em>VisiumLarge</em> which contains the barcoded spots of the 11mm capture area.</dd>
      </dl></div><div class="section level2">
      <h2 id="activate-content">Activate Content<a class="anchor" aria-label="anchor" href="#activate-content"></a></h2>

      <div class="section-desc"><p>Functions to manage object specific defaults.</p></div>


    </div><div class="section level2">




      <dl><dt>

          <code><a href="activateAssay.html">activateAssay()</a></code> <code><a href="activateAssay.html">activeAssay()</a></code>

        </dt>
        <dd>Default assay name</dd>
      </dl><dl><dt>

          <code><a href="activateImage.html">activeImage()</a></code> <code><a href="activateImage.html">activateImage()</a></code> <code><a href="activateImage.html">activateImageInt()</a></code>

        </dt>
        <dd>Default image name</dd>
      </dl><dl><dt>

          <code><a href="activateMatrix.html">activateMatrix()</a></code> <code><a href="activateMatrix.html">activeMatrix()</a></code>

        </dt>
        <dd>Default matrix name</dd>
      </dl></div><div class="section level2">
      <h2 id="counting">Counting<a class="anchor" aria-label="anchor" href="#counting"></a></h2>

      <div class="section-desc"><p>Functions that quantify the abundance of specific ascpects in the <code>SPATA2</code> object.</p></div>


    </div><div class="section level2">




      <dl><dt>

          <code><a href="nMolecules.html">nGenes()</a></code> <code><a href="nMolecules.html">nMolecules()</a></code> <code><a href="nMolecules.html">nMetabolites()</a></code> <code><a href="nMolecules.html">nProteins()</a></code>

        </dt>
        <dd>Number of molecules</dd>
      </dl><dl><dt>

          <code><a href="nObs.html">nObs()</a></code>

        </dt>
        <dd>Number of observations</dd>
      </dl><dl><dt>

          <code><a href="nSpatialAnnotations.html">nSpatialAnnotations()</a></code>

        </dt>
        <dd>Number of spatial annotations</dd>
      </dl><dl><dt>

          <code><a href="nSpatialTrajectories.html">nSpatialTrajectories()</a></code> <code><a href="nSpatialTrajectories.html">nTrajectories()</a></code>

        </dt>
        <dd>Number of spatial trajectories</dd>
      </dl></div><div class="section level2">
      <h2 id="extract-data">Extract Data<a class="anchor" aria-label="anchor" href="#extract-data"></a></h2>




    </div><div class="section level2">

      <h3 id="copy-number-variations">Copy Number Variations<a class="anchor" aria-label="anchor" href="#copy-number-variations"></a></h3>
      <div class="section-desc"><p>Functions to extract CNV related content.</p></div>


    </div><div class="section level2">




      <dl><dt>

          <code><a href="getChrRegionsDf.html">getChrRegionsDf()</a></code>

        </dt>
        <dd>Obtain chromosome information</dd>
      </dl><dl><dt>

          <code><a href="getCnvFeatureNames.html">getCnvFeatureNames()</a></code>

        </dt>
        <dd>Obtain features names under which cnv-analysis results are stored.</dd>
      </dl><dl><dt>

          <code><a href="getCnvGenesDf.html">getCnvGenesDf()</a></code>

        </dt>
        <dd>Obtain CNV results by gene</dd>
      </dl><dl><dt>

          <code><a href="getCnvResults.html">getCnvResults()</a></code>

        </dt>
        <dd>Obtain CNV results</dd>
      </dl></div><div class="section level2">

      <h3 id="molecular-data">Molecular Data<a class="anchor" aria-label="anchor" href="#molecular-data"></a></h3>
      <div class="section-desc"><p>Functions to extract molecular data and particularly names of molecular variables.</p></div>


    </div><div class="section level2">




      <dl><dt>

          <code><a href="getAssay.html">getAssay()</a></code>

        </dt>
        <dd>Obtain molecular assay</dd>
      </dl><dl><dt>

          <code><a href="getAssayModalities.html">getAssayModalities()</a></code> <code><a href="getAssayModalities.html">getAssayNames()</a></code>

        </dt>
        <dd>Obtain assay names/modalites</dd>
      </dl><dl><dt>

          <code><a href="getGenePosDf.html">getGenePosDf()</a></code>

        </dt>
        <dd>Obtain gene CNV information</dd>
      </dl><dl><dt>

          <code><a href="getSignature.html">getGeneSet()</a></code> <code><a href="getSignature.html">getMetaboliteSet()</a></code> <code><a href="getSignature.html">getProteinSet()</a></code> <code><a href="getSignature.html">getSignature()</a></code>

        </dt>
        <dd>Obtain molecular signature</dd>
      </dl><dl><dt>

          <code><a href="getSignatureList.html">getGeneSetList()</a></code> <code><a href="getSignatureList.html">getMetaboliteSetList()</a></code> <code><a href="getSignatureList.html">getProteinSetList()</a></code> <code><a href="getSignatureList.html">getSignatureList()</a></code>

        </dt>
        <dd>Obtain molecular signatures</dd>
      </dl><dl><dt>

          <code><a href="getSignatureNames.html">getGeneSets()</a></code> <code><a href="getSignatureNames.html">getMetaboliteSets()</a></code> <code><a href="getSignatureNames.html">getProteinSets()</a></code> <code><a href="getSignatureNames.html">getSignatureNames()</a></code>

        </dt>
        <dd>Obtain molecular signature names</dd>
      </dl><dl><dt>

          <code><a href="getSignatureOverview.html">getGeneSetOverview()</a></code> <code><a href="getSignatureOverview.html">getMetaboliteOverview()</a></code> <code><a href="getSignatureOverview.html">getProteinSetOverview()</a></code> <code><a href="getSignatureOverview.html">getSignatureOverview()</a></code>

        </dt>
        <dd>Overview about the current signature collection</dd>
      </dl><dl><dt>

          <code><a href="getCountMatrix.html">getCountMatrix()</a></code>

        </dt>
        <dd>Obtain raw counts</dd>
      </dl><dl><dt>

          <code><a href="getMatrix.html">getMatrix()</a></code> <code><a href="getMatrix.html">getMatrices()</a></code>

        </dt>
        <dd>Obtain a data matrix</dd>
      </dl><dl><dt>

          <code><a href="getMatrixNames.html">getMatrixNames()</a></code>

        </dt>
        <dd>Obtain matrix names</dd>
      </dl><dl><dt>

          <code><a href="getProcessedMatrix.html">getProcessedMatrix()</a></code>

        </dt>
        <dd>Obtain processed data matrix</dd>
      </dl><dl><dt>

          <code><a href="getProcessedMatrixNames.html">getProcessedMatrixNames()</a></code>

        </dt>
        <dd>Obtain names of processed matrices</dd>
      </dl><dl><dt>

          <code><a href="getMetaVarDf.html">getMetaVarDf()</a></code>

        </dt>
        <dd>Obtain molecular meta data.frame</dd>
      </dl><dl><dt>

          <code><a href="getModality.html">getModality()</a></code>

        </dt>
        <dd>Obatin molecular modality</dd>
      </dl><dl><dt>

          <code><a href="getMoleculeCoordinates.html">getMoleculeCoordinates()</a></code>

        </dt>
        <dd>Obtain molecule coordinates</dd>
      </dl><dl><dt>

          <code><a href="getMoleculeTypeList.html">getMoleculeTypeList()</a></code>

        </dt>
        <dd>Obtain a list of molecules</dd>
      </dl><dl><dt>

          <code><a href="getMolecules.html">getGenes()</a></code> <code><a href="getMolecules.html">getMetabolites()</a></code> <code><a href="getMolecules.html">getMolecules()</a></code> <code><a href="getMolecules.html">getProteins()</a></code>

        </dt>
        <dd>Obtain molecule names</dd>
      </dl><dl><dt>

          <code><a href="getSignatureTypeList.html">getSignatureTypeList()</a></code>

        </dt>
        <dd>Obtain a list of signatures</dd>
      </dl><dl><dt>

          <code><a href="getVariableMolecules.html">getVariableMolecules()</a></code>

        </dt>
        <dd>Obtain molecules of high variability</dd>
      </dl></div><div class="section level2">

      <h3 id="spatial-coordinates">Spatial Coordinates<a class="anchor" aria-label="anchor" href="#spatial-coordinates"></a></h3>
      <div class="section-desc"><p>Functions to extract coordinates of observations.</p></div>


    </div><div class="section level2">




      <dl><dt>

          <code><a href="getCoordsCenter.html">getCoordsCenter()</a></code>

        </dt>
        <dd>Obtain coordinate center</dd>
      </dl><dl><dt>

          <code><a href="getCoordsDf.html">getCoordsDf()</a></code>

        </dt>
        <dd>Obtain coordinates</dd>
      </dl><dl><dt>

          <code><a href="getCoordsDfSA.html">getCoordsDfSA()</a></code> <code><a href="getCoordsDfSA.html">getCoordsDfST()</a></code>

        </dt>
        <dd>Relate points to spatial annotations</dd>
      </dl><dl><dt>

          <code><a href="getCoordsMtr.html">getCoordsMtr()</a></code>

        </dt>
        <dd>Obtain coordinates matrix</dd>
      </dl><dl><dt>

          <code><a href="getCoordsRange.html">getCoordsRange()</a></code>

        </dt>
        <dd>Obtain coordinate range</dd>
      </dl></div><div class="section level2">

      <h3 id="dimensional-reduction">Dimensional Reduction<a class="anchor" aria-label="anchor" href="#dimensional-reduction"></a></h3>
      <div class="section-desc"><p>Functions to extract dimensional reduction results.</p></div>


    </div><div class="section level2">




      <dl><dt>

          <code><a href="getDimRedDf.html">getDimRedDf()</a></code> <code><a href="getDimRedDf.html">getPcaDf()</a></code> <code><a href="getDimRedDf.html">getPcaMtr()</a></code> <code><a href="getDimRedDf.html">getTsneDf()</a></code> <code><a href="getDimRedDf.html">getUmapDf()</a></code>

        </dt>
        <dd>Obtain dim red data.frame</dd>
      </dl></div><div class="section level2">

      <h3 id="features-grouping">Features &amp; Grouping<a class="anchor" aria-label="anchor" href="#features-grouping"></a></h3>
      <div class="section-desc"><p>Functions to extract grouping related variables and names.</p></div>


    </div><div class="section level2">




      <dl><dt>

          <code><a href="getCnvFeatureNames.html">getCnvFeatureNames()</a></code>

        </dt>
        <dd>Obtain features names under which cnv-analysis results are stored.</dd>
      </dl><dl><dt>

          <code><a href="getFeatureNames.html">getFeatureNames()</a></code>

        </dt>
        <dd>Obtain metadata column names</dd>
      </dl><dl><dt>

          <code><a href="getGroupingOptions.html">getGroupingOptions()</a></code>

        </dt>
        <dd>Obtain variable names that group data points</dd>
      </dl><dl><dt>

          <code><a href="getGroupNames.html">getGroupNames()</a></code>

        </dt>
        <dd>Obtain group names a grouping variable contains</dd>
      </dl><dl><dt>

          <code><a href="getMetaDf.html">getMetaDf()</a></code>

        </dt>
        <dd>Obtain meta data.frame for observations</dd>
      </dl><dl><dt>

          <code><a href="getSpatSegmVarNames.html">getSpatSegmVarNames()</a></code>

        </dt>
        <dd>Obtain segmentation variable names</dd>
      </dl></div><div class="section level2">

      <h3 id="differential-expression-analysis">Differential Expression Analysis<a class="anchor" aria-label="anchor" href="#differential-expression-analysis"></a></h3>
      <div class="section-desc"><p>Functions to extract DEA related content.</p></div>


    </div><div class="section level2">




      <dl><dt>

          <code><a href="getDeaResultsDf.html">getDeaGenes()</a></code> <code><a href="getDeaResultsDf.html">getDeaResultsDf()</a></code>

        </dt>
        <dd>Obtain DEA results</dd>
      </dl></div><div class="section level2">

      <h3 id="gene-set-enrichment-analysis">Gene Set Enrichment Analysis<a class="anchor" aria-label="anchor" href="#gene-set-enrichment-analysis"></a></h3>
      <div class="section-desc"><p>Functions to extract GSEA related content.</p></div>


    </div><div class="section level2">




      <dl><dt>

          <code><a href="getGseaDf.html">getGseaDf()</a></code>

        </dt>
        <dd>Obtain enrichment data.frame</dd>
      </dl><dl><dt>

          <code><a href="getGseaResults.html">getGseaResults()</a></code>

        </dt>
        <dd>Obtain enrichment results</dd>
      </dl></div><div class="section level2">

      <h3 id="spatial-gradient-screening">Spatial Gradient Screening<a class="anchor" aria-label="anchor" href="#spatial-gradient-screening"></a></h3>
      <div class="section-desc"><p>Functions to extract SGS related content.</p></div>


    </div><div class="section level2">




      <dl><dt>

          <code><a href="getSgsResultsDf.html">getSgsResultsDf()</a></code>

        </dt>
        <dd>Obtain spatial gradient screening results</dd>
      </dl><dl><dt>

          <code><a href="getSgsResultsVec.html">getSgsResultsVec()</a></code>

        </dt>
        <dd>Obtain spatial gradient screening results</dd>
      </dl><dl><dt>

          <code><a href="getSasDf.html">getSasDf()</a></code>

        </dt>
        <dd>Obtain spatial annotation screening data.frame</dd>
      </dl><dl><dt>

          <code><a href="getStsDf.html">getStsDf()</a></code>

        </dt>
        <dd>Obtain spatial trajectory screening data.frame</dd>
      </dl></div><div class="section level2">

      <h3 id="image">Image<a class="anchor" aria-label="anchor" href="#image"></a></h3>
      <div class="section-desc"><p>Functions to extract image related content.</p></div>


    </div><div class="section level2">




      <dl><dt>

          <code><a href="getHistoImage.html">getHistoImage()</a></code> <code><a href="getHistoImage.html">getHistoImageRef()</a></code>

        </dt>
        <dd>Obtain HistoImage object</dd>
      </dl><dl><dt>

          <code><a href="getImage.html">getImage()</a></code>

        </dt>
        <dd>Obtain image</dd>
      </dl><dl><dt>

          <code><a href="getImageCenter.html">getImageCenter()</a></code>

        </dt>
        <dd>Obtain image center</dd>
      </dl><dl><dt>

          <code><a href="getImageDf.html">getImageDf()</a></code>

        </dt>
        <dd>Obtain image as a data.frame</dd>
      </dl><dl><dt>

          <code><a href="getImageDims.html">getImageDims()</a></code> <code><a href="getImageDims.html">getImageRange()</a></code>

        </dt>
        <dd>Obtain image dimensions/ranges</dd>
      </dl><dl><dt>

          <code><a href="getImageDir.html">getImageDir()</a></code> <code><a href="getImageDir.html">getImageDirectories()</a></code>

        </dt>
        <dd>Obtain directory to image</dd>
      </dl><dl><dt>

          <code><a href="getImageNames.html">getImageNames()</a></code>

        </dt>
        <dd>Obtain names of registered images</dd>
      </dl><dl><dt>

          <code><a href="getImageSectionsByBarcode.html">getImageSectionsByBarcode()</a></code>

        </dt>
        <dd>Obtain image sections by barcode</dd>
      </dl><dl><dt>

          <code><a href="getImageTransformations.html">getImageTransformations()</a></code>

        </dt>
        <dd>Obtain image transformation instructions</dd>
      </dl><dl><dt>

          <code><a href="getPixelDf.html">getPixelDf()</a></code>

        </dt>
        <dd>Obtain pixel data.frame</dd>
      </dl><dl><dt>

          <code><a href="getPixelScaleFactor.html">getPixelScaleFactor()</a></code>

        </dt>
        <dd>Obtain scale factor for pixel to SI conversion</dd>
      </dl></div><div class="section level2">

      <h3 id="spatial-annotations">Spatial Annotations<a class="anchor" aria-label="anchor" href="#spatial-annotations"></a></h3>
      <div class="section-desc"><p>Functions to extract data related to spatial annotations.</p></div>


    </div><div class="section level2">




      <dl><dt>

          <code><a href="getSpatialAnnotation.html">getSpatialAnnotation()</a></code>

        </dt>
        <dd>Obtain a SpatialAnnotation object</dd>
      </dl><dl><dt>

          <code><a href="getSpatialAnnotations.html">getSpatialAnnotations()</a></code>

        </dt>
        <dd>Obtain list of SpatialAnnotation objects</dd>
      </dl><dl><dt>

          <code><a href="getSpatAnnArea.html">getSpatAnnArea()</a></code>

        </dt>
        <dd>Obtain area of spatial annotation</dd>
      </dl><dl><dt>

          <code><a href="getSpatAnnBarcodes.html">getSpatAnnBarcodes()</a></code>

        </dt>
        <dd>Obtain barcodes by spatial annotations</dd>
      </dl><dl><dt>

          <code><a href="getSpatAnnCenter.html">getSpatAnnCenter()</a></code> <code><a href="getSpatAnnCenter.html">getSpatAnnCenters()</a></code>

        </dt>
        <dd>Obtain center of a spatial annotation</dd>
      </dl><dl><dt>

          <code><a href="getSpatAnnCenterBcsp.html">getSpatAnnCenterBcsp()</a></code>

        </dt>
        <dd>Obtain center data point</dd>
      </dl><dl><dt>

          <code><a href="getSpatAnnIds.html">getSpatAnnIds()</a></code>

        </dt>
        <dd>Obtain IDs of spatial annotations</dd>
      </dl><dl><dt>

          <code><a href="getSpatAnnOutlineDf.html">getSpatAnnOutlineDf()</a></code>

        </dt>
        <dd>Obtain the outline of a spatial annotation</dd>
      </dl><dl><dt>

          <code><a href="getSpatAnnRange.html">getSpatAnnRange()</a></code>

        </dt>
        <dd>Obtain spatial annotations range</dd>
      </dl><dl><dt>

          <code><a href="getSpatAnnSf.html">getSpatAnnSf()</a></code>

        </dt>
        <dd>Obtain the outline of a spatial anontation</dd>
      </dl><dl><dt>

          <code><a href="getSpatAnnTags.html">getSpatAnnTags()</a></code>

        </dt>
        <dd>Obtain spatial annotation tags</dd>
      </dl></div><div class="section level2">

      <h3 id="spatial-trajectories">Spatial Trajectories<a class="anchor" aria-label="anchor" href="#spatial-trajectories"></a></h3>
      <div class="section-desc"><p>Functions to extract data related to spatial trajectories.</p></div>


    </div><div class="section level2">




      <dl><dt>

          <code><a href="getProjectionDf.html">getProjectionDf()</a></code>

        </dt>
        <dd>Obtain trajectory projection</dd>
      </dl><dl><dt>

          <code><a href="getSpatialTrajectory.html">getSpatialTrajectory()</a></code> <code><a href="getSpatialTrajectory.html">getSpatialTrajectories()</a></code>

        </dt>
        <dd>Obtain SpatialTrajectory objects</dd>
      </dl><dl><dt>

          <code><a href="getSpatialTrajectoryIds.html">getSpatialTrajectoryIds()</a></code>

        </dt>
        <dd>Obtain trajectory IDs</dd>
      </dl><dl><dt>

          <code><a href="getTrajectoryLength.html">getTrajectoryLength()</a></code>

        </dt>
        <dd>Obtain length of spatial trajectory</dd>
      </dl><dl><dt>

          <code><a href="getTrajectorySegmentDf.html">getTrajectorySegmentDf()</a></code>

        </dt>
        <dd>Obtain trajectory course</dd>
      </dl><dl><dt>

          <code><a href="getTrajectoryWidth.html">getTrajectoryWidth()</a></code>

        </dt>
        <dd>Obtain trajectory width</dd>
      </dl></div><div class="section level2">

      <h3 id="tissue-tissue-outline-spatial-measures">Tissue, Tissue Outline &amp; Spatial Measures<a class="anchor" aria-label="anchor" href="#tissue-tissue-outline-spatial-measures"></a></h3>
      <div class="section-desc"><p>Functions to extract tissue related data.</p></div>


    </div><div class="section level2">




      <dl><dt>

          <code><a href="getCaptureArea.html">getCaptureArea()</a></code>

        </dt>
        <dd>Obtain capture area</dd>
      </dl><dl><dt>

          <code><a href="getCCD.html">getCCD()</a></code>

        </dt>
        <dd>Obtain center to center distance</dd>
      </dl><dl><dt>

          <code><a href="getDefaultUnit.html">getDefaultUnit()</a></code>

        </dt>
        <dd>Obtain default unit</dd>
      </dl><dl><dt>

          <code><a href="getSpatialMethod.html">getSpatialMethod()</a></code>

        </dt>
        <dd>Obtain spatial method</dd>
      </dl><dl><dt>

          <code><a href="getTissueArea.html">getTissueArea()</a></code>

        </dt>
        <dd>Obtain tissue area size</dd>
      </dl><dl><dt>

          <code><a href="getTissueOutlineCentroid.html">getTissueOutlineCentroid()</a></code>

        </dt>
        <dd>Obtain tissue outline centroid</dd>
      </dl><dl><dt>

          <code><a href="getTissueOutlineDf.html">getTissueOutlineDf()</a></code>

        </dt>
        <dd>Obtain the outline of tissue sections</dd>
      </dl><dl><dt>

          <code><a href="getTissueSections.html">getTissueSections()</a></code>

        </dt>
        <dd>Obtain the names of tissue sections</dd>
      </dl></div><div class="section level2">

      <h3 id="miscellaneous">Miscellaneous<a class="anchor" aria-label="anchor" href="#miscellaneous"></a></h3>
      <div class="section-desc"><p>Functions to extract miscellaneous types of data.</p></div>


    </div><div class="section level2">




      <dl><dt>

          <code><a href="getBarcodes.html">getBarcodes()</a></code>

        </dt>
        <dd>Obtain barcodes</dd>
      </dl><dl><dt>

          <code><a href="getDefaultInstructions.html">getDefault()</a></code> <code><a href="getDefaultInstructions.html">getDefaultInstructions()</a></code>

        </dt>
        <dd>Obtain default argument inputs</dd>
      </dl><dl><dt>

          <code><a href="getLogfileDf.html">getLogfileDf()</a></code>

        </dt>
        <dd>Obtain object logfile data.frame</dd>
      </dl><dl><dt>

          <code><a href="getMethodSpecifics.html">getMethodSpecifics()</a></code>

        </dt>
        <dd>Obtain platform details</dd>
      </dl><dl><dt>

          <code><a href="getPlatform.html">getPlatform()</a></code>

        </dt>
        <dd>Obtain platform name</dd>
      </dl><dl><dt>

          <code><a href="getSampleName.html">getSampleName()</a></code>

        </dt>
        <dd>Obtain name of SPATA2 object</dd>
      </dl><dl><dt>

          <code><a href="getScaleFactor.html">getScaleFactor()</a></code>

        </dt>
        <dd>Obtain scale factors</dd>
      </dl><dl><dt>

          <code><a href="getSpataDf.html">getSpataDf()</a></code>

        </dt>
        <dd>Obtain a data.frame of observations</dd>
      </dl><dl><dt>

          <code><a href="getSpataDir.html">getSpataDir()</a></code>

        </dt>
        <dd>Obtain SPATA2 object directory</dd>
      </dl><dl><dt>

          <code><a href="getSpatialData.html">getSpatialData()</a></code>

        </dt>
        <dd>Obtain SpatialData object</dd>
      </dl><dl><dt>

          <code><a href="getVariableNames.html">getVariableNames()</a></code>

        </dt>
        <dd>Obtain variable names of the SPATA2 object</dd>
      </dl></div><div class="section level2">
      <h2 id="remove-content">Remove Content<a class="anchor" aria-label="anchor" href="#remove-content"></a></h2>

      <div class="section-desc"><p>Remove miscellaneous kinds of data from a SPATA2 object.</p></div>


    </div><div class="section level2">




      <dl><dt>

          <code><a href="registerImage.html">registerImage()</a></code> <code><a href="registerImage.html">removeImage()</a></code>

        </dt>
        <dd>Register or remove images</dd>
      </dl><dl><dt>

          <code><a href="removeMetaFeatures.html">removeMetaFeatures()</a></code>

        </dt>
        <dd>Remove meta features</dd>
      </dl><dl><dt>

          <code><a href="removeMolecules.html">removeGenes()</a></code> <code><a href="removeMolecules.html">removeGenesStress()</a></code> <code><a href="removeMolecules.html">removeGenesZeroCounts()</a></code> <code><a href="removeMolecules.html">removeMolecules()</a></code>

        </dt>
        <dd>Remove molecules from the <code>SPATA2</code> object</dd>
      </dl><dl><dt>

          <code><a href="removeObs.html">removeObs()</a></code> <code><a href="removeObs.html">removeObsZeroCounts()</a></code>

        </dt>
        <dd>Remove observations</dd>
      </dl><dl><dt>

          <code><a href="removeProcessedMatrix.html">removeProcessedMatrix()</a></code>

        </dt>
        <dd>Remove a processed matrix</dd>
      </dl><dl><dt>

          <code><a href="removeSpatialAnnotations.html">removeSpatialAnnotations()</a></code>

        </dt>
        <dd>Remove spatial annotations</dd>
      </dl><dl><dt>

          <code><a href="removeSpatialOutliers.html">removeSpatialOutliers()</a></code>

        </dt>
        <dd>Remove spatial outliers</dd>
      </dl><dl><dt>

          <code><a href="removeSpatialTrajectories.html">removeSpatialTrajectories()</a></code>

        </dt>
        <dd>Remove spatial trajectories</dd>
      </dl><dl><dt>

          <code><a href="removeTissueFragments.html">removeTissueFragments()</a></code>

        </dt>
        <dd>Remove data points from tissue fragments</dd>
      </dl></div><div class="section level2">
      <h2 id="management-of-grouping-variables">Management of Grouping Variables<a class="anchor" aria-label="anchor" href="#management-of-grouping-variables"></a></h2>

      <div class="section-desc"><p>Functions to change grouping variables.</p></div>


    </div><div class="section level2">




      <dl><dt>

          <code><a href="mergeGroups.html">mergeGroups()</a></code>

        </dt>
        <dd>Lump groups together</dd>
      </dl><dl><dt>

          <code><a href="dissolveGroups.html">dissolveGroups()</a></code>

        </dt>
        <dd>Dissolve groups in a SPATA2 object</dd>
      </dl><dl><dt>

          <code><a href="renameGroups.html">renameGroups()</a></code> <code><a href="renameGroups.html">renameTissueSection()</a></code>

        </dt>
        <dd>Rename cluster/group names</dd>
      </dl><dl><dt>

          <code><a href="relevelGroups.html">relevelGroups()</a></code>

        </dt>
        <dd>Relevel groups of grouping variable</dd>
      </dl></div><div class="section level2">
      <h2 id="spatial-annotation-handling">Spatial Annotation Handling<a class="anchor" aria-label="anchor" href="#spatial-annotation-handling"></a></h2>

      <div class="section-desc"><p>Functions to create and work with spatial annotations.</p></div>


    </div><div class="section level2">

      <h3 id="create-spatial-annotations">Create Spatial Annotations<a class="anchor" aria-label="anchor" href="#create-spatial-annotations"></a></h3>



    </div><div class="section level2">




      <dl><dt>

          <code><a href="addSpatialAnnotation.html">addSpatialAnnotation()</a></code>

        </dt>
        <dd>Create and add a spatial annotation manually</dd>
      </dl><dl><dt>

          <code><a href="barcodesToSpatialAnnotation.html">barcodesToSpatialAnnotation()</a></code>

        </dt>
        <dd>Create spatial annotations from a list of barcodes</dd>
      </dl><dl><dt>

          <code><a href="createGroupAnnotations.html">createGroupAnnotations()</a></code>

        </dt>
        <dd>Create spatial annotations from a group of data points</dd>
      </dl><dl><dt>

          <code><a href="createImageAnnotations.html">createImageAnnotations()</a></code>

        </dt>
        <dd>Create spatial annotations based on histo-morphological features</dd>
      </dl><dl><dt>

          <code><a href="createNumericAnnotations.html">createNumericAnnotations()</a></code>

        </dt>
        <dd>Create spatial annotations based on numeric values</dd>
      </dl></div><div class="section level2">

      <h3 id="manipulate-spatial-annotations">Manipulate Spatial Annotations<a class="anchor" aria-label="anchor" href="#manipulate-spatial-annotations"></a></h3>



    </div><div class="section level2">




      <dl><dt>

          <code><a href="centerSpatialAnnotation.html">centerSpatialAnnotation()</a></code>

        </dt>
        <dd>Center the borders of a spatial annotation</dd>
      </dl><dl><dt>

          <code><a href="expandSpatialAnnotation.html">expandSpatialAnnotation()</a></code>

        </dt>
        <dd>Expand the outline of spatial annotations</dd>
      </dl><dl><dt>

          <code><a href="mergeSpatialAnnotations.html">mergeSpatialAnnotations()</a></code>

        </dt>
        <dd>Merge spatial annotations</dd>
      </dl><dl><dt>

          <code><a href="renameSpatialAnnotation.html">renameSpatialAnnotation()</a></code>

        </dt>
        <dd>Rename a spatial annotation</dd>
      </dl><dl><dt>

          <code><a href="rotateSpatialAnnotation.html">rotateSpatialAnnotation()</a></code>

        </dt>
        <dd>Rotate the outline of a spatial annotation</dd>
      </dl><dl><dt>

          <code><a href="shiftSpatialAnnotation.html">shiftSpatialAnnotation()</a></code>

        </dt>
        <dd>Shift the borders of a spatial annotation</dd>
      </dl><dl><dt>

          <code><a href="smoothSpatialAnnotation.html">smoothSpatialAnnotation()</a></code>

        </dt>
        <dd>Smooth the border of a spatial annotation</dd>
      </dl><dl><dt>

          <code><a href="tagSpatialAnnotations.html">tagSpatialAnnotations()</a></code>

        </dt>
        <dd>Tag spatial annotations</dd>
      </dl></div><div class="section level2">

      <h3 id="use-spatial-annotations">Use Spatial Annotations<a class="anchor" aria-label="anchor" href="#use-spatial-annotations"></a></h3>



    </div><div class="section level2">




      <dl><dt>

          <code><a href="mapSpatialAnnotationTags.html">mapSpatialAnnotationTags()</a></code>

        </dt>
        <dd>Spatial annotation and barcode intersection</dd>
      </dl><dl><dt>

          <code><a href="spatialAnnotationToGrouping.html">spatialAnnotationToGrouping()</a></code>

        </dt>
        <dd>Convert spatial annotation to grouping variable</dd>
      </dl></div><div class="section level2">
      <h2 id="image-handling-and-image-coordinate-alignment">Image handling and Image-Coordinate Alignment<a class="anchor" aria-label="anchor" href="#image-handling-and-image-coordinate-alignment"></a></h2>

      <div class="section-desc"><p>Functions to work with and ensure alignment between images and coordinates of observations.</p></div>


    </div><div class="section level2">




      <dl><dt>

          <code><a href="flipAll.html">flipAll()</a></code> <code><a href="flipAll.html">flipCoordinates()</a></code> <code><a href="flipAll.html">flipCoordsDf()</a></code> <code><a href="flipAll.html">flipImage()</a></code> <code><a href="flipAll.html">flipSpatialAnnotations()</a></code> <code><a href="flipAll.html">flipSpatialTrajectories()</a></code>

        </dt>
        <dd>Mirror invert image and coordinates</dd>
      </dl><dl><dt>

          <code><a href="loadImage.html">loadImage()</a></code> <code><a href="loadImage.html">loadImages()</a></code>

        </dt>
        <dd>Load image slot content</dd>
      </dl><dl><dt>

          <code><a href="refImage.html">refImage()</a></code>

        </dt>
        <dd>Obtain name of reference iamge</dd>
      </dl><dl><dt>

          <code><a href="registerImage.html">registerImage()</a></code> <code><a href="registerImage.html">removeImage()</a></code>

        </dt>
        <dd>Register or remove images</dd>
      </dl><dl><dt>

          <code><a href="resetImageTransformations.html">resetImageTransformations()</a></code>

        </dt>
        <dd>Reset image transformations</dd>
      </dl><dl><dt>

          <code><a href="rotateAll.html">rotateAll()</a></code> <code><a href="rotateAll.html">rotateImage()</a></code> <code><a href="rotateAll.html">rotateCoordinates()</a></code> <code><a href="rotateAll.html">rotateCoordsDf()</a></code> <code><a href="rotateAll.html">rotateTissueOutlineDf()</a></code> <code><a href="rotateAll.html">rotateSpatialAnnotations()</a></code> <code><a href="rotateAll.html">rotateSpatialTrajectories()</a></code>

        </dt>
        <dd>Rotate image and coordinates</dd>
      </dl><dl><dt>

          <code><a href="rotateSpatialAnnotation.html">rotateSpatialAnnotation()</a></code>

        </dt>
        <dd>Rotate the outline of a spatial annotation</dd>
      </dl><dl><dt>

          <code><a href="unloadImage.html">unloadImage()</a></code> <code><a href="unloadImage.html">unloadImages()</a></code>

        </dt>
        <dd>Unload image slot content</dd>
      </dl></div><div class="section level2">
      <h2 id="join-information">Join Information<a class="anchor" aria-label="anchor" href="#join-information"></a></h2>

      <div class="section-desc"><p>These functions access the information of interest, perform additional customized computations and join it as additional variables to already existing data.frames in a tidy-data fashion.</p></div>


    </div><div class="section level2">




      <dl><dt>

          <code><a href="joinWithPca.html">joinWithPca()</a></code> <code><a href="joinWithPca.html">joinWithTsne()</a></code> <code><a href="joinWithPca.html">joinWithUmap()</a></code>

        </dt>
        <dd>Join barcodes with additional variables</dd>
      </dl><dl><dt>

          <code><a href="joinWithVariables.html">joinWithVariables()</a></code>

        </dt>
        <dd>Join data with variables</dd>
      </dl></div><div class="section level2">
      <h2 id="implemented-algorithms">Implemented Algorithms<a class="anchor" aria-label="anchor" href="#implemented-algorithms"></a></h2>

      <div class="section-desc"><p>These functions perform calculations that might take some time. The results are stored inside the SPATA2 object.</p></div>


    </div><div class="section level2">




      <dl><dt>

          <code><a href="computeCnvByChrArm.html">computeCnvByChrArm()</a></code>

        </dt>
        <dd>Compute CNV by chromosome arm</dd>
      </dl><dl><dt>

          <code><a href="computeMetaFeatures.html">computeMetaFeatures()</a></code>

        </dt>
        <dd>Compute meta features</dd>
      </dl><dl><dt>

          <code><a href="computePixelScaleFactor.html">computePixelScaleFactor()</a></code>

        </dt>
        <dd>Compute pixel scale factor</dd>
      </dl><dl><dt>

          <code><a href="compute_img_scale_fct.html">compute_img_scale_fct()</a></code>

        </dt>
        <dd>Compute scale factor of two images</dd>
      </dl><dl><dt>

          <code><a href="runBayesSpaceClustering.html">runBayesSpaceClustering()</a></code>

        </dt>
        <dd>Clustering with BayesSpace</dd>
      </dl><dl><dt>

          <code><a href="runCNV.html">runCNV()</a></code> <code><a href="runCNV.html">runCnvAnalysis()</a></code>

        </dt>
        <dd>Identify large-scale chromosomal copy number variations</dd>
      </dl><dl><dt>

          <code><a href="runDEA.html">runDEA()</a></code> <code><a href="runDEA.html">runDeAnalysis()</a></code>

        </dt>
        <dd>Differential expression analysis (DEA)</dd>
      </dl><dl><dt>

          <code><a href="runGSEA.html">runGSEA()</a></code>

        </dt>
        <dd>Compute gene set enrichment</dd>
      </dl><dl><dt>

          <code><a href="runKmeansClustering.html">runKmeansClustering()</a></code>

        </dt>
        <dd>Clustering with Kmeans</dd>
      </dl><dl><dt>

          <code><a href="runNearestNeighborClustering.html">runNearestNeighborClustering()</a></code>

        </dt>
        <dd>Clustering with nearest neighbor search</dd>
      </dl><dl><dt>

          <code><a href="runPCA.html">runPCA()</a></code> <code><a href="runPCA.html">runPca()</a></code>

        </dt>
        <dd>Run Principal Component Analysis</dd>
      </dl><dl><dt>

          <code><a href="runSPARKX.html">getSparkxGeneDf()</a></code> <code><a href="runSPARKX.html">getSparkxGenes()</a></code> <code><a href="runSPARKX.html">getSparkxResults()</a></code> <code><a href="runSPARKX.html">runSPARKX()</a></code>

        </dt>
        <dd>Identify spatially significant features with SPARKX</dd>
      </dl><dl><dt>

          <code><a href="runSeuratClustering.html">runSeuratClustering()</a></code>

        </dt>
        <dd>Clustering with Seurat</dd>
      </dl><dl><dt>

          <code><a href="runTSNE.html">runTSNE()</a></code> <code><a href="runTSNE.html">runTsne()</a></code>

        </dt>
        <dd>Run t-Stochastic Neighbour Embedding</dd>
      </dl><dl><dt>

          <code><a href="runUMAP.html">runUMAP()</a></code> <code><a href="runUMAP.html">runUmap()</a></code>

        </dt>
        <dd>Run Uniform Manifold Approximation and Projection</dd>
      </dl></div><div class="section level2">
      <h2 id="tissue-outline-tissue-sections">Tissue Outline &amp; Tissue Sections<a class="anchor" aria-label="anchor" href="#tissue-outline-tissue-sections"></a></h2>

      <div class="section-desc"><p>Functions that incorporate the spatial dimensions and outline of the tissue using data points or images as information.</p></div>


    </div><div class="section level2">




      <dl><dt>

          <code><a href="distToEdge.html">distToEdge()</a></code>

        </dt>
        <dd>Distance to edge of tissue section</dd>
      </dl><dl><dt>

          <code><a href="identifyPixelContent.html">identifyPixelContent()</a></code>

        </dt>
        <dd>Identify pixel content</dd>
      </dl><dl><dt>

          <code><a href="identifyTissueOutline.html">identifyTissueOutline()</a></code>

        </dt>
        <dd>Identify tissue outline</dd>
      </dl><dl><dt>

          <code><a href="identifySpatialOutliers.html">identifySpatialOutliers()</a></code>

        </dt>
        <dd>Identify spatial outliers</dd>
      </dl><dl><dt>

          <code><a href="mergeTissueSections.html">mergeTissueSections()</a></code>

        </dt>
        <dd>Merge tissue sections</dd>
      </dl><dl><dt>

          <code><a href="mergeWithTissueOutline.html">mergeWithTissueOutline()</a></code>

        </dt>
        <dd>Integrate tissue outline in spatial annotation</dd>
      </dl><dl><dt>

          <code><a href="renameGroups.html">renameGroups()</a></code> <code><a href="renameGroups.html">renameTissueSection()</a></code>

        </dt>
        <dd>Rename cluster/group names</dd>
      </dl><dl><dt>

          <code><a href="useVarForTissueOutline.html">useVarForTissueOutline()</a></code>

        </dt>
        <dd>Use specified variable for tissue outline</dd>
      </dl><dl><dt>

          <code><a href="whichTissueSection.html">whichTissueSection()</a></code>

        </dt>
        <dd>Tissue section belonging of spatial annotations</dd>
      </dl></div><div class="section level2">
      <h2 id="miscellaneous-spatial-functions">Miscellaneous Spatial Functions<a class="anchor" aria-label="anchor" href="#miscellaneous-spatial-functions"></a></h2>




    </div><div class="section level2">




      <dl><dt>

          <code><a href="add_dbscan_variable.html">add_dbscan_variable()</a></code> <code><a href="add_dbscan_variable.html">add_tissue_section_variable()</a></code>

        </dt>
        <dd>Add tissue section variable to coordinate data.frame</dd>
      </dl><dl><dt>

          <code><a href="add_edge_variable.html">add_edge_variable()</a></code>

        </dt>
        <dd>Add edge variable to coordinate data.frame</dd>
      </dl><dl><dt>

          <code><a href="add_grid_variable.html">add_grid_variable()</a></code>

        </dt>
        <dd>Add grid variable to coordinate data.rame</dd>
      </dl><dl><dt>

          <code><a href="buffer_area.html">buffer_area()</a></code>

        </dt>
        <dd>Buffer area</dd>
      </dl><dl><dt>

          <code><a href="identify_obs_in_polygon.html">identify_obs_in_polygon()</a></code> <code><a href="identify_obs_in_polygon.html">identify_obs_in_spat_ann()</a></code>

        </dt>
        <dd>Identify observations inside a polygon</dd>
      </dl><dl><dt>

          <code><a href="include_tissue_outline.html">include_tissue_outline()</a></code>

        </dt>
        <dd>Include spatial extent of tissue sections in analysis</dd>
      </dl><dl><dt>

          <code><a href="intersect_polygons.html">intersect_polygons()</a></code>

        </dt>
        <dd>Test polygon intersection</dd>
      </dl><dl><dt>

          <code><a href="map_to_tissue_section.html">map_to_tissue_section()</a></code>

        </dt>
        <dd>Map observations to tissue sections</dd>
      </dl><dl><dt>

          <code><a href="merge_intersecting_polygons.html">merge_intersecting_polygons()</a></code>

        </dt>
        <dd>Merge polygons</dd>
      </dl></div><div class="section level2">
      <h2 id="spatial-gradient-screening-1">Spatial Gradient Screening<a class="anchor" aria-label="anchor" href="#spatial-gradient-screening-1"></a></h2>

      <div class="section-desc"><p>Functions around Spatial Trajectory Screening and Image Annotation Screening.</p></div>


    </div><div class="section level2">




      <dl><dt>

          <code><a href="spatial_gradient_screening.html">spatial_gradient_screening()</a></code>

        </dt>
        <dd>Low level implementation of the spatial gradient screening algorithm</dd>
      </dl><dl><dt>

          <code><a href="spatialAnnotationScreening.html">spatialAnnotationScreening()</a></code>

        </dt>
        <dd>The Spatial Annotation Screening algorithm</dd>
      </dl><dl><dt>

          <code><a href="spatialTrajectoryScreening.html">spatialTrajectoryScreening()</a></code>

        </dt>
        <dd>The Spatial Trajectory Screening algorithm</dd>
      </dl><dl><dt>

          <code><a href="getStsDf.html">getStsDf()</a></code>

        </dt>
        <dd>Obtain spatial trajectory screening data.frame</dd>
      </dl><dl><dt>

          <code><a href="getSasDf.html">getSasDf()</a></code>

        </dt>
        <dd>Obtain spatial annotation screening data.frame</dd>
      </dl><dl><dt>

          <code><a href="plotSasBarplot.html">plotSasBarplot()</a></code>

        </dt>
        <dd>Plot SAS barplot</dd>
      </dl><dl><dt>

          <code><a href="plotSasHeatmap.html">plotSasHeatmap()</a></code>

        </dt>
        <dd>Plot SAS heatmap</dd>
      </dl><dl><dt>

          <code><a href="plotSasLineplot.html">plotSasLineplot()</a></code>

        </dt>
        <dd>Plot SAS lineplot</dd>
      </dl><dl><dt>

          <code><a href="plotSasRidgeplot.html">plotSasRidgeplot()</a></code>

        </dt>
        <dd>Plot SAS rideplot</dd>
      </dl></div><div class="section level2">
      <h2 id="spatial-measures">Spatial Measures<a class="anchor" aria-label="anchor" href="#spatial-measures"></a></h2>

      <div class="section-desc"><p>Functions to work with spatial measures in SI units.</p></div>


    </div><div class="section level2">




      <dl><dt>

          <code><a href="are_all_area_or_dist.html">are_all_area_or_dist()</a></code> <code><a href="are_all_area_or_dist.html">are_all_dist()</a></code>

        </dt>
        <dd>Test area or distance input</dd>
      </dl><dl><dt>

          <code><a href="as_unit.html">as_meter()</a></code> <code><a href="as_unit.html">as_meter2()</a></code> <code><a href="as_unit.html">as_micrometer()</a></code> <code><a href="as_unit.html">as_micrometer2()</a></code> <code><a href="as_unit.html">as_millimeter()</a></code> <code><a href="as_unit.html">as_millimeter2()</a></code> <code><a href="as_unit.html">as_nanometer()</a></code> <code><a href="as_unit.html">as_nanometer2()</a></code> <code><a href="as_unit.html">as_pixel()</a></code> <code><a href="as_unit.html">as_unit()</a></code> <code><a href="as_unit.html">as_centimeter()</a></code> <code><a href="as_unit.html">as_centimeter2()</a></code> <code><a href="as_unit.html">as_decimeter()</a></code> <code><a href="as_unit.html">as_decimeter2()</a></code>

        </dt>
        <dd>Transform distance and area values</dd>
      </dl><dl><dt>

          <code><a href="as_SPATA2_dist.html">as_SPATA2_dist()</a></code>

        </dt>
        <dd>Distance transformation</dd>
      </dl><dl><dt>

          <code><a href="extract_unit.html">extract_unit()</a></code>

        </dt>
        <dd>Extract distance units</dd>
      </dl><dl><dt>

          <code><a href="extract_value.html">extract_value()</a></code>

        </dt>
        <dd>Extract distance value</dd>
      </dl><dl><dt>

          <code><a href="is_area.html">is_area()</a></code> <code><a href="is_area.html">is_area_pixel()</a></code> <code><a href="is_area.html">is_area_si()</a></code>

        </dt>
        <dd>Test area input</dd>
      </dl><dl><dt>

          <code><a href="is_dist.html">is_dist()</a></code> <code><a href="is_dist.html">is_dist_si()</a></code> <code><a href="is_dist.html">is_dist_pixel()</a></code>

        </dt>
        <dd>Test distance input</dd>
      </dl><dl><dt>

          <code><a href="is_unit_area.html">is_unit_area()</a></code>

        </dt>
        <dd>Test unit of area input</dd>
      </dl><dl><dt>

          <code><a href="is_unit_dist.html">is_unit_dist()</a></code>

        </dt>
        <dd>Test unit of length input</dd>
      </dl></div><div class="section level2">
      <h2 id="s-classes">S4 Classes<a class="anchor" aria-label="anchor" href="#s-classes"></a></h2>

      <div class="section-desc"><p>The S4 classes SPATA2 uses to abstract concepts of spatial multi-omic experiments.</p></div>


    </div><div class="section level2">




      <dl><dt>

          <code><a href="GroupAnnotation-class.html">GroupAnnotation-class</a></code> <code><a href="GroupAnnotation-class.html">GroupAnnotation</a></code>

        </dt>
        <dd>The <code>GroupAnnotation</code> - Class</dd>
      </dl><dl><dt>

          <code><a href="HistoImage-class.html">HistoImage-class</a></code> <code><a href="HistoImage-class.html">HistoImage</a></code>

        </dt>
        <dd>The <code>HistoImage</code> - Class</dd>
      </dl><dl><dt>

          <code><a href="ImageAnnotation-class.html">ImageAnnotation-class</a></code> <code><a href="ImageAnnotation-class.html">ImageAnnotation</a></code>

        </dt>
        <dd>The <code>ImageAnnotation</code> - Class</dd>
      </dl><dl><dt>

          <code><a href="MolecularAssay-class.html">MolecularAssay-class</a></code> <code><a href="MolecularAssay-class.html">MolecularAssay</a></code>

        </dt>
        <dd>The <code>MolecularAssay</code> - Class</dd>
      </dl><dl><dt>

          <code><a href="NumericAnnotation-class.html">NumericAnnotation-class</a></code> <code><a href="NumericAnnotation-class.html">NumericAnnotation</a></code>

        </dt>
        <dd>The <code>NumericAnnotation</code> - Class</dd>
      </dl><dl><dt>

          <code><a href="SPATA2.html">SPATA2-class</a></code> <code><a href="SPATA2.html">SPATA2</a></code>

        </dt>
        <dd>The <code>SPATA2</code> - class</dd>
      </dl><dl><dt>

          <code><a href="SpatialAnnotation-class.html">SpatialAnnotation-class</a></code> <code><a href="SpatialAnnotation-class.html">SpatialAnnotation</a></code>

        </dt>
        <dd>The <code>SpatialAnnotation</code> - Class</dd>
      </dl><dl><dt>

          <code><a href="SpatialAnnotationScreening-class.html">SpatialAnnotationScreening-class</a></code> <code><a href="SpatialAnnotationScreening-class.html">SpatialAnnotationScreening</a></code>

        </dt>
        <dd>The <code>SpatialAnnotationScreening</code> - Class</dd>
      </dl><dl><dt>

          <code><a href="SpatialData-class.html">SpatialData-class</a></code> <code><a href="SpatialData-class.html">SpatialData</a></code>

        </dt>
        <dd>The <code>SpatialData</code> - Class</dd>
      </dl><dl><dt>

          <code><a href="SpatialGradientScreening-class.html">SpatialGradientScreening-class</a></code> <code><a href="SpatialGradientScreening-class.html">SpatialGradientScreening</a></code>

        </dt>
        <dd>The <code>SpatialGradientScreening</code> - Class</dd>
      </dl><dl><dt>

          <code><a href="SpatialMethod-class.html">SpatialMethod-class</a></code> <code><a href="SpatialMethod-class.html">SpatialMethod</a></code>

        </dt>
        <dd>The <code>SpatialMethod</code> - Class</dd>
      </dl><dl><dt>

          <code><a href="SpatialTrajectory-class.html">SpatialTrajectory-class</a></code> <code><a href="SpatialTrajectory-class.html">SpatialTrajectory</a></code>

        </dt>
        <dd>The <code>SpatialTrajectory</code> - Class</dd>
      </dl><dl><dt>

          <code><a href="SpatialTrajectoryScreening-class.html">SpatialTrajectoryScreening-class</a></code> <code><a href="SpatialTrajectoryScreening-class.html">SpatialTrajectoryScreening</a></code>

        </dt>
        <dd>The <code>SpatialTrajectoryScreening</code> - Class</dd>
      </dl></div><div class="section level2">
      <h2 id="platforms">Platforms<a class="anchor" aria-label="anchor" href="#platforms"></a></h2>

      <div class="section-desc"><p>Variables that have been defined as objects of class <code>SpatialMethod</code> to store information related to the spatial platforms.</p></div>


    </div><div class="section level2">




      <dl><dt>

          <code><a href="MERFISH.html">MERFISH</a></code>

        </dt>
        <dd>MERFISH Spatial Method</dd>
      </dl><dl><dt>

          <code><a href="VisiumSmall.html">VisiumSmall</a></code>

        </dt>
        <dd>VisiumSmall Spatial Method</dd>
      </dl><dl><dt>

          <code><a href="VisiumHD.html">VisiumHD</a></code>

        </dt>
        <dd>VisiumHD Spatial Method</dd>
      </dl><dl><dt>

          <code><a href="VisiumLarge.html">VisiumLarge</a></code>

        </dt>
        <dd>VisiumLarge Spatial Method</dd>
      </dl><dl><dt>

          <code><a href="SlideSeqV1.html">SlideSeqV1</a></code>

        </dt>
        <dd>SlideSeqV1 Spatial Method</dd>
      </dl><dl><dt>

          <code><a href="Xenium.html">Xenium</a></code>

        </dt>
        <dd>Xenium Spatial Method</dd>
      </dl></div><div class="section level2">
      <h2 id="visualization">Visualization<a class="anchor" aria-label="anchor" href="#visualization"></a></h2>




    </div><div class="section level2">

      <h3 id="copy-number-variation">Copy Number Variation<a class="anchor" aria-label="anchor" href="#copy-number-variation"></a></h3>



    </div><div class="section level2">




      <dl><dt>

          <code><a href="plotCnvHeatmap.html">plotCnvHeatmap()</a></code>

        </dt>
        <dd>Plot CNV Heatmap</dd>
      </dl><dl><dt>

          <code><a href="plotCnvLineplot.html">plotCnvLineplot()</a></code>

        </dt>
        <dd>Plot CNV Lineplot</dd>
      </dl></div><div class="section level2">

      <h3 id="dea-gsea">DEA &amp; GSEA<a class="anchor" aria-label="anchor" href="#dea-gsea"></a></h3>
      <div class="section-desc"><p>Visualize the results of differential expression analysis and gene set enrichment analysis.</p></div>


    </div><div class="section level2">




      <dl><dt>

          <code><a href="plotDeaDotPlot.html">plotDeaDotPlot()</a></code>

        </dt>
        <dd>Plot DEA results via dot plots</dd>
      </dl><dl><dt>

          <code><a href="plotDeaHeatmap.html">plotDeaHeatmap()</a></code>

        </dt>
        <dd>Plot DEA results via heatmaps</dd>
      </dl><dl><dt>

          <code><a href="plotDeaVolcano.html">plotDeaVolcano()</a></code> <code><a href="plotDeaVolcano.html">plotDeaVolcano1v1()</a></code>

        </dt>
        <dd>Plot DEA results via volcano plots</dd>
      </dl><dl><dt>

          <code><a href="plotGseaDotPlot.html">plotGseaDotPlot()</a></code>

        </dt>
        <dd>Plot GSEA results via dot plots</dd>
      </dl></div><div class="section level2">

      <h3 id="dimensional-reduction-1">Dimensional Reduction<a class="anchor" aria-label="anchor" href="#dimensional-reduction-1"></a></h3>
      <div class="section-desc"><p>Visualize the barcode spots in a reduced dimension fashion.</p></div>


    </div><div class="section level2">




      <dl><dt>

          <code><a href="plotUMAP.html">plotPCA()</a></code> <code><a href="plotUMAP.html">plotPca()</a></code> <code><a href="plotUMAP.html">plotTSNE()</a></code> <code><a href="plotUMAP.html">plotTsne()</a></code> <code><a href="plotUMAP.html">plotTsneComparison()</a></code> <code><a href="plotUMAP.html">plotUMAP()</a></code> <code><a href="plotUMAP.html">plotUmap()</a></code> <code><a href="plotUMAP.html">plotUmapComparison()</a></code>

        </dt>
        <dd>Plot dimensional reduction</dd>
      </dl><dl><dt>

          <code><a href="plotPcaElbow.html">plotPcaElbow()</a></code>

        </dt>
        <dd>Plot PCA Elbow Plot</dd>
      </dl><dl><dt>

          <code><a href="plotPcaVariation.html">plotPcaVariation()</a></code>

        </dt>
        <dd>Plot PCA Variation</dd>
      </dl></div><div class="section level2">

      <h3 id="ggplayer-and-other-gg-objects">ggpLayer and other gg objects<a class="anchor" aria-label="anchor" href="#ggplayer-and-other-gg-objects"></a></h3>
      <div class="section-desc"><p>Add ggroto objects to ggplot2 plots.</p></div>


    </div><div class="section level2">




      <dl><dt>

          <code><a href="ggpInit.html">ggpInit()</a></code>

        </dt>
        <dd>Initiate ggplot2 layering</dd>
      </dl><dl><dt>

          <code><a href="ggpLayerAxesClean.html">ggpLayerAxesClean()</a></code>

        </dt>
        <dd>Display clean axes</dd>
      </dl><dl><dt>

          <code><a href="ggpLayerAxesSI.html">ggpLayerAxesSI()</a></code>

        </dt>
        <dd>Display axes with SI units of length</dd>
      </dl><dl><dt>

          <code><a href="ggpLayerCaptureArea.html">ggpLayerCaptureArea()</a></code>

        </dt>
        <dd>Add capture area to surface plot</dd>
      </dl><dl><dt>

          <code><a href="ggpLayerColorGroupScale.html">ggpLayerColorGroupScale()</a></code>

        </dt>
        <dd>Add group specific color spectrum</dd>
      </dl><dl><dt>

          <code><a href="ggpLayerExprEstimatesSAS.html">ggpLayerExprEstimatesSAS()</a></code>

        </dt>
        <dd>Add SAS expression estimates</dd>
      </dl><dl><dt>

          <code><a href="ggpLayerFrame.html">ggpLayerFrame()</a></code> <code><a href="ggpLayerFrame.html">ggpLayerFixFrame()</a></code>

        </dt>
        <dd>Fix ggplot frame</dd>
      </dl><dl><dt>

          <code><a href="ggpLayerFrameByCoords.html">ggpLayerFrameByCoords()</a></code> <code><a href="ggpLayerFrameByCoords.html">ggpLayerFrameByImage()</a></code>

        </dt>
        <dd>Set plot limits</dd>
      </dl><dl><dt>

          <code><a href="ggpLayerGroupOutline.html">ggpLayerGroupOutline()</a></code>

        </dt>
        <dd>Add group outline</dd>
      </dl><dl><dt>

          <code><a href="ggpLayerHorizonSAS.html">ggpLayerHorizonSAS()</a></code>

        </dt>
        <dd>Add SAS screening horizon</dd>
      </dl><dl><dt>

          <code><a href="ggpLayerImage.html">ggpLayerImage()</a></code>

        </dt>
        <dd>Add histology image</dd>
      </dl><dl><dt>

          <code><a href="ggpLayerPoints.html">ggpLayerPoints()</a></code>

        </dt>
        <dd>Add the observations to the surface plot</dd>
      </dl><dl><dt>

          <code><a href="ggpLayerRect.html">ggpLayerRect()</a></code>

        </dt>
        <dd>Add a rectangular to the plot</dd>
      </dl><dl><dt>

          <code><a href="ggpLayerScaleBarSI.html">ggpLayerScaleBarSI()</a></code>

        </dt>
        <dd>Add a scale bar in SI units</dd>
      </dl><dl><dt>

          <code><a href="ggpLayerScreeningDirectionSAS.html">ggpLayerScreeningDirectionSAS()</a></code>

        </dt>
        <dd>Add SAS screening direction</dd>
      </dl><dl><dt>

          <code><a href="ggpLayerSpatAnnOutline.html">ggpLayerSpatAnnOutline()</a></code>

        </dt>
        <dd>Add outline of spatial annotations</dd>
      </dl><dl><dt>

          <code><a href="ggpLayerSpatAnnPointer.html">ggpLayerSpatAnnPointer()</a></code>

        </dt>
        <dd>Add pointer towards spatial annotations</dd>
      </dl><dl><dt>

          <code><a href="ggpLayerSpatAnnRect.html">ggpLayerSpatAnnRect()</a></code>

        </dt>
        <dd>Add a rectangular around an spatial annotation</dd>
      </dl><dl><dt>

          <code><a href="ggpLayerSpatialTrajectories.html">ggpLayerSpatialTrajectories()</a></code>

        </dt>
        <dd>Add spatial trajectories</dd>
      </dl><dl><dt>

          <code><a href="ggpLayerThemeCoords.html">ggpLayerThemeCoords()</a></code>

        </dt>
        <dd>Add coordinates theme</dd>
      </dl><dl><dt>

          <code><a href="ggpLayerTissueOutline.html">ggpLayerTissueOutline()</a></code>

        </dt>
        <dd>Add a hull outlining the tissue</dd>
      </dl><dl><dt>

          <code><a href="ggpLayerTrajectoryFrame.html">ggpLayerTrajectoryFrame()</a></code>

        </dt>
        <dd>Add the screening frame of a spatial trajectory</dd>
      </dl><dl><dt>

          <code><a href="ggpLayerZoom.html">ggpLayerZoom()</a></code>

        </dt>
        <dd>Set plot limits manually</dd>
      </dl><dl><dt>

          <code><a href="theme_lineplot_gradient.html">theme_lineplot_gradient()</a></code> <code><a href="theme_lineplot_gradient.html">theme_ridgeplot_gradient()</a></code> <code><a href="theme_lineplot_gradient.html">theme_trajectory_fit()</a></code> <code><a href="theme_lineplot_gradient.html">theme_transparent()</a></code>

        </dt>
        <dd>ggplot2 themes</dd>
      </dl><dl><dt>

          <code><a href="legendBottom.html">labsNone()</a></code> <code><a href="legendBottom.html">legendBottom()</a></code> <code><a href="legendBottom.html">legendColor()</a></code> <code><a href="legendBottom.html">legendLeft()</a></code> <code><a href="legendBottom.html">legendNone()</a></code> <code><a href="legendBottom.html">legendRight()</a></code> <code><a href="legendBottom.html">legendTop()</a></code>

        </dt>
        <dd>ggplot2 basic manipulation</dd>
      </dl></div><div class="section level2">

      <h3 id="geoms">geoms<a class="anchor" aria-label="anchor" href="#geoms"></a></h3>



    </div><div class="section level2">




      <dl><dt>

          <code><a href="geom_point_fixed.html">geom_point_fixed()</a></code>

        </dt>
        <dd>Points (fixed size ~ window ratio)</dd>
      </dl><dl><dt>

          <code><a href="geom_segment_fixed.html">geom_segment_fixed()</a></code>

        </dt>
        <dd>Segments (fixed size ~ window ratio)</dd>
      </dl><dl><dt>

          <code><a href="geom_text_fixed.html">geom_text_fixed()</a></code>

        </dt>
        <dd>Text (fixed size ~ window ratio)</dd>
      </dl></div><div class="section level2">

      <h3 id="image-1">Image<a class="anchor" aria-label="anchor" href="#image-1"></a></h3>



    </div><div class="section level2">




      <dl><dt>

          <code><a href="plotImage.html">plotImage()</a></code>

        </dt>
        <dd>Plot image with ggplot2</dd>
      </dl><dl><dt>

          <code><a href="plotImages.html">plotImages()</a></code>

        </dt>
        <dd>Plot histology images (ggplot2)</dd>
      </dl><dl><dt>

          <code><a href="plotImageBase.html">plotImageBase()</a></code>

        </dt>
        <dd>Plot image with R base plotting</dd>
      </dl><dl><dt>

          <code><a href="plotImageMask.html">plotImageMask()</a></code> <code><a href="plotImageMask.html">plotPixelContent()</a></code>

        </dt>
        <dd>Plot pixel content</dd>
      </dl></div><div class="section level2">

      <h3 id="miscellaneous-1">Miscellaneous<a class="anchor" aria-label="anchor" href="#miscellaneous-1"></a></h3>
      <div class="section-desc"><p>Visualize a variety of aspects.</p></div>


    </div><div class="section level2">




      <dl><dt>

          <code><a href="plotLoglik.html">plotLoglik()</a></code>

        </dt>
        <dd>Plot Bayes Space logliks</dd>
      </dl><dl><dt>

          <code><a href="plotScatterplot.html">plotScatterplot()</a></code>

        </dt>
        <dd>Plot numeric variables as a scatterplot</dd>
      </dl><dl><dt>

          <code><a href="showColors.html">showColors()</a></code> <code><a href="showColors.html">showColorPalettes()</a></code> <code><a href="showColors.html">showColorSpectra()</a></code>

        </dt>
        <dd>Show color palettes and spectra</dd>
      </dl><dl><dt>

          <code><a href="showModels.html">showModels()</a></code>

        </dt>
        <dd>Show spatial gradient screening models</dd>
      </dl></div><div class="section level2">

      <h3 id="spatial-gradient">Spatial Gradient<a class="anchor" aria-label="anchor" href="#spatial-gradient"></a></h3>



    </div><div class="section level2">




      <dl><dt>

          <code><a href="plotExprVsDistSA.html">plotExprVsDistSA()</a></code> <code><a href="plotExprVsDistSA.html">plotExprVsDistST()</a></code>

        </dt>
        <dd>Plot expression as a function of distance to a spatial references</dd>
      </dl><dl><dt>

          <code><a href="plotSpatialAnnotations.html">plotSpatialAnnotations()</a></code>

        </dt>
        <dd>Plot spatial annotations</dd>
      </dl><dl><dt>

          <code><a href="plotSpatialTrajectories.html">plotSpatialTrajectories()</a></code>

        </dt>
        <dd>Plot spatial trajectories</dd>
      </dl><dl><dt>

          <code><a href="plotSasBarplot.html">plotSasBarplot()</a></code>

        </dt>
        <dd>Plot SAS barplot</dd>
      </dl><dl><dt>

          <code><a href="plotSasHeatmap.html">plotSasHeatmap()</a></code>

        </dt>
        <dd>Plot SAS heatmap</dd>
      </dl><dl><dt>

          <code><a href="plotSasLineplot.html">plotSasLineplot()</a></code>

        </dt>
        <dd>Plot SAS lineplot</dd>
      </dl><dl><dt>

          <code><a href="plotSasRidgeplot.html">plotSasRidgeplot()</a></code>

        </dt>
        <dd>Plot SAS rideplot</dd>
      </dl><dl><dt>

          <code><a href="plotStsBarplot.html">plotStsBarplot()</a></code>

        </dt>
        <dd>Plot STS barplot</dd>
      </dl><dl><dt>

          <code><a href="plotStsHeatmap.html">plotStsHeatmap()</a></code>

        </dt>
        <dd>Plot STS heatmap</dd>
      </dl><dl><dt>

          <code><a href="plotStsLineplot.html">plotStsLineplot()</a></code> <code><a href="plotStsLineplot.html">plotStsRidgeplot()</a></code>

        </dt>
        <dd>Plot STS line- and ridgeplot</dd>
      </dl></div><div class="section level2">

      <h3 id="statistics">Statistics<a class="anchor" aria-label="anchor" href="#statistics"></a></h3>



    </div><div class="section level2">




      <dl><dt>

          <code><a href="plotBarchart.html">plotBarchart()</a></code>

        </dt>
        <dd>Plot distribution of discrete variables</dd>
      </dl><dl><dt>

          <code><a href="plotBoxplot.html">plotBoxplot()</a></code> <code><a href="plotBoxplot.html">plotDensityplot()</a></code> <code><a href="plotBoxplot.html">plotHistogram()</a></code> <code><a href="plotBoxplot.html">plotRidgeplot()</a></code> <code><a href="plotBoxplot.html">plotVioBoxplot()</a></code> <code><a href="plotBoxplot.html">plotViolinplot()</a></code>

        </dt>
        <dd>Plot numeric distribution and statistical tests</dd>
      </dl><dl><dt>

          <code><a href="plotMosaicplot.html">plotMosaicplot()</a></code>

        </dt>
        <dd>Plot mosaic plot</dd>
      </dl><dl><dt>

          <code><a href="plotRiverplot.html">plotRiverplot()</a></code>

        </dt>
        <dd>Plot riverplots</dd>
      </dl><dl><dt>

          <code><a href="plotStatisticsInteractive.html">plotStatisticsInteractive()</a></code>

        </dt>
        <dd>Plot distribution of variables interactively</dd>
      </dl></div><div class="section level2">

      <h3 id="surface-plots-d-data">Surface Plots &amp; 2D data<a class="anchor" aria-label="anchor" href="#surface-plots-d-data"></a></h3>



    </div><div class="section level2">




      <dl><dt>

          <code><a href="plotSurface.html">plotSurface()</a></code> <code><a href="plotSurface.html">plotSurfaceInteractive()</a></code>

        </dt>
        <dd>Plot the surface of the sample</dd>
      </dl><dl><dt>

          <code><a href="plotSurfaceBase.html">plotSurfaceBase()</a></code>

        </dt>
        <dd>Plot surface with R base plotting</dd>
      </dl><dl><dt>

          <code><a href="plotSurfaceComparison.html">plotSurfaceComparison()</a></code>

        </dt>
        <dd>Plot several surface plots colored by numeric variables</dd>
      </dl><dl><dt>

          <code><a href="plotSurfaceSAS.html">plotSurfaceSAS()</a></code>

        </dt>
        <dd>Plot screening area of SAS set up</dd>
      </dl><dl><dt>

          <code><a href="plotSurfaceSC.html">plotSurfaceSC()</a></code>

        </dt>
        <dd>Plot single cells on surface</dd>
      </dl><dl><dt>

          <code><a href="plotMolecules2D.html">plotMolecules2D()</a></code>

        </dt>
        <dd>Plot molecules in 2D space</dd>
      </dl></div><div class="section level2">
      <h2 id="miscellaneous-functions">Miscellaneous Functions<a class="anchor" aria-label="anchor" href="#miscellaneous-functions"></a></h2>




    </div><div class="section level2">




      <dl><dt>

          <code><a href="currentSpata2Version.html">currentSpata2Version()</a></code>

        </dt>
        <dd>The current version of SPATA2</dd>
      </dl><dl><dt>

          <code><a href="deprecatedInfo.html">deprecatedInfo()</a></code>

        </dt>
        <dd>Information about deprecated aspects</dd>
      </dl><dl><dt>

          <code><a href="is_outlier.html">is_outlier()</a></code>

        </dt>
        <dd>Check for outliers in a numeric vector</dd>
      </dl><dl><dt>

          <code><a href="model_ascending.html">model_ascending()</a></code> <code><a href="model_ascending.html">model_descending()</a></code>

        </dt>
        <dd>Generate model-based ascending or descending sequence</dd>
      </dl><dl><dt>

          <code><a href="model_peak.html">model_peak()</a></code> <code><a href="model_peak.html">model_trough()</a></code>

        </dt>
        <dd>Model a peaking pattern</dd>
      </dl><dl><dt>

          <code><a href="recSgsRes.html">recDbscanEps()</a></code> <code><a href="recSgsRes.html">recDbscanMinPts()</a></code> <code><a href="recSgsRes.html">recSgsRes()</a></code>

        </dt>
        <dd>Platform dependent input recommendations</dd>
      </dl><dl><dt>

          <code><a href="read_coords.html">read_coords()</a></code> <code><a href="read_coords.html">read_coords_merfish()</a></code> <code><a href="read_coords.html">read_coords_slide_seq_v1()</a></code> <code><a href="read_coords.html">read_coords_visium()</a></code> <code><a href="read_coords.html">read_coords_xenium()</a></code>

        </dt>
        <dd>Read coordinate data.frames</dd>
      </dl><dl><dt>

          <code><a href="read_matrix_from_folder.html">read_matrix_from_folder()</a></code>

        </dt>
        <dd>Read Matrix from Folder</dd>
      </dl><dl><dt>

          <code><a href="transform_coords.html">transform_coords()</a></code>

        </dt>
        <dd>Transform coordinates</dd>
      </dl><dl><dt>

          <code><a href="transform_image.html">transform_image()</a></code>

        </dt>
        <dd>Transform image</dd>
      </dl></div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Jan Kueckelhaus, Dieter-Henrik Heiland.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.0.</p>
</div>

    </footer></div>





  </body></html>

