<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Function reference â€¢ SPATA2</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Function reference"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-index">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">SPATA2</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">3.0.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/adding-data.html">Adding Data to SPATA2 objects</a>
    </li>
    <li>
      <a href="../articles/clustering.html">Clustering</a>
    </li>
    <li>
      <a href="../articles/cnv.html">Copy Number Variations (CNV)</a>
    </li>
    <li>
      <a href="../articles/creating-spatial-annotations.html">Creating Spatial Annotations</a>
    </li>
    <li>
      <a href="../articles/creating-spatial-trajectories.html">Spatial Trajectories</a>
    </li>
    <li>
      <a href="../articles/dea.html">Differential Expression Analysis (DEA)</a>
    </li>
    <li>
      <a href="../articles/gsea.html">Gene Set Enrichment Analysis</a>
    </li>
    <li>
      <a href="../articles/image-handling.html">Image Handling</a>
    </li>
    <li>
      <a href="../articles/initiation-and-preprocessing-customized.html">Initiation of individual SPATA2 objects</a>
    </li>
    <li>
      <a href="../articles/initiation-and-preprocessing-visium.html">Initiation &amp; Processing: Visium</a>
    </li>
    <li>
      <a href="../articles/newslog.html">Newslog</a>
    </li>
    <li>
      <a href="../articles/spatial-annotation-screening.html">Spatial Annotation Screening</a>
    </li>
    <li>
      <a href="../articles/spatial-measures.html">Spatial measures</a>
    </li>
    <li>
      <a href="../articles/spatial-segmentation.html">Spatial Segmentation</a>
    </li>
    <li>
      <a href="../articles/spatial-trajectory-screening.html">Spatial Trajectory Screening</a>
    </li>
    <li>
      <a href="../articles/surface-plotting.html">Surface Plotting</a>
    </li>
    <li>
      <a href="../articles/using-spatial-annotations.html">Using Spatial Annotations</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"></ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">
    <div class="page-header">
      <h1>Reference</h1>
    </div>

    <table class="ref-index"><colgroup><col class="alias"><col class="title"></colgroup><tbody><tr><th colspan="2">
          <h2 id="all-functions">All functions <a href="#all-functions" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="GroupAnnotation-class.html">GroupAnnotation-class</a></code> <code><a href="GroupAnnotation-class.html">GroupAnnotation</a></code> </p>
        </td>
        <td><p>The <code>GroupAnnotation</code> - Class</p></td>
      </tr><tr><td>
          <p><code><a href="HistoImage-class.html">HistoImage-class</a></code> <code><a href="HistoImage-class.html">HistoImage</a></code> </p>
        </td>
        <td><p>The <code>HistoImage</code> - Class</p></td>
      </tr><tr><td>
          <p><code><a href="HistologyImage-class.html">HistologyImage-class</a></code> <code><a href="HistologyImage-class.html">HistologyImage</a></code> </p>
        </td>
        <td><p>The <code>HistologyImage</code> - Class</p></td>
      </tr><tr><td>
          <p><code><a href="ImageAnnotation-class.html">ImageAnnotation-class</a></code> <code><a href="ImageAnnotation-class.html">ImageAnnotation</a></code> </p>
        </td>
        <td><p>The <code>ImageAnnotation</code> - Class</p></td>
      </tr><tr><td>
          <p><code><a href="MolecularAssay-class.html">MolecularAssay-class</a></code> <code><a href="MolecularAssay-class.html">MolecularAssay</a></code> </p>
        </td>
        <td><p>The <code>MolecularAssay</code> - Class</p></td>
      </tr><tr><td>
          <p><code><a href="NumericAnnotation-class.html">NumericAnnotation-class</a></code> <code><a href="NumericAnnotation-class.html">NumericAnnotation</a></code> </p>
        </td>
        <td><p>The <code>NumericAnnotation</code> - Class</p></td>
      </tr><tr><td>
          <p><code><a href="SPATA2.html">SPATA2-class</a></code> <code><a href="SPATA2.html">SPATA2</a></code> </p>
        </td>
        <td><p>The <code>SPATA2</code> - class</p></td>
      </tr><tr><td>
          <p><code><a href="SpatialAnnotation-class.html">SpatialAnnotation-class</a></code> <code><a href="SpatialAnnotation-class.html">SpatialAnnotation</a></code> </p>
        </td>
        <td><p>The <code>SpatialAnnotation</code> - Class</p></td>
      </tr><tr><td>
          <p><code><a href="SpatialAnnotationScreening-class.html">SpatialAnnotationScreening-class</a></code> <code><a href="SpatialAnnotationScreening-class.html">SpatialAnnotationScreening</a></code> </p>
        </td>
        <td><p>The <code>SpatialAnnotationScreening</code> - Class</p></td>
      </tr><tr><td>
          <p><code><a href="SpatialData-class.html">SpatialData-class</a></code> <code><a href="SpatialData-class.html">SpatialData</a></code> </p>
        </td>
        <td><p>The <code>SpatialData</code> - Class</p></td>
      </tr><tr><td>
          <p><code><a href="SpatialGradientScreening-class.html">SpatialGradientScreening-class</a></code> <code><a href="SpatialGradientScreening-class.html">SpatialGradientScreening</a></code> </p>
        </td>
        <td><p>The <code>SpatialGradientScreening</code> - Class</p></td>
      </tr><tr><td>
          <p><code><a href="SpatialMethod-class.html">SpatialMethod-class</a></code> <code><a href="SpatialMethod-class.html">SpatialMethod</a></code> </p>
        </td>
        <td><p>The <code>SpatialMethod</code> - Class</p></td>
      </tr><tr><td>
          <p><code><a href="SpatialTrajectory-class.html">SpatialTrajectory-class</a></code> <code><a href="SpatialTrajectory-class.html">SpatialTrajectory</a></code> </p>
        </td>
        <td><p>The <code>SpatialTrajectory</code> - Class</p></td>
      </tr><tr><td>
          <p><code><a href="SpatialTrajectoryScreening-class.html">SpatialTrajectoryScreening-class</a></code> <code><a href="SpatialTrajectoryScreening-class.html">SpatialTrajectoryScreening</a></code> </p>
        </td>
        <td><p>The <code>SpatialTrajectoryScreening</code> - Class</p></td>
      </tr><tr><td>
          <p><code><a href="Trajectory-class.html">Trajectory-class</a></code> <code><a href="Trajectory-class.html">Trajectory</a></code> </p>
        </td>
        <td><p>The <code>Trajectory</code> - Class</p></td>
      </tr><tr><td>
          <p><code><a href="activateAssay.html">activateAssay()</a></code> <code><a href="activateAssay.html">activeAssay()</a></code> </p>
        </td>
        <td><p>Default assay name</p></td>
      </tr><tr><td>
          <p><code><a href="activateImage.html">activeImage()</a></code> <code><a href="activateImage.html">activateImage()</a></code> <code><a href="activateImage.html">activateImageInt()</a></code> </p>
        </td>
        <td><p>Default image name</p></td>
      </tr><tr><td>
          <p><code><a href="activateMatrix.html">activateMatrix()</a></code> <code><a href="activateMatrix.html">activeMatrix()</a></code> </p>
        </td>
        <td><p>Default matrix name</p></td>
      </tr><tr><td>
          <p><code><a href="addFeatures.html">addFeatures()</a></code> </p>
        </td>
        <td><p>Add meta features</p></td>
      </tr><tr><td>
          <p><code><a href="addGeneSet.html">addGeneSet()</a></code> <code><a href="addGeneSet.html">addGeneSetsInteractive()</a></code> </p>
        </td>
        <td><p>Add a new gene set</p></td>
      </tr><tr><td>
          <p><code><a href="addInnerBorder.html">addInnerBorder()</a></code> </p>
        </td>
        <td><p>Add holes to spatial annotations</p></td>
      </tr><tr><td>
          <p><code><a href="addProcessedMatrix.html">addProcessedMatrix()</a></code> <code><a href="addProcessedMatrix.html">addExpressionMatrix()</a></code> </p>
        </td>
        <td><p>Add a processed matrix</p></td>
      </tr><tr><td>
          <p><code><a href="addSpatialAnnotation.html">addSpatialAnnotation()</a></code> </p>
        </td>
        <td><p>Add an spatial annotation manually</p></td>
      </tr><tr><td>
          <p><code><a href="add_dbscan_variable.html">add_dbscan_variable()</a></code> <code><a href="add_dbscan_variable.html">add_tissue_section_variable()</a></code> </p>
        </td>
        <td><p>Add tissue section variable</p></td>
      </tr><tr><td>
          <p><code><a href="add_edge_variable.html">add_edge_variable()</a></code> </p>
        </td>
        <td><p>Add edge variable</p></td>
      </tr><tr><td>
          <p><code><a href="add_grid_variable.html">add_grid_variable()</a></code> </p>
        </td>
        <td><p>Add grid variable to coordinate data.rame</p></td>
      </tr><tr><td>
          <p><code><a href="are_all_area_or_dist.html">are_all_area_or_dist()</a></code> <code><a href="are_all_area_or_dist.html">are_all_dist()</a></code> </p>
        </td>
        <td><p>Test area or distance input</p></td>
      </tr><tr><td>
          <p><code><a href="asGiotto.html">asGiotto()</a></code> </p>
        </td>
        <td><p>Transform <code>SPATA2</code> object to <code>Giotto</code></p></td>
      </tr><tr><td>
          <p><code><a href="asSPATA2.html">asSPATA2()</a></code> </p>
        </td>
        <td><p>Transform to SPATA2 object object</p></td>
      </tr><tr><td>
          <p><code><a href="asSeurat.html">asSeurat()</a></code> </p>
        </td>
        <td><p>Transform <code>SPATA2</code> to <code>Seurat</code></p></td>
      </tr><tr><td>
          <p><code><a href="asSingleCellExperiment.html">asSingleCellExperiment()</a></code> </p>
        </td>
        <td><p>Transform to <code>SingleCellExperiment</code></p></td>
      </tr><tr><td>
          <p><code><a href="asSpatialTrajectory.html">asSpatialTrajectory()</a></code> </p>
        </td>
        <td><p>Transform to <code>SpatialTrajectory</code></p></td>
      </tr><tr><td>
          <p><code><a href="asSummarizedExperiment.html">asSummarizedExperiment()</a></code> </p>
        </td>
        <td><p>Transform to <code>SummarizedExperiment</code></p></td>
      </tr><tr><td>
          <p><code><a href="as_SPATA2_dist.html">as_SPATA2_dist()</a></code> </p>
        </td>
        <td><p>Distance transformation</p></td>
      </tr><tr><td>
          <p><code><a href="as_unit.html">as_meter()</a></code> <code><a href="as_unit.html">as_meter2()</a></code> <code><a href="as_unit.html">as_micrometer()</a></code> <code><a href="as_unit.html">as_micrometer2()</a></code> <code><a href="as_unit.html">as_millimeter()</a></code> <code><a href="as_unit.html">as_millimeter2()</a></code> <code><a href="as_unit.html">as_nanometer()</a></code> <code><a href="as_unit.html">as_nanometer2()</a></code> <code><a href="as_unit.html">as_pixel()</a></code> <code><a href="as_unit.html">as_unit()</a></code> <code><a href="as_unit.html">as_centimeter()</a></code> <code><a href="as_unit.html">as_centimeter2()</a></code> <code><a href="as_unit.html">as_decimeter()</a></code> <code><a href="as_unit.html">as_decimeter2()</a></code> </p>
        </td>
        <td><p>Transform distance and area values</p></td>
      </tr><tr><td>
          <p><code><a href="barcodesToSpatialAnnotation.html">barcodesToSpatialAnnotation()</a></code> </p>
        </td>
        <td><p>Create spatial annotations from a list of barcodes</p></td>
      </tr><tr><td>
          <p><code><a href="centerSpatialAnnotation.html">centerSpatialAnnotation()</a></code> </p>
        </td>
        <td><p>Center the borders of a spatial annotation</p></td>
      </tr><tr><td>
          <p><code><a href="cnv_ref.html">cnv_ref</a></code> </p>
        </td>
        <td><p>A list of reference data for copy number variation analysis (CNV)</p></td>
      </tr><tr><td>
          <p><code><a href="cnv_regions_df.html">cnv_regions_df</a></code> </p>
        </td>
        <td><p>A data.frame necessary for cnv-analysis. Contains information about start and end points of chromosomes.</p></td>
      </tr><tr><td>
          <p><code><a href="computeCnvByChrArm.html">computeCnvByChrArm()</a></code> </p>
        </td>
        <td><p>Compute CNV by chromosome arm</p></td>
      </tr><tr><td>
          <p><code><a href="computeMetaFeatures.html">computeMetaFeatures()</a></code> </p>
        </td>
        <td><p>Compute meta features for molecular data</p></td>
      </tr><tr><td>
          <p><code><a href="computePixelScaleFactor.html">computePixelScaleFactor()</a></code> </p>
        </td>
        <td><p>Compute pixel scale factor</p></td>
      </tr><tr><td>
          <p><code><a href="compute_img_scale_fct.html">compute_img_scale_fct()</a></code> </p>
        </td>
        <td><p>Compute scale factor of two images</p></td>
      </tr><tr><td>
          <p><code><a href="containsAssay.html">containsAssay()</a></code> </p>
        </td>
        <td><p>Check availability of an assay</p></td>
      </tr><tr><td>
          <p><code><a href="containsCCD.html">containsCCD()</a></code> </p>
        </td>
        <td><p>Check availability of center to center distance</p></td>
      </tr><tr><td>
          <p><code><a href="containsCNV.html">containsCNV()</a></code> </p>
        </td>
        <td><p>Check availability CNV results</p></td>
      </tr><tr><td>
          <p><code><a href="containsCells.html">containsCells()</a></code> </p>
        </td>
        <td><p>Check availability of cells</p></td>
      </tr><tr><td>
          <p><code><a href="containsHistoImages.html">containsHistoImages()</a></code> </p>
        </td>
        <td><p>Check availability of image containers</p></td>
      </tr><tr><td>
          <p><code><a href="containsImage.html">containsImage()</a></code> </p>
        </td>
        <td><p>Check availability of an image</p></td>
      </tr><tr><td>
          <p><code><a href="containsInnerBorders.html">containsInnerBorders()</a></code> </p>
        </td>
        <td><p>Check for inner borders in a spatial annotation</p></td>
      </tr><tr><td>
          <p><code><a href="containsMethod.html">containsMethod()</a></code> </p>
        </td>
        <td><p>Check availability of specific methods</p></td>
      </tr><tr><td>
          <p><code><a href="containsPixelContent.html">containsPixelContent()</a></code> </p>
        </td>
        <td><p>Check availability pixel content</p></td>
      </tr><tr><td>
          <p><code><a href="containsScaleFactor.html">containsScaleFactor()</a></code> </p>
        </td>
        <td><p>Check availability of specific scale factors</p></td>
      </tr><tr><td>
          <p><code><a href="containsSectionVariable.html">containsSectionVariable()</a></code> </p>
        </td>
        <td><p>Check availability of section variable</p></td>
      </tr><tr><td>
          <p><code><a href="containsSpatialAnnotations.html">containsSpatialAnnotations()</a></code> </p>
        </td>
        <td><p>Check availability of spatial annotations</p></td>
      </tr><tr><td>
          <p><code><a href="containsSpatialData.html">containsSpatialData()</a></code> </p>
        </td>
        <td><p>Checks availability of <code>SpatialData</code> object</p></td>
      </tr><tr><td>
          <p><code><a href="containsSpatialOutliers.html">containsSpatialOutliers()</a></code> </p>
        </td>
        <td><p>Check if spatial outliers exist</p></td>
      </tr><tr><td>
          <p><code><a href="containsSpots.html">containsSpots()</a></code> </p>
        </td>
        <td><p>Check availability of spots</p></td>
      </tr><tr><td>
          <p><code><a href="containsTissueOutline.html">containsTissueOutline()</a></code> </p>
        </td>
        <td><p>Check availability of tissue outline</p></td>
      </tr><tr><td>
          <p><code><a href="createGroupAnnotations.html">createGroupAnnotations()</a></code> </p>
        </td>
        <td><p>Create spatial annotations from a group of data points</p></td>
      </tr><tr><td>
          <p><code><a href="createHistoImage.html">createHistoImage()</a></code> </p>
        </td>
        <td><p>Create an object of class <code>HistoImage</code></p></td>
      </tr><tr><td>
          <p><code><a href="createImageAnnotations.html">createImageAnnotations()</a></code> </p>
        </td>
        <td><p>Create spatial annotations based on histo-morphological features</p></td>
      </tr><tr><td>
          <p><code><a href="createMolecularAssay.html">createMolecularAssay()</a></code> </p>
        </td>
        <td><p>Add a molecular assay</p></td>
      </tr><tr><td>
          <p><code><a href="createNumericAnnotations.html">createNumericAnnotations()</a></code> </p>
        </td>
        <td><p>Create spatial annotations based on numeric values</p></td>
      </tr><tr><td>
          <p><code><a href="createSpatialData.html">createSpatialData()</a></code> <code><a href="createSpatialData.html">createSpatialDataMERFISH()</a></code> <code><a href="createSpatialData.html">createSpatialDataSlideSeqV1()</a></code> <code><a href="createSpatialData.html">createSpatialDataVisium()</a></code> <code><a href="createSpatialData.html">createSpatialDataXenium()</a></code> </p>
        </td>
        <td><p>Create an object of class <code>SpatialData</code></p></td>
      </tr><tr><td>
          <p><code><a href="createSpatialMethod.html">createSpatialMethod()</a></code> </p>
        </td>
        <td><p>Create an object of class <code>SpatialMethod</code></p></td>
      </tr><tr><td>
          <p><code><a href="createSpatialSegmentation.html">createSpatialSegmentation()</a></code> </p>
        </td>
        <td><p>Interactive sample segmentation</p></td>
      </tr><tr><td>
          <p><code><a href="createSpatialTrajectories.html">addSpatialTrajectory()</a></code> <code><a href="createSpatialTrajectories.html">createSpatialTrajectories()</a></code> </p>
        </td>
        <td><p>Create and add spatial trajectories</p></td>
      </tr><tr><td>
          <p><code><a href="create_model_df.html">create_model_df()</a></code> </p>
        </td>
        <td><p>Create model data.frame</p></td>
      </tr><tr><td>
          <p><code><a href="cropSpataObject.html">cropSpataObject()</a></code> </p>
        </td>
        <td><p>Subset SPATA2 object</p></td>
      </tr><tr><td>
          <p><code><a href="currentSpata2Version.html">currentSpata2Version()</a></code> </p>
        </td>
        <td><p>The current version of SPATA2</p></td>
      </tr><tr><td>
          <p><code><a href="deprecatedInfo.html">deprecatedInfo()</a></code> </p>
        </td>
        <td><p>Information about deprecated aspects</p></td>
      </tr><tr><td>
          <p><code><a href="dissolveGroups.html">dissolveGroups()</a></code> </p>
        </td>
        <td><p>Dissolve groups in a SPATA2 object</p></td>
      </tr><tr><td>
          <p><code><a href="distToEdge.html">distToEdge()</a></code> </p>
        </td>
        <td><p>Distance to edge of tissue section</p></td>
      </tr><tr><td>
          <p><code><a href="downloadFromPublication.html">downloadFromPublication()</a></code> </p>
        </td>
        <td><p>Download data from publications</p></td>
      </tr><tr><td>
          <p><code><a href="downloadRawData.html">downloadRawData()</a></code> </p>
        </td>
        <td><p>Download raw Visium output</p></td>
      </tr><tr><td>
          <p><code><a href="downloadSpataObject.html">downloadSpataObject()</a></code> <code><a href="downloadSpataObject.html">downloadSpataObjects()</a></code> </p>
        </td>
        <td><p>Download <code>spata2</code> objects</p></td>
      </tr><tr><td>
          <p><code><a href="expandSpatialAnnotation.html">expandSpatialAnnotation()</a></code> </p>
        </td>
        <td><p>Expand the outline of spatial annotations</p></td>
      </tr><tr><td>
          <p><code><a href="extract_unit.html">extract_unit()</a></code> </p>
        </td>
        <td><p>Extract distance units</p></td>
      </tr><tr><td>
          <p><code><a href="extract_value.html">extract_value()</a></code> </p>
        </td>
        <td><p>Extract distance value</p></td>
      </tr><tr><td>
          <p><code><a href="flipAll.html">flipAll()</a></code> <code><a href="flipAll.html">flipCoordinates()</a></code> <code><a href="flipAll.html">flipCoordsDf()</a></code> <code><a href="flipAll.html">flipImage()</a></code> <code><a href="flipAll.html">flipSpatialAnnotations()</a></code> <code><a href="flipAll.html">flipSpatialTrajectories()</a></code> </p>
        </td>
        <td><p>Mirror invert image and coordinates</p></td>
      </tr><tr><td>
          <p><code><a href="gene_pos_df.html">gene_pos_df</a></code> </p>
        </td>
        <td><p>A data.frame necessary for cnv-analysis. Contains information about the gene positions on chromosomes. Contains the following variables:</p></td>
      </tr><tr><td>
          <p><code><a href="geom_point_fixed.html">geom_point_fixed()</a></code> </p>
        </td>
        <td><p>Points (fixed size ~ window ratio)</p></td>
      </tr><tr><td>
          <p><code><a href="geom_segment_fixed.html">geom_segment_fixed()</a></code> </p>
        </td>
        <td><p>Segments (fixed size ~ window ratio)</p></td>
      </tr><tr><td>
          <p><code><a href="geom_text_fixed.html">geom_text_fixed()</a></code> </p>
        </td>
        <td><p>Text (fixed size ~ window ratio)</p></td>
      </tr><tr><td>
          <p><code><a href="getAssay.html">getAssay()</a></code> </p>
        </td>
        <td><p>Obtain molecular assay</p></td>
      </tr><tr><td>
          <p><code><a href="getAssayNames.html">getAssayNames()</a></code> </p>
        </td>
        <td><p>Obtain assay names</p></td>
      </tr><tr><td>
          <p><code><a href="getAutoencoderAssessment.html">getAutoencoderAssessment()</a></code> </p>
        </td>
        <td><p>Obtain information about the optimal neural network set up</p></td>
      </tr><tr><td>
          <p><code><a href="getAutoencoderSetUp.html">getAutoencoderSetUp()</a></code> </p>
        </td>
        <td><p>Obtain information on neural network</p></td>
      </tr><tr><td>
          <p><code><a href="getBackgroundColor.html">getBackgroundColor()</a></code> </p>
        </td>
        <td><p>Obtain background color</p></td>
      </tr><tr><td>
          <p><code><a href="getBarcodeSpotDistances.html">getBarcodeSpotDistances()</a></code> </p>
        </td>
        <td><p>Obtain distances between barcodes</p></td>
      </tr><tr><td>
          <p><code><a href="getBarcodes.html">getBarcodes()</a></code> </p>
        </td>
        <td><p>Obtain barcodes</p></td>
      </tr><tr><td>
          <p><code><a href="getBarcodesInPolygon.html">getBarcodesInPolygon()</a></code> <code><a href="getBarcodesInPolygon.html">getBarcodesInPolygonList()</a></code> </p>
        </td>
        <td><p>Obtain barcodes in polygon</p></td>
      </tr><tr><td>
          <p><code><a href="getCCD.html">getCCD()</a></code> </p>
        </td>
        <td><p>Obtain center to center distance</p></td>
      </tr><tr><td>
          <p><code><a href="getCaptureArea.html">getCaptureArea()</a></code> </p>
        </td>
        <td><p>Obtain capture area</p></td>
      </tr><tr><td>
          <p><code><a href="getChrRegionsDf.html">getChrRegionsDf()</a></code> </p>
        </td>
        <td><p>Obtain chromosome information</p></td>
      </tr><tr><td>
          <p><code><a href="getCnvFeatureNames.html">getCnvFeatureNames()</a></code> </p>
        </td>
        <td><p>Obtain features names under which cnv-analysis results are stored.</p></td>
      </tr><tr><td>
          <p><code><a href="getCnvGenesDf.html">getCnvGenesDf()</a></code> </p>
        </td>
        <td><p>Obtain CNV results by gene</p></td>
      </tr><tr><td>
          <p><code><a href="getCnvResults.html">getCnvResults()</a></code> </p>
        </td>
        <td><p>Obtain copy-number-variations results</p></td>
      </tr><tr><td>
          <p><code><a href="getCoordsCenter.html">getCoordsCenter()</a></code> </p>
        </td>
        <td><p>Obtain coordinate center</p></td>
      </tr><tr><td>
          <p><code><a href="getCoordsDf.html">getCoordsDf()</a></code> </p>
        </td>
        <td><p>Obtain coordinates</p></td>
      </tr><tr><td>
          <p><code><a href="getCoordsDfSA.html">getCoordsDfSA()</a></code> <code><a href="getCoordsDfSA.html">getCoordsDfST()</a></code> </p>
        </td>
        <td><p>Relate points to spatial annotations</p></td>
      </tr><tr><td>
          <p><code><a href="getCoordsMtr.html">getCoordsMtr()</a></code> </p>
        </td>
        <td><p>Obtain coordinates matrix</p></td>
      </tr><tr><td>
          <p><code><a href="getCoordsRange.html">getCoordsRange()</a></code> </p>
        </td>
        <td><p>Obtain coordinate range</p></td>
      </tr><tr><td>
          <p><code><a href="getCountMatrix.html">getCountMatrix()</a></code> </p>
        </td>
        <td><p>Obtain raw counts</p></td>
      </tr><tr><td>
          <p><code><a href="getDeaResultsDf.html">getDeaGenes()</a></code> <code><a href="getDeaResultsDf.html">getDeaResultsDf()</a></code> </p>
        </td>
        <td><p>Obtain DEA results</p></td>
      </tr><tr><td>
          <p><code><a href="getDefaultUnit.html">getDefaultUnit()</a></code> </p>
        </td>
        <td><p>Obtain default unit</p></td>
      </tr><tr><td>
          <p><code><a href="getDimRedDf.html">getDimRedDf()</a></code> <code><a href="getDimRedDf.html">getPcaDf()</a></code> <code><a href="getDimRedDf.html">getPcaMtr()</a></code> <code><a href="getDimRedDf.html">getTsneDf()</a></code> <code><a href="getDimRedDf.html">getUmapDf()</a></code> </p>
        </td>
        <td><p>Obtain dim red data.frame</p></td>
      </tr><tr><td>
          <p><code><a href="getFeatureNames.html">getFeatureNames()</a></code> </p>
        </td>
        <td><p>Obtain metadata column names</p></td>
      </tr><tr><td>
          <p><code><a href="getGeneMetaData.html">getGeneMetaData()</a></code> <code><a href="getGeneMetaData.html">getGeneMetaDf()</a></code> </p>
        </td>
        <td><p>Obtain gene meta data</p></td>
      </tr><tr><td>
          <p><code><a href="getGenePosDf.html">getGenePosDf()</a></code> </p>
        </td>
        <td><p>Obtain gene information</p></td>
      </tr><tr><td>
          <p><code><a href="getGeneSetDf.html">getGeneSetDf()</a></code> <code><a href="getGeneSetDf.html">getGeneSetList()</a></code> </p>
        </td>
        <td><p>Obtain gene sets</p></td>
      </tr><tr><td>
          <p><code><a href="getGeneSetOverview.html">getGeneSetOverview()</a></code> </p>
        </td>
        <td><p>Overview about the current gene sets</p></td>
      </tr><tr><td>
          <p><code><a href="getGeneSets.html">getGeneSets()</a></code> <code><a href="getGeneSets.html">getGeneSetsInteractive()</a></code> </p>
        </td>
        <td><p>Obtain gene set names</p></td>
      </tr><tr><td>
          <p><code><a href="getGroupNames.html">getGroupNames()</a></code> </p>
        </td>
        <td><p>Obtain group names a grouping variable contains</p></td>
      </tr><tr><td>
          <p><code><a href="getGroupingOptions.html">getGroupingOptions()</a></code> </p>
        </td>
        <td><p>Obtain variable names that group data points</p></td>
      </tr><tr><td>
          <p><code><a href="getGseaDf.html">getGseaDf()</a></code> </p>
        </td>
        <td><p>Obtain enrichment data.frame</p></td>
      </tr><tr><td>
          <p><code><a href="getGseaResults.html">getGseaResults()</a></code> </p>
        </td>
        <td><p>Obtain enrichment results</p></td>
      </tr><tr><td>
          <p><code><a href="getHistoImage.html">getHistoImage()</a></code> <code><a href="getHistoImage.html">getHistoImageRef()</a></code> </p>
        </td>
        <td><p>Obtain object of class <code>HistoImage</code></p></td>
      </tr><tr><td>
          <p><code><a href="getImage.html">getImage()</a></code> </p>
        </td>
        <td><p>Obtain object of class <code>Image</code></p></td>
      </tr><tr><td>
          <p><code><a href="getImageCenter.html">getImageCenter()</a></code> </p>
        </td>
        <td><p>Obtain image center</p></td>
      </tr><tr><td>
          <p><code><a href="getImageDf.html">getImageDf()</a></code> </p>
        </td>
        <td><p>Obtain image as a data.frame</p></td>
      </tr><tr><td>
          <p><code><a href="getImageDims.html">getImageDims()</a></code> <code><a href="getImageDims.html">getImageRange()</a></code> </p>
        </td>
        <td><p>Obtain image dimensions/ranges</p></td>
      </tr><tr><td>
          <p><code><a href="getImageDir.html">getImageDir()</a></code> <code><a href="getImageDir.html">getImageDirectories()</a></code> </p>
        </td>
        <td><p>Obtain directory to image</p></td>
      </tr><tr><td>
          <p><code><a href="getImageNames.html">getImageNames()</a></code> </p>
        </td>
        <td><p>Obtain names of registered <code>HistoImage</code> objects</p></td>
      </tr><tr><td>
          <p><code><a href="getImageRaster.html">getImageRaster()</a></code> <code><a href="getImageRaster.html">getImageRasterInfo()</a></code> </p>
        </td>
        <td><p>Obtain image raster-(information)</p></td>
      </tr><tr><td>
          <p><code><a href="getImageSectionsByBarcode.html">getImageSectionsByBarcode()</a></code> </p>
        </td>
        <td><p>Obtain image sections by barcode spot</p></td>
      </tr><tr><td>
          <p><code><a href="getImageTransformations.html">getImageTransformations()</a></code> </p>
        </td>
        <td><p>Obtain image transformation instructions</p></td>
      </tr><tr><td>
          <p><code><a href="getLogfileDf.html">getLogfileDf()</a></code> </p>
        </td>
        <td><p>Obtain object logfile data.frame</p></td>
      </tr><tr><td>
          <p><code><a href="getMatrix.html">getMatrix()</a></code> <code><a href="getMatrix.html">getMatrices()</a></code> </p>
        </td>
        <td><p>Obtain a data matrix</p></td>
      </tr><tr><td>
          <p><code><a href="getMatrixNames.html">getMatrixNames()</a></code> </p>
        </td>
        <td><p>Obtain matrix names</p></td>
      </tr><tr><td>
          <p><code><a href="getMetaDf.html">getMetaDf()</a></code> </p>
        </td>
        <td><p>Obtain meta data.frame</p></td>
      </tr><tr><td>
          <p><code><a href="getMethodSpecifics.html">getMethodSpecifics()</a></code> </p>
        </td>
        <td><p>Obtain method specifics</p></td>
      </tr><tr><td>
          <p><code><a href="getMoleculeTypeList.html">getMoleculeTypeList()</a></code> </p>
        </td>
        <td><p>Obtain a list of molecules</p></td>
      </tr><tr><td>
          <p><code><a href="getMolecules.html">getGenes()</a></code> <code><a href="getMolecules.html">getGenesInteractive()</a></code> <code><a href="getMolecules.html">getMolecules()</a></code> </p>
        </td>
        <td><p>Obtain molecule names</p></td>
      </tr><tr><td>
          <p><code><a href="getPixelDf.html">getPixelDf()</a></code> </p>
        </td>
        <td><p>Obtain pixel data.frame</p></td>
      </tr><tr><td>
          <p><code><a href="getPixelScaleFactor.html">getPixelScaleFactor()</a></code> </p>
        </td>
        <td><p>Obtain scale factor for pixel to SI conversion</p></td>
      </tr><tr><td>
          <p><code><a href="getProcessedMatrix.html">getProcessedMatrix()</a></code> </p>
        </td>
        <td><p>Obtain processed data matrix</p></td>
      </tr><tr><td>
          <p><code><a href="getProcessedMatrixNames.html">getProcessedMatrixNames()</a></code> </p>
        </td>
        <td><p>Obtain names of processed matrices</p></td>
      </tr><tr><td>
          <p><code><a href="getProjectionDf.html">getProjectionDf()</a></code> </p>
        </td>
        <td><p>Obtain trajectory projection</p></td>
      </tr><tr><td>
          <p><code><a href="getSampleName.html">getSampleName()</a></code> </p>
        </td>
        <td><p>Obtain name of <code>SPATA2</code> object</p></td>
      </tr><tr><td>
          <p><code><a href="getSasDf.html">getSasDf()</a></code> </p>
        </td>
        <td><p>Obtain spatial annotation screening data.frame</p></td>
      </tr><tr><td>
          <p><code><a href="getScaleFactor.html">getScaleFactor()</a></code> </p>
        </td>
        <td><p>Obtain scale factors</p></td>
      </tr><tr><td>
          <p><code><a href="getSgsResultsDf.html">getSgsResultsDf()</a></code> </p>
        </td>
        <td><p>Obtain spatial gradient screening results</p></td>
      </tr><tr><td>
          <p><code><a href="getSgsResultsVec.html">getSgsResultsVec()</a></code> </p>
        </td>
        <td><p>Obtain spatial gradient screening results</p></td>
      </tr><tr><td>
          <p><code><a href="getSignatureTypeList.html">getSignatureTypeList()</a></code> </p>
        </td>
        <td><p>Obtain a list of signatures</p></td>
      </tr><tr><td>
          <p><code><a href="getSignatures.html">getSignatures()</a></code> </p>
        </td>
        <td><p>Obtain molecular signatures</p></td>
      </tr><tr><td>
          <p><code><a href="getSpatAnnArea.html">getSpatAnnArea()</a></code> </p>
        </td>
        <td><p>Obtain area of spatial annotation</p></td>
      </tr><tr><td>
          <p><code><a href="getSpatAnnBarcodes.html">getSpatAnnBarcodes()</a></code> </p>
        </td>
        <td><p>Obtain barcodes by spatial annotations</p></td>
      </tr><tr><td>
          <p><code><a href="getSpatAnnCenter.html">getSpatAnnCenter()</a></code> <code><a href="getSpatAnnCenter.html">getSpatAnnCenters()</a></code> </p>
        </td>
        <td><p>Obtain center of an spatial annotation</p></td>
      </tr><tr><td>
          <p><code><a href="getSpatAnnCenterBcsp.html">getSpatAnnCenterBcsp()</a></code> </p>
        </td>
        <td><p>Obtain center data point</p></td>
      </tr><tr><td>
          <p><code><a href="getSpatAnnIds.html">getSpatAnnIds()</a></code> </p>
        </td>
        <td><p>Obtain IDs of spatial annotations</p></td>
      </tr><tr><td>
          <p><code><a href="getSpatAnnOutlineDf.html">getSpatAnnOutlineDf()</a></code> </p>
        </td>
        <td><p>Obtain the outline of a spatial annotation</p></td>
      </tr><tr><td>
          <p><code><a href="getSpatAnnRange.html">getSpatAnnRange()</a></code> </p>
        </td>
        <td><p>Obtain spatial annotations range</p></td>
      </tr><tr><td>
          <p><code><a href="getSpatAnnSf.html">getSpatAnnSf()</a></code> </p>
        </td>
        <td><p>Obtain the outline of a spatial anontation</p></td>
      </tr><tr><td>
          <p><code><a href="getSpatAnnTags.html">getSpatAnnTags()</a></code> </p>
        </td>
        <td><p>Obtain spatial annotation tags</p></td>
      </tr><tr><td>
          <p><code><a href="getSpatSegmVarNames.html">getSpatSegmVarNames()</a></code> </p>
        </td>
        <td><p>Obtain segmentation variable names</p></td>
      </tr><tr><td>
          <p><code><a href="getSpataDir.html">getSpataDir()</a></code> </p>
        </td>
        <td><p>Obtain SPATA2 object directory</p></td>
      </tr><tr><td>
          <p><code><a href="getSpatialAnnotation.html">getSpatialAnnotation()</a></code> </p>
        </td>
        <td><p>Obtain object of class <code>SpatialAnnotation</code></p></td>
      </tr><tr><td>
          <p><code><a href="getSpatialAnnotations.html">getSpatialAnnotations()</a></code> </p>
        </td>
        <td><p>Obtain list of <code>SpatialAnnotation</code>-objects</p></td>
      </tr><tr><td>
          <p><code><a href="getSpatialData.html">getSpatialData()</a></code> </p>
        </td>
        <td><p>Obtain object of class <code>SpatialData</code></p></td>
      </tr><tr><td>
          <p><code><a href="getSpatialMethod.html">getSpatialMethod()</a></code> </p>
        </td>
        <td><p>Obtain spatial method</p></td>
      </tr><tr><td>
          <p><code><a href="getSpatialTrajectory.html">getSpatialTrajectory()</a></code> <code><a href="getSpatialTrajectory.html">getSpatialTrajectories()</a></code> </p>
        </td>
        <td><p>Obtain spatial trajectories</p></td>
      </tr><tr><td>
          <p><code><a href="getSpatialTrajectoryIds.html">getSpatialTrajectoryIds()</a></code> </p>
        </td>
        <td><p>Obtain trajectory IDs</p></td>
      </tr><tr><td>
          <p><code><a href="getSpotSize.html">getSpotSize()</a></code> </p>
        </td>
        <td><p>Obtain spot size</p></td>
      </tr><tr><td>
          <p><code><a href="getStsDf.html">getStsDf()</a></code> </p>
        </td>
        <td><p>Obtain spatial trajectory screening data.frame</p></td>
      </tr><tr><td>
          <p><code><a href="getTissueArea.html">getTissueArea()</a></code> </p>
        </td>
        <td><p>Obtain tissue area size</p></td>
      </tr><tr><td>
          <p><code><a href="getTissueOutlineCentroid.html">getTissueOutlineCentroid()</a></code> </p>
        </td>
        <td><p>Obtain tissue outline centroid</p></td>
      </tr><tr><td>
          <p><code><a href="getTissueOutlineDf.html">getTissueOutlineDf()</a></code> </p>
        </td>
        <td><p>Obtain the outline of tissue sections</p></td>
      </tr><tr><td>
          <p><code><a href="getTissueSections.html">getTissueSections()</a></code> </p>
        </td>
        <td><p>Obtain the names of tissue sections</p></td>
      </tr><tr><td>
          <p><code><a href="getTrajectoryLength.html">getTrajectoryLength()</a></code> </p>
        </td>
        <td><p>Obtain length of spatial trajectory</p></td>
      </tr><tr><td>
          <p><code><a href="getTrajectorySegmentDf.html">getTrajectorySegmentDf()</a></code> </p>
        </td>
        <td><p>Obtain trajectory course</p></td>
      </tr><tr><td>
          <p><code><a href="getTrajectoryWidth.html">getTrajectoryWidth()</a></code> </p>
        </td>
        <td><p>Obtain trajectory width</p></td>
      </tr><tr><td>
          <p><code><a href="getVarTypeList.html">getVarTypeList()</a></code> </p>
        </td>
        <td><p>Get variable type list</p></td>
      </tr><tr><td>
          <p><code><a href="getVariableNames.html">getVariableNames()</a></code> </p>
        </td>
        <td><p>Obtain variable names</p></td>
      </tr><tr><td>
          <p><code><a href="ggpInit.html">ggpInit()</a></code> </p>
        </td>
        <td><p>Initiate ggplot2 layering</p></td>
      </tr><tr><td>
          <p><code><a href="ggpLayerAxesClean.html">ggpLayerAxesClean()</a></code> </p>
        </td>
        <td><p>Display clean axes</p></td>
      </tr><tr><td>
          <p><code><a href="ggpLayerAxesSI.html">ggpLayerAxesSI()</a></code> </p>
        </td>
        <td><p>Display axes with SI units of length</p></td>
      </tr><tr><td>
          <p><code><a href="ggpLayerCaptureArea.html">ggpLayerCaptureArea()</a></code> </p>
        </td>
        <td><p>Add capture area to surface plot</p></td>
      </tr><tr><td>
          <p><code><a href="ggpLayerColorGroupScale.html">ggpLayerColorGroupScale()</a></code> </p>
        </td>
        <td><p>Add group specific color spectrum</p></td>
      </tr><tr><td>
          <p><code><a href="ggpLayerExprEstimatesSAS.html">ggpLayerExprEstimatesSAS()</a></code> </p>
        </td>
        <td><p>Add SAS expression estimates</p></td>
      </tr><tr><td>
          <p><code><a href="ggpLayerFrame.html">ggpLayerFrame()</a></code> <code><a href="ggpLayerFrame.html">ggpLayerFixFrame()</a></code> </p>
        </td>
        <td><p>Fix ggplot frame</p></td>
      </tr><tr><td>
          <p><code><a href="ggpLayerFrameByCoords.html">ggpLayerFrameByCoords()</a></code> <code><a href="ggpLayerFrameByCoords.html">ggpLayerFrameByImage()</a></code> </p>
        </td>
        <td><p>Set plot limits</p></td>
      </tr><tr><td>
          <p><code><a href="ggpLayerGroupOutline.html">ggpLayerGroupOutline()</a></code> </p>
        </td>
        <td><p>Add group outline</p></td>
      </tr><tr><td>
          <p><code><a href="ggpLayerHorizonSAS.html">ggpLayerHorizonSAS()</a></code> </p>
        </td>
        <td><p>Add SAS screening horizon</p></td>
      </tr><tr><td>
          <p><code><a href="ggpLayerImage.html">ggpLayerImage()</a></code> </p>
        </td>
        <td><p>Add histology image</p></td>
      </tr><tr><td>
          <p><code><a href="ggpLayerPoints.html">ggpLayerPoints()</a></code> </p>
        </td>
        <td><p>Add the observations to the surface plot</p></td>
      </tr><tr><td>
          <p><code><a href="ggpLayerRect.html">ggpLayerRect()</a></code> </p>
        </td>
        <td><p>Add a rectangular to the plot</p></td>
      </tr><tr><td>
          <p><code><a href="ggpLayerScaleBarSI.html">ggpLayerScaleBarSI()</a></code> </p>
        </td>
        <td><p>Add a scale bar in SI units</p></td>
      </tr><tr><td>
          <p><code><a href="ggpLayerScreeningDirectionSAS.html">ggpLayerScreeningDirectionSAS()</a></code> </p>
        </td>
        <td><p>Add SAS screening direction</p></td>
      </tr><tr><td>
          <p><code><a href="ggpLayerSpatAnnOutline.html">ggpLayerSpatAnnOutline()</a></code> </p>
        </td>
        <td><p>Add outline of spatial annotations</p></td>
      </tr><tr><td>
          <p><code><a href="ggpLayerSpatAnnPointer.html">ggpLayerSpatAnnPointer()</a></code> </p>
        </td>
        <td><p>Add pointer towards spatial annotations</p></td>
      </tr><tr><td>
          <p><code><a href="ggpLayerSpatAnnRect.html">ggpLayerSpatAnnRect()</a></code> </p>
        </td>
        <td><p>Add a rectangular around an spatial annotation</p></td>
      </tr><tr><td>
          <p><code><a href="ggpLayerSpatialTrajectories.html">ggpLayerSpatialTrajectories()</a></code> </p>
        </td>
        <td><p>Add spatial trajectories</p></td>
      </tr><tr><td>
          <p><code><a href="ggpLayerThemeCoords.html">ggpLayerThemeCoords()</a></code> </p>
        </td>
        <td><p>Add coordinates theme</p></td>
      </tr><tr><td>
          <p><code><a href="ggpLayerTissueOutline.html">ggpLayerTissueOutline()</a></code> </p>
        </td>
        <td><p>Add a hull outlining the tissue</p></td>
      </tr><tr><td>
          <p><code><a href="ggpLayerTrajectoryFrame.html">ggpLayerTrajectoryFrame()</a></code> </p>
        </td>
        <td><p>Add the screening frame of a spatial trajectory</p></td>
      </tr><tr><td>
          <p><code><a href="ggpLayerZoom.html">ggpLayerZoom()</a></code> </p>
        </td>
        <td><p>Set plot limits manually</p></td>
      </tr><tr><td>
          <p><code><a href="identifyPixelContent.html">identifyPixelContent()</a></code> </p>
        </td>
        <td><p>Identify pixel content</p></td>
      </tr><tr><td>
          <p><code><a href="identifySpatialOutliers.html">identifySpatialOutliers()</a></code> </p>
        </td>
        <td><p>Identify spatial outliers</p></td>
      </tr><tr><td>
          <p><code><a href="identifyTissueOutline.html">identifyTissueOutline()</a></code> </p>
        </td>
        <td><p>Identify tissue outline</p></td>
      </tr><tr><td>
          <p><code><a href="identify_obs_in_polygon.html">identify_obs_in_polygon()</a></code> <code><a href="identify_obs_in_polygon.html">identify_obs_in_spat_ann()</a></code> </p>
        </td>
        <td><p>Identify observations inside a polygon</p></td>
      </tr><tr><td>
          <p><code><a href="initiateSpataObject.html">initiateSpataObject()</a></code> </p>
        </td>
        <td><p>Initiate a <code>SPATA2</code> object</p></td>
      </tr><tr><td>
          <p><code><a href="initiateSpataObjectMERFISH.html">initiateSpataObjectMERFISH()</a></code> </p>
        </td>
        <td><p>Initiate a <code>SPATA2</code> object from platform MERFISH</p></td>
      </tr><tr><td>
          <p><code><a href="initiateSpataObjectSlideSeqV1.html">initiateSpataObjectSlideSeqV1()</a></code> </p>
        </td>
        <td><p>Initiate a <code>SPATA2</code> object from platform SlideSeq</p></td>
      </tr><tr><td>
          <p><code><a href="initiateSpataObjectVisium.html">initiateSpataObjectVisium()</a></code> </p>
        </td>
        <td><p>Initiate <code>SPATA2</code> object from platform Visium</p></td>
      </tr><tr><td>
          <p><code><a href="initiateSpataObjectXenium.html">initiateSpataObjectXenium()</a></code> </p>
        </td>
        <td><p>Initiate <code>SPATA2</code> object from platform Xenium</p></td>
      </tr><tr><td>
          <p><code><a href="isDirVisium.html">isDirVisium()</a></code> </p>
        </td>
        <td><p>Directory tests</p></td>
      </tr><tr><td>
          <p><code><a href="is_area.html">is_area()</a></code> <code><a href="is_area.html">is_area_pixel()</a></code> <code><a href="is_area.html">is_area_si()</a></code> </p>
        </td>
        <td><p>Test area input</p></td>
      </tr><tr><td>
          <p><code><a href="is_dist.html">is_dist()</a></code> <code><a href="is_dist.html">is_dist_si()</a></code> <code><a href="is_dist.html">is_dist_pixel()</a></code> </p>
        </td>
        <td><p>Test distance input</p></td>
      </tr><tr><td>
          <p><code><a href="is_outlier.html">is_outlier()</a></code> </p>
        </td>
        <td><p>Check for outliers in a numeric vector</p></td>
      </tr><tr><td>
          <p><code><a href="is_unit_area.html">is_unit_area()</a></code> </p>
        </td>
        <td><p>Test unit of area input</p></td>
      </tr><tr><td>
          <p><code><a href="is_unit_dist.html">is_unit_dist()</a></code> </p>
        </td>
        <td><p>Test unit of length input</p></td>
      </tr><tr><td>
          <p><code><a href="joinWithPca.html">joinWithPca()</a></code> <code><a href="joinWithPca.html">joinWithTsne()</a></code> <code><a href="joinWithPca.html">joinWithUmap()</a></code> </p>
        </td>
        <td><p>Join barcodes with additional variables</p></td>
      </tr><tr><td>
          <p><code><a href="joinWithVariables.html">joinWithVariables()</a></code> </p>
        </td>
        <td><p>Join data with variables</p></td>
      </tr><tr><td>
          <p><code><a href="legendBottom.html">labsNone()</a></code> <code><a href="legendBottom.html">legendBottom()</a></code> <code><a href="legendBottom.html">legendColor()</a></code> <code><a href="legendBottom.html">legendLeft()</a></code> <code><a href="legendBottom.html">legendNone()</a></code> <code><a href="legendBottom.html">legendRight()</a></code> <code><a href="legendBottom.html">legendTop()</a></code> </p>
        </td>
        <td><p>ggplot2 basic manipulation</p></td>
      </tr><tr><td>
          <p><code><a href="loadImage.html">loadImage()</a></code> <code><a href="loadImage.html">loadImages()</a></code> </p>
        </td>
        <td><p>Load image slot content</p></td>
      </tr><tr><td>
          <p><code><a href="loadSpataObject.html">loadSpataObject()</a></code> </p>
        </td>
        <td><p>Load SPATA2 object</p></td>
      </tr><tr><td>
          <p><code><a href="mapSpatialAnnotationTags.html">mapSpatialAnnotationTags()</a></code> </p>
        </td>
        <td><p>Spatial annotation and barcode intersection</p></td>
      </tr><tr><td>
          <p><code><a href="map_to_tissue_section.html">map_to_tissue_section()</a></code> </p>
        </td>
        <td><p>Map observations to tissue sections</p></td>
      </tr><tr><td>
          <p><code><a href="matrix_dummy.html">matrix_dummy()</a></code> </p>
        </td>
        <td><p>matrix_dummy</p></td>
      </tr><tr><td>
          <p><code><a href="mergeGroups.html">mergeGroups()</a></code> </p>
        </td>
        <td><p>Lump groups together</p></td>
      </tr><tr><td>
          <p><code><a href="mergeSpatialAnnotations.html">mergeSpatialAnnotations()</a></code> </p>
        </td>
        <td><p>Merge spatial annotations</p></td>
      </tr><tr><td>
          <p><code><a href="mergeTissueSections.html">mergeTissueSections()</a></code> </p>
        </td>
        <td><p>Merge tissue sections</p></td>
      </tr><tr><td>
          <p><code><a href="mergeWithTissueOutline.html">mergeWithTissueOutline()</a></code> </p>
        </td>
        <td><p>Integrate tissue outline in spatial annotation</p></td>
      </tr><tr><td>
          <p><code><a href="model_ascending.html">model_ascending()</a></code> <code><a href="model_ascending.html">model_descending()</a></code> </p>
        </td>
        <td><p>Generate model-based ascending or descending sequence</p></td>
      </tr><tr><td>
          <p><code><a href="model_peak.html">model_peak()</a></code> <code><a href="model_peak.html">model_trough()</a></code> </p>
        </td>
        <td><p>Model a peaking pattern</p></td>
      </tr><tr><td>
          <p><code><a href="nMolecules.html">nMolecules()</a></code> </p>
        </td>
        <td><p>Number of molecules</p></td>
      </tr><tr><td>
          <p><code><a href="nObs.html">nObs()</a></code> </p>
        </td>
        <td><p>Number of observations</p></td>
      </tr><tr><td>
          <p><code><a href="nSpatialAnnotations.html">nSpatialAnnotations()</a></code> </p>
        </td>
        <td><p>Number of spatial annotations</p></td>
      </tr><tr><td>
          <p><code><a href="nSpatialTrajectories.html">nSpatialTrajectories()</a></code> <code><a href="nSpatialTrajectories.html">nTrajectories()</a></code> </p>
        </td>
        <td><p>Number of spatial trajectories</p></td>
      </tr><tr><td>
          <p><code><a href="normalizeCounts.html">normalizeCounts()</a></code> </p>
        </td>
        <td><p>Normalize raw counts</p></td>
      </tr><tr><td>
          <p><code><a href="plotAutoencoderAssessment.html">plotAutoencoderAssessment()</a></code> </p>
        </td>
        <td><p>Plot total variance of different neural networks</p></td>
      </tr><tr><td>
          <p><code><a href="plotAutoencoderResults.html">plotAutoencoderResults()</a></code> </p>
        </td>
        <td><p>Plot scaled vs. denoised expression</p></td>
      </tr><tr><td>
          <p><code><a href="plotBarchart.html">plotBarchart()</a></code> </p>
        </td>
        <td><p>Plot distribution of discrete variables</p></td>
      </tr><tr><td>
          <p><code><a href="plotBoxplot.html">plotBoxplot()</a></code> <code><a href="plotBoxplot.html">plotDensityplot()</a></code> <code><a href="plotBoxplot.html">plotHistogram()</a></code> <code><a href="plotBoxplot.html">plotRidgeplot()</a></code> <code><a href="plotBoxplot.html">plotVioBoxplot()</a></code> <code><a href="plotBoxplot.html">plotViolinplot()</a></code> </p>
        </td>
        <td><p>Plot numeric distribution and statistical tests</p></td>
      </tr><tr><td>
          <p><code><a href="plotCnvHeatmap.html">plotCnvHeatmap()</a></code> </p>
        </td>
        <td><p>Plot CNV Heatmap</p></td>
      </tr><tr><td>
          <p><code><a href="plotCnvLineplot.html">plotCnvLineplot()</a></code> </p>
        </td>
        <td><p>Plot CNV Lineplot</p></td>
      </tr><tr><td>
          <p><code><a href="plotDeaDotPlot.html">plotDeaDotPlot()</a></code> </p>
        </td>
        <td><p>Plot DEA results via dot plots</p></td>
      </tr><tr><td>
          <p><code><a href="plotDeaHeatmap.html">plotDeaHeatmap()</a></code> </p>
        </td>
        <td><p>Plot DEA results via heatmaps</p></td>
      </tr><tr><td>
          <p><code><a href="plotDeaVolcano.html">plotDeaVolcano()</a></code> <code><a href="plotDeaVolcano.html">plotDeaVolcano1v1()</a></code> </p>
        </td>
        <td><p>Plot DEA results via volcano plots</p></td>
      </tr><tr><td>
          <p><code><a href="plotExprVsDistSA.html">plotExprVsDistSA()</a></code> <code><a href="plotExprVsDistSA.html">plotExprVsDistST()</a></code> </p>
        </td>
        <td><p>Plot Expression as a function of distance to a spatial references</p></td>
      </tr><tr><td>
          <p><code><a href="plotGseaDotPlot.html">plotGseaDotPlot()</a></code> </p>
        </td>
        <td><p>Plot gene set enrichment</p></td>
      </tr><tr><td>
          <p><code><a href="plotImage.html">plotImage()</a></code> </p>
        </td>
        <td><p>Plot image with ggplot2</p></td>
      </tr><tr><td>
          <p><code><a href="plotImageBase.html">plotImageBase()</a></code> </p>
        </td>
        <td><p>Plot image with R base plotting</p></td>
      </tr><tr><td>
          <p><code><a href="plotImageMask.html">plotImageMask()</a></code> <code><a href="plotImageMask.html">plotPixelContent()</a></code> </p>
        </td>
        <td><p>Plot pixel content</p></td>
      </tr><tr><td>
          <p><code><a href="plotImages.html">plotImages()</a></code> </p>
        </td>
        <td><p>Plot histology images (ggplot2)</p></td>
      </tr><tr><td>
          <p><code><a href="plotLoglik.html">plotLoglik()</a></code> </p>
        </td>
        <td><p>Plot Bayes Space logliks</p></td>
      </tr><tr><td>
          <p><code><a href="plotMosaicplot.html">plotMosaicplot()</a></code> </p>
        </td>
        <td><p>Plot mosaic plot</p></td>
      </tr><tr><td>
          <p><code><a href="plotOverview.html">plotOverview()</a></code> </p>
        </td>
        <td><p>Plot overview of S4 objects</p></td>
      </tr><tr><td>
          <p><code><a href="plotPcaVariation.html">plotPcaVariation()</a></code> </p>
        </td>
        <td><p>Plot PCA Variation</p></td>
      </tr><tr><td>
          <p><code><a href="plotRiverplot.html">plotRiverplot()</a></code> </p>
        </td>
        <td><p>Plot riverplots</p></td>
      </tr><tr><td>
          <p><code><a href="plotSasBarplot.html">plotSasBarplot()</a></code> </p>
        </td>
        <td><p>Plot SAS barplot</p></td>
      </tr><tr><td>
          <p><code><a href="plotSasHeatmap.html">plotSasHeatmap()</a></code> </p>
        </td>
        <td><p>Plot SAS heatmap</p></td>
      </tr><tr><td>
          <p><code><a href="plotSasLineplot.html">plotSasLineplot()</a></code> </p>
        </td>
        <td><p>Plot SAS lineplot</p></td>
      </tr><tr><td>
          <p><code><a href="plotSasRidgeplot.html">plotSasRidgeplot()</a></code> </p>
        </td>
        <td><p>Plot SAS rideplot</p></td>
      </tr><tr><td>
          <p><code><a href="plotScatterplot.html">plotScatterplot()</a></code> </p>
        </td>
        <td><p>Plot numeric variables as a scatterplot</p></td>
      </tr><tr><td>
          <p><code><a href="plotSpatialAnnotations.html">plotSpatialAnnotations()</a></code> </p>
        </td>
        <td><p>Plot spatial annotations</p></td>
      </tr><tr><td>
          <p><code><a href="plotSpatialTrajectories.html">plotSpatialTrajectories()</a></code> </p>
        </td>
        <td><p>Plot spatial trajectories</p></td>
      </tr><tr><td>
          <p><code><a href="plotStatisticsInteractive.html">plotStatisticsInteractive()</a></code> </p>
        </td>
        <td><p>Plot distribution of variables interactively</p></td>
      </tr><tr><td>
          <p><code><a href="plotStsBarplot.html">plotStsBarplot()</a></code> </p>
        </td>
        <td><p>Plot STS barplot</p></td>
      </tr><tr><td>
          <p><code><a href="plotStsHeatmap.html">plotStsHeatmap()</a></code> </p>
        </td>
        <td><p>Plot STS heatmap</p></td>
      </tr><tr><td>
          <p><code><a href="plotStsLineplot.html">plotStsLineplot()</a></code> <code><a href="plotStsLineplot.html">plotStsRidgeplot()</a></code> </p>
        </td>
        <td><p>Plot STS line- and ridgeplot</p></td>
      </tr><tr><td>
          <p><code><a href="plotSurface.html">plotSurface()</a></code> <code><a href="plotSurface.html">plotSurfaceInteractive()</a></code> </p>
        </td>
        <td><p>Plot the surface of the sample</p></td>
      </tr><tr><td>
          <p><code><a href="plotSurfaceBase.html">plotSurfaceBase()</a></code> </p>
        </td>
        <td><p>Plot surface with R base plotting</p></td>
      </tr><tr><td>
          <p><code><a href="plotSurfaceComparison.html">plotSurfaceComparison()</a></code> </p>
        </td>
        <td><p>Plot several surface plots colored by numeric variables</p></td>
      </tr><tr><td>
          <p><code><a href="plotSurfaceSAS.html">plotSurfaceSAS()</a></code> </p>
        </td>
        <td><p>Plot screening area of SAS set up</p></td>
      </tr><tr><td>
          <p><code><a href="plotSurfaceSC.html">plotSurfaceSC()</a></code> </p>
        </td>
        <td><p>Plot single cells on surface</p></td>
      </tr><tr><td>
          <p><code><a href="plotUMAP.html">plotPCA()</a></code> <code><a href="plotUMAP.html">plotPca()</a></code> <code><a href="plotUMAP.html">plotTSNE()</a></code> <code><a href="plotUMAP.html">plotTsne()</a></code> <code><a href="plotUMAP.html">plotTsneComparison()</a></code> <code><a href="plotUMAP.html">plotUMAP()</a></code> <code><a href="plotUMAP.html">plotUmap()</a></code> <code><a href="plotUMAP.html">plotUmapComparison()</a></code> </p>
        </td>
        <td><p>Plot dimensional reduction</p></td>
      </tr><tr><td>
          <p><code><a href="printAutoencoderSummary.html">printAutoencoderSummary()</a></code> </p>
        </td>
        <td><p>Print autoencoder summary</p></td>
      </tr><tr><td>
          <p><code><a href="printDefaultInstructions.html">printDefaultInstructions()</a></code> </p>
        </td>
        <td><p>Print current default settings</p></td>
      </tr><tr><td>
          <p><code><a href="printGeneSetOverview.html">printGeneSetOverview()</a></code> </p>
        </td>
        <td><p>Print overview about the current gene sets</p></td>
      </tr><tr><td>
          <p><code><a href="read_coords.html">read_coords()</a></code> <code><a href="read_coords.html">read_coords_merfish()</a></code> <code><a href="read_coords.html">read_coords_slide_seq_v1()</a></code> <code><a href="read_coords.html">read_coords_visium()</a></code> <code><a href="read_coords.html">read_coords_xenium()</a></code> </p>
        </td>
        <td><p>Read coordinate data.frames</p></td>
      </tr><tr><td>
          <p><code><a href="recSgsRes.html">recDbscanEps()</a></code> <code><a href="recSgsRes.html">recDbscanMinPts()</a></code> <code><a href="recSgsRes.html">recSgsRes()</a></code> </p>
        </td>
        <td><p>Platform dependent input recommendations</p></td>
      </tr><tr><td>
          <p><code><a href="reduceSpataObject.html">reduceSpataObject()</a></code> </p>
        </td>
        <td><p>Reduce SPATA2 object to minimal version</p></td>
      </tr><tr><td>
          <p><code><a href="refImage.html">refImage()</a></code> </p>
        </td>
        <td><p>Obtain name of reference iamge</p></td>
      </tr><tr><td>
          <p><code><a href="registerImage.html">registerImage()</a></code> <code><a href="registerImage.html">removeImage()</a></code> </p>
        </td>
        <td><p>Register or remove images</p></td>
      </tr><tr><td>
          <p><code><a href="relevelGroups.html">relevelGroups()</a></code> </p>
        </td>
        <td><p>Relevel groups of grouping variable</p></td>
      </tr><tr><td>
          <p><code><a href="removeMetaFeatures.html">removeMetaFeatures()</a></code> </p>
        </td>
        <td><p>Remove meta features</p></td>
      </tr><tr><td>
          <p><code><a href="removeMolecules.html">removeGenes()</a></code> <code><a href="removeMolecules.html">removeGenesStress()</a></code> <code><a href="removeMolecules.html">removeGenesZeroCounts()</a></code> <code><a href="removeMolecules.html">removeMolecules()</a></code> </p>
        </td>
        <td><p>Remove molecules from the <code>SPATA2</code> object</p></td>
      </tr><tr><td>
          <p><code><a href="removeObs.html">removeObs()</a></code> <code><a href="removeObs.html">removeObsNoCounts()</a></code> </p>
        </td>
        <td><p>Remove observations</p></td>
      </tr><tr><td>
          <p><code><a href="removeProcessedMatrix.html">removeProcessedMatrix()</a></code> </p>
        </td>
        <td><p>Remove a processed matrix</p></td>
      </tr><tr><td>
          <p><code><a href="removeSpatialAnnotations.html">removeSpatialAnnotations()</a></code> </p>
        </td>
        <td><p>Remove spatial annotations</p></td>
      </tr><tr><td>
          <p><code><a href="removeSpatialOutliers.html">removeSpatialOutliers()</a></code> </p>
        </td>
        <td><p>Remove spatial outliers</p></td>
      </tr><tr><td>
          <p><code><a href="removeSpatialTrajectories.html">removeSpatialTrajectories()</a></code> </p>
        </td>
        <td><p>Remove spatial trajectories</p></td>
      </tr><tr><td>
          <p><code><a href="removeTissueFragments.html">removeTissueFragments()</a></code> </p>
        </td>
        <td><p>Remove data points from tissue fragments</p></td>
      </tr><tr><td>
          <p><code><a href="renameGroups.html">renameGroups()</a></code> <code><a href="renameGroups.html">renameTissueSection()</a></code> </p>
        </td>
        <td><p>Rename cluster/group names</p></td>
      </tr><tr><td>
          <p><code><a href="renameMetaFeatures.html">renameMetaFeatures()</a></code> </p>
        </td>
        <td><p>Rename features</p></td>
      </tr><tr><td>
          <p><code><a href="renameSpataObject.html">renameSpataObject()</a></code> </p>
        </td>
        <td><p>Rename the object</p></td>
      </tr><tr><td>
          <p><code><a href="renameSpatialAnnotation.html">renameSpatialAnnotation()</a></code> </p>
        </td>
        <td><p>Rename a spatial annotation</p></td>
      </tr><tr><td>
          <p><code><a href="renameSpatialTrajectory.html">renameSpatialTrajectory()</a></code> </p>
        </td>
        <td><p>Rename a spatial trajectory.</p></td>
      </tr><tr><td>
          <p><code><a href="resetImageTransformations.html">resetImageTransformations()</a></code> </p>
        </td>
        <td><p>Reset image transformations</p></td>
      </tr><tr><td>
          <p><code><a href="rotateAll.html">rotateAll()</a></code> <code><a href="rotateAll.html">rotateImage()</a></code> <code><a href="rotateAll.html">rotateCoordinates()</a></code> <code><a href="rotateAll.html">rotateCoordsDf()</a></code> <code><a href="rotateAll.html">rotateSpatialAnnotations()</a></code> <code><a href="rotateAll.html">rotateSpatialTrajectories()</a></code> </p>
        </td>
        <td><p>Rotate image and coordinates</p></td>
      </tr><tr><td>
          <p><code><a href="rotateSpatialAnnotation.html">rotateSpatialAnnotation()</a></code> </p>
        </td>
        <td><p>Rotate the outline of a spatial annotation</p></td>
      </tr><tr><td>
          <p><code><a href="runAutoencoderDenoising.html">runAutoencoderDenoising()</a></code> </p>
        </td>
        <td><p>Denoise expression matrix</p></td>
      </tr><tr><td>
          <p><code><a href="runBayesSpaceClustering.html">runBayesSpaceClustering()</a></code> </p>
        </td>
        <td><p>Clustering with BayesSpace</p></td>
      </tr><tr><td>
          <p><code><a href="runCNV.html">runCNV()</a></code> <code><a href="runCNV.html">runCnvAnalysis()</a></code> </p>
        </td>
        <td><p>Identify large-scale chromosomal copy number variations</p></td>
      </tr><tr><td>
          <p><code><a href="runDEA.html">runDEA()</a></code> <code><a href="runDEA.html">runDeAnalysis()</a></code> </p>
        </td>
        <td><p>Find differently expressed genes</p></td>
      </tr><tr><td>
          <p><code><a href="runGSEA.html">runGSEA()</a></code> </p>
        </td>
        <td><p>Compute gene set enrichment</p></td>
      </tr><tr><td>
          <p><code><a href="runKmeansClustering.html">runKmeansClustering()</a></code> </p>
        </td>
        <td><p>Clustering with Kmeans</p></td>
      </tr><tr><td>
          <p><code><a href="runPCA.html">runPCA()</a></code> <code><a href="runPCA.html">runPca()</a></code> </p>
        </td>
        <td><p>Run Principal Component Analysis</p></td>
      </tr><tr><td>
          <p><code><a href="runSPARKX.html">getSparkxGeneDf()</a></code> <code><a href="runSPARKX.html">getSparkxGenes()</a></code> <code><a href="runSPARKX.html">getSparkxResults()</a></code> <code><a href="runSPARKX.html">runSPARKX()</a></code> </p>
        </td>
        <td><p>Identify spatially significant features with SPARKX</p></td>
      </tr><tr><td>
          <p><code><a href="runSeuratClustering.html">runSeuratClustering()</a></code> </p>
        </td>
        <td><p>Clustering with Seurat</p></td>
      </tr><tr><td>
          <p><code><a href="runTSNE.html">runTSNE()</a></code> <code><a href="runTSNE.html">runTsne()</a></code> </p>
        </td>
        <td><p>Run t-Stochastic Neighbour Embedding</p></td>
      </tr><tr><td>
          <p><code><a href="runUMAP.html">runUMAP()</a></code> <code><a href="runUMAP.html">runUmap()</a></code> </p>
        </td>
        <td><p>Run UMAP</p></td>
      </tr><tr><td>
          <p><code><a href="saveSpataObject.html">saveSpataObject()</a></code> </p>
        </td>
        <td><p>Save a SPATA2 object with a default directory</p></td>
      </tr><tr><td>
          <p><code><a href="setAssay.html">setAssay()</a></code> </p>
        </td>
        <td><p>Set molecular assay</p></td>
      </tr><tr><td>
          <p><code><a href="setAutoencoderAssessment.html">setAutoencoderAssessment()</a></code> </p>
        </td>
        <td><p>Set results of autoencoder assessment</p></td>
      </tr><tr><td>
          <p><code><a href="setBarcodes.html">setBarcodes()</a></code> </p>
        </td>
        <td><p>Set barcodes</p></td>
      </tr><tr><td>
          <p><code><a href="setCCD.html">setCCD()</a></code> </p>
        </td>
        <td><p>Set center to center distance</p></td>
      </tr><tr><td>
          <p><code><a href="setCaptureArea.html">setCaptureArea()</a></code> </p>
        </td>
        <td><p>Set capture area</p></td>
      </tr><tr><td>
          <p><code><a href="setCnvResults.html">setCnvResults()</a></code> </p>
        </td>
        <td><p>Set cnv-results</p></td>
      </tr><tr><td>
          <p><code><a href="setCoordsDf.html">setCoordsDf()</a></code> </p>
        </td>
        <td><p>Set the coordinates</p></td>
      </tr><tr><td>
          <p><code><a href="setCountMatrix.html">setCountMatrix()</a></code> <code><a href="setCountMatrix.html">setProcessedMatrix()</a></code> </p>
        </td>
        <td><p>Set data matrices</p></td>
      </tr><tr><td>
          <p><code><a href="setDefault.html">setDefault()</a></code> </p>
        </td>
        <td><p>Set object specific default</p></td>
      </tr><tr><td>
          <p><code><a href="setDefaultInstructions.html">setDefaultInstructions()</a></code> <code><a href="setDefaultInstructions.html">setDirectoryInstructions()</a></code> </p>
        </td>
        <td><p>Set default instructions</p></td>
      </tr><tr><td>
          <p><code><a href="setHistoImage.html">setHistoImage()</a></code> </p>
        </td>
        <td><p>Set histo image</p></td>
      </tr><tr><td>
          <p><code><a href="setImageDir.html">setImageDir()</a></code> </p>
        </td>
        <td><p>Set image directory</p></td>
      </tr><tr><td>
          <p><code><a href="setImageTransformations.html">setImageTransformations()</a></code> </p>
        </td>
        <td><p>Set image transformation instructions</p></td>
      </tr><tr><td>
          <p><code><a href="setLogfileDf.html">setLogfileDf()</a></code> </p>
        </td>
        <td><p>Set logfile data.frame</p></td>
      </tr><tr><td>
          <p><code><a href="setMetaDf.html">setMetaDf()</a></code> </p>
        </td>
        <td><p>Set meta data.frame</p></td>
      </tr><tr><td>
          <p><code><a href="setPcaDf.html">setPcaDf()</a></code> <code><a href="setPcaDf.html">setTsneDf()</a></code> <code><a href="setPcaDf.html">setUmapDf()</a></code> </p>
        </td>
        <td><p>Set dimensional reductions data.frames</p></td>
      </tr><tr><td>
          <p><code><a href="setScaleFactor.html">setScaleFactor()</a></code> </p>
        </td>
        <td><p>Set scale factors</p></td>
      </tr><tr><td>
          <p><code><a href="setSpataDir.html">setSpataDir()</a></code> </p>
        </td>
        <td><p>Set SPATA2 directory</p></td>
      </tr><tr><td>
          <p><code><a href="setSpatialAnnotation.html">setSpatialAnnotation()</a></code> <code><a href="setSpatialAnnotation.html">setSpatialAnnotations()</a></code> </p>
        </td>
        <td><p>Set spatial annotations</p></td>
      </tr><tr><td>
          <p><code><a href="setSpatialData.html">setSpatialData()</a></code> </p>
        </td>
        <td><p>Set spatial data</p></td>
      </tr><tr><td>
          <p><code><a href="setSpatialMethod.html">setSpatialMethod()</a></code> </p>
        </td>
        <td><p>Set spatial method</p></td>
      </tr><tr><td>
          <p><code><a href="setSpatialMethodInfo.html">setSpatialMethodInfo()</a></code> </p>
        </td>
        <td><p>Set information of <code>SpatialMethod</code> object</p></td>
      </tr><tr><td>
          <p><code><a href="setSpatialMethodSlot.html">setSpatialMethodSlot()</a></code> </p>
        </td>
        <td><p>Set slot content of <code>SpatialMethod</code> object</p></td>
      </tr><tr><td>
          <p><code><a href="shiftSpatialAnnotation.html">shiftSpatialAnnotation()</a></code> </p>
        </td>
        <td><p>Shift the borders of a spatial annotation</p></td>
      </tr><tr><td>
          <p><code><a href="shinyModuleZoomingServer.html">shinyModuleZoomingServer()</a></code> <code><a href="shinyModuleZoomingServer.html">shinyModuleZoomingUI()</a></code> </p>
        </td>
        <td><p>Shiny module <em>Zooming</em></p></td>
      </tr><tr><td>
          <p><code><a href="showColors.html">showColors()</a></code> <code><a href="showColors.html">showColorPalettes()</a></code> <code><a href="showColors.html">showColorSpectra()</a></code> </p>
        </td>
        <td><p>Show color palettes and spectra</p></td>
      </tr><tr><td>
          <p><code><a href="showModels.html">showModels()</a></code> </p>
        </td>
        <td><p>Show spatial gradient screening models</p></td>
      </tr><tr><td>
          <p><code><a href="signatures.html">gsdf</a></code> </p>
        </td>
        <td><p>The default collection of frequently used signatures</p></td>
      </tr><tr><td>
          <p><code><a href="smoothSpatialAnnotation.html">smoothSpatialAnnotation()</a></code> </p>
        </td>
        <td><p>Smooth the border of a spatial annotation</p></td>
      </tr><tr><td>
          <p><code><a href="spatialAnnotationScreening.html">spatialAnnotationScreening()</a></code> </p>
        </td>
        <td><p>The Spatial Annotation Screening algorithm</p></td>
      </tr><tr><td>
          <p><code><a href="spatialAnnotationToGrouping.html">spatialAnnotationToGrouping()</a></code> </p>
        </td>
        <td><p>Convert spatial annotation to grouping variable</p></td>
      </tr><tr><td>
          <p><code><a href="spatialTrajectoryScreening.html">spatialTrajectoryScreening()</a></code> </p>
        </td>
        <td><p>The Spatial Trajectory Screening algorithm</p></td>
      </tr><tr><td>
          <p><code><a href="spatial_gradient_screening.html">spatial_gradient_screening()</a></code> </p>
        </td>
        <td><p>Low level implementation of the spatial gradient screening algorithm</p></td>
      </tr><tr><td>
          <p><code><a href="splitSpataObject.html">splitSpataObject()</a></code> </p>
        </td>
        <td><p>Split SPATA2 object</p></td>
      </tr><tr><td>
          <p><code><a href="subsetByBarcodes.html">subsetByBarcodes()</a></code> </p>
        </td>
        <td><p>Subset SPATA2 object</p></td>
      </tr><tr><td>
          <p><code><a href="summarize_formulas.html">summarize_formulas</a></code> </p>
        </td>
        <td><p>List of summarizing formulas</p></td>
      </tr><tr><td>
          <p><code><a href="tagSpatialAnnotations.html">tagSpatialAnnotations()</a></code> </p>
        </td>
        <td><p>Tag spatial annotations</p></td>
      </tr><tr><td>
          <p><code><a href="theme_lineplot_gradient.html">theme_lineplot_gradient()</a></code> <code><a href="theme_lineplot_gradient.html">theme_ridgeplot_gradient()</a></code> <code><a href="theme_lineplot_gradient.html">theme_trajectory_fit()</a></code> <code><a href="theme_lineplot_gradient.html">theme_transparent()</a></code> </p>
        </td>
        <td><p>ggplot2 themes</p></td>
      </tr><tr><td>
          <p><code><a href="transform_coords.html">transform_coords()</a></code> </p>
        </td>
        <td><p>Transform coordinates</p></td>
      </tr><tr><td>
          <p><code><a href="transform_image.html">transform_image()</a></code> </p>
        </td>
        <td><p>Transform image</p></td>
      </tr><tr><td>
          <p><code><a href="unloadImage.html">unloadImage()</a></code> <code><a href="unloadImage.html">unloadImages()</a></code> </p>
        </td>
        <td><p>Unload image slot content</p></td>
      </tr><tr><td>
          <p><code><a href="updateS4.html">updateS4()</a></code> </p>
        </td>
        <td><p>Update S4 objects</p></td>
      </tr><tr><td>
          <p><code><a href="updateSpataObject.html">updateSpataObject()</a></code> </p>
        </td>
        <td><p>Update <code>SPATA2</code> object</p></td>
      </tr><tr><td>
          <p><code><a href="useVarForTissueOutline.html">useVarForTissueOutline()</a></code> </p>
        </td>
        <td><p>Use specified variable for tissue outline</p></td>
      </tr><tr><td>
          <p><code><a href="validActivationFunctions.html">validActivationFunctions()</a></code> <code><a href="validActivationFunctions.html">validAgglomerationMethods()</a></code> <code><a href="validActivationFunctions.html">validAlluvialTypes()</a></code> <code><a href="validActivationFunctions.html">validColorPalettes()</a></code> <code><a href="validActivationFunctions.html">validColorSpectra()</a></code> <code><a href="validActivationFunctions.html">validDeaMethods()</a></code> <code><a href="validActivationFunctions.html">validDefaultInstructionSlots()</a></code> <code><a href="validActivationFunctions.html">validDistanceMethods()</a></code> <code><a href="validActivationFunctions.html">validHierarchicalClusterMethods()</a></code> <code><a href="validActivationFunctions.html">validModelNames()</a></code> <code><a href="validActivationFunctions.html">validPadjMethods()</a></code> <code><a href="validActivationFunctions.html">validSpatialMethods()</a></code> <code><a href="validActivationFunctions.html">validUnits()</a></code> <code><a href="validActivationFunctions.html">validUnitsOfArea()</a></code> <code><a href="validActivationFunctions.html">validUnitsOfAreaSI()</a></code> <code><a href="validActivationFunctions.html">validUnitsOfLength()</a></code> <code><a href="validActivationFunctions.html">validUnitsOfLengthSI()</a></code> </p>
        </td>
        <td><p>Obtain valid argument inputs</p></td>
      </tr><tr><td>
          <p><code><a href="visiumSpotDistances.html">visiumSpotDistances()</a></code> </p>
        </td>
        <td><p>Calculate Distances Between Visium Spots</p></td>
      </tr><tr><td>
          <p><code><a href="visium_spots.html">visium_spots</a></code> </p>
        </td>
        <td><p>A list of a slot named <em>VisiumSmall</em>, which contains the data.frame of all barcoded spots of 6.5mm capture area and a slot named <em>VisiumLarge</em> which contains the barcoded spots of the 11mm capture area.</p></td>
      </tr><tr><td>
          <p><code><a href="whichTissueSection.html">whichTissueSection()</a></code> </p>
        </td>
        <td><p>Tissue section belonging of spatial annotations</p></td>
      </tr></tbody></table></div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Jan Kueckelhaus, Dieter-Henrik Heiland, Simon Frerich.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

      </footer></div>

  


  

  </body></html>

