<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Plot CNV Lineplot — plotCnvLineplot • SPATA2</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Plot CNV Lineplot — plotCnvLineplot"><meta name="description" content="Plots the results of runCNV() in form of a lineplot.
Use arguments across and across_subset to visualize CNV differences
between subgroups of cluster variables or other grouping variables (e.g. based on
histology created with createSpatialSegmentation())."><meta property="og:description" content="Plots the results of runCNV() in form of a lineplot.
Use arguments across and across_subset to visualize CNV differences
between subgroups of cluster variables or other grouping variables (e.g. based on
histology created with createSpatialSegmentation())."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">SPATA2</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">3.0.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item"><a class="nav-link" href="../articles/installation.html">Installation</a></li>
<li class="nav-item"><a class="nav-link" href="../articles/spata-data.html">SPATAData</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><h6 class="dropdown-header" data-toc-skip>Initiating a SPATA2 object</h6></li>
    <li><a class="dropdown-item" href="../articles/initiation-and-preprocessing-customized.html">Initiation &amp; Preprocessing with individual data</a></li>
    <li><a class="dropdown-item" href="../articles/initiation-and-preprocessing-merfish.html">Initiation &amp; Preprocessing with MERFISH data</a></li>
    <li><a class="dropdown-item" href="../articles/initiation-and-preprocessing-visium.html">Initiation &amp; Preprocessing with Visium data</a></li>
    <li><a class="dropdown-item" href="../articles/initiation-and-preprocessing-visium-hd.html">Initiation &amp; Preprocessing with VisiumHD data</a></li>
    <li><h6 class="dropdown-header" data-toc-skip>Basic Manipulation of a SPATA2 object</h6></li>
    <li><a class="dropdown-item" href="../articles/adding-data.html">Adding Data to SPATA2 objects</a></li>
    <li><a class="dropdown-item" href="../articles/package-compatibility.html">Platform Compatibility</a></li>
    <li><a class="dropdown-item" href="../articles/molecular-variables.html">Molecular Variables</a></li>
    <li><a class="dropdown-item" href="../articles/image-handling.html">Image Handling</a></li>
    <li><a class="dropdown-item" href="../articles/spatial-segmentation.html">Spatial Segmentation</a></li>
    <li><h6 class="dropdown-header" data-toc-skip>Implemented Algorithms &amp; Concepts</h6></li>
    <li><a class="dropdown-item" href="../articles/cnv.html">Copy Number Variations</a></li>
    <li><a class="dropdown-item" href="../articles/clustering.html">Clustering</a></li>
    <li><a class="dropdown-item" href="../articles/dimensional-reduction.html">Dimensional Reduction</a></li>
    <li><a class="dropdown-item" href="../articles/dea.html">Differential Expression</a></li>
    <li><a class="dropdown-item" href="../articles/gsea.html">Gene Set Enrichment</a></li>
    <li><a class="dropdown-item" href="../articles/spatial-measures.html">Spatial Measures</a></li>
    <li><h6 class="dropdown-header" data-toc-skip>Spatial Gradient Screening</h6></li>
    <li><a class="dropdown-item" href="../articles/creating-spatial-annotations.html">Creating Spatial Annotations</a></li>
    <li><a class="dropdown-item" href="../articles/using-spatial-annotations.html">Using Spatial Annotations</a></li>
    <li><a class="dropdown-item" href="../articles/spatial-annotation-screening.html">Spatial Annotation Screening</a></li>
    <li><a class="dropdown-item" href="../articles/creating-spatial-trajectories.html">Creating Spatial Trajectories</a></li>
    <li><a class="dropdown-item" href="../articles/spatial-trajectory-screening.html">Spatial Trajectory Screening</a></li>
    <li><h6 class="dropdown-header" data-toc-skip>Visualization</h6></li>
    <li><a class="dropdown-item" href="../articles/plotting-surface.html">Surface Plotting</a></li>
  </ul></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Plot CNV Lineplot</h1>

      <div class="d-none name"><code>plotCnvLineplot.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Plots the results of <code><a href="runCNV.html">runCNV()</a></code> in form of a lineplot.
Use arguments <code>across</code> and <code>across_subset</code> to visualize CNV differences
between subgroups of cluster variables or other grouping variables (e.g. based on
histology created with <code><a href="createSpatialSegmentation.html">createSpatialSegmentation()</a></code>).</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">plotCnvLineplot</span><span class="op">(</span></span>
<span>  <span class="va">object</span>,</span>
<span>  across <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  across_subset <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  arm_subset <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"p"</span>, <span class="st">"q"</span><span class="op">)</span>,</span>
<span>  chrom_subset <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">22</span>,</span>
<span>  chrom_separate <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">22</span>,</span>
<span>  chrom_arm_subset <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  smooth_span <span class="op">=</span> <span class="fl">0.08</span>,</span>
<span>  line_alpha <span class="op">=</span> <span class="fl">0.9</span>,</span>
<span>  line_color <span class="op">=</span> <span class="st">"blue"</span>,</span>
<span>  line_size <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  display_ribbon <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  ribbon_alpha <span class="op">=</span> <span class="fl">0.25</span>,</span>
<span>  ribbon_fill <span class="op">=</span> <span class="st">"lightgrey"</span>,</span>
<span>  vline_alpha <span class="op">=</span> <span class="fl">0.75</span>,</span>
<span>  vline_color <span class="op">=</span> <span class="st">"black"</span>,</span>
<span>  vline_size <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>  vline_type <span class="op">=</span> <span class="st">"dashed"</span>,</span>
<span>  summarize_with <span class="op">=</span> <span class="st">"mean"</span>,</span>
<span>  nrow <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  ncol <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  breaks_y <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.9</span>, <span class="fl">0.95</span>, <span class="fl">1</span>, <span class="fl">1.05</span>, <span class="fl">1.1</span><span class="op">)</span>,</span>
<span>  labels_y <span class="op">=</span> <span class="va">breaks_y</span>,</span>
<span>  limits_y <span class="op">=</span> <span class="fu">base</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/base/range.html" class="external-link">range</a></span><span class="op">(</span><span class="va">breaks_y</span><span class="op">)</span>,</span>
<span>  expand_y <span class="op">=</span> <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/waiver.html" class="external-link">waiver</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-object">object<a class="anchor" aria-label="anchor" href="#arg-object"></a></dt>
<dd><p>An object of class <code><a href="SPATA2.html">SPATA2</a></code> or, in case of S4 generics,
objects of classes for which a method has been defined.</p></dd>


<dt id="arg-across">across<a class="anchor" aria-label="anchor" href="#arg-across"></a></dt>
<dd><p>Character value or NULL. Specifies the grouping variable of interest.</p>
<p>Use <code><a href="getGroupingOptions.html">getGroupingOptions()</a></code> to obtain all variable names that group the
barcode spots of your object in a certain manner.</p></dd>


<dt id="arg-across-subset">across_subset<a class="anchor" aria-label="anchor" href="#arg-across-subset"></a></dt>
<dd><p>Character vector or NULL. Specifies the particular groups
of interest the grouping variable specified in argument <code>across</code> contains.</p>
<p>If set to NULL all of them are chosen. You can prefix groups you are NOT interested in
with a <em>'-'</em>. (Saves writing if there are more groups you are interested in
than groups you are not interested in.)</p>
<p>Use <code><a href="getGroupNames.html">getGroupNames()</a></code> to obtain all valid input options.</p></dd>


<dt id="arg-arm-subset">arm_subset<a class="anchor" aria-label="anchor" href="#arg-arm-subset"></a></dt>
<dd><p>Character vector. A combination of <em>'p'</em> and/or <em>'q'</em>.
Denotes which chromosome arms are included. Defaults to both.</p></dd>


<dt id="arg-chrom-subset">chrom_subset<a class="anchor" aria-label="anchor" href="#arg-chrom-subset"></a></dt>
<dd><p>Character or numeric vector. Denotes the chromosomes that
are included. Defaults to all 1-22.</p></dd>


<dt id="arg-chrom-separate">chrom_separate<a class="anchor" aria-label="anchor" href="#arg-chrom-separate"></a></dt>
<dd><p>Character or numeric vector. Denotes the chromosomes that
are separated from their neighbors by vertical lines. Defaults to all 1-22. If FALSE or NULL,
no vertical lines are drawn. Requires <code>display_vlines</code> to be set to TRUE.</p></dd>


<dt id="arg-chrom-arm-subset">chrom_arm_subset<a class="anchor" aria-label="anchor" href="#arg-chrom-arm-subset"></a></dt>
<dd><p>Character vector. Denotes the exact chromosome-arm combinations
that are included.</p></dd>


<dt id="arg-smooth-span">smooth_span<a class="anchor" aria-label="anchor" href="#arg-smooth-span"></a></dt>
<dd><p>Numeric value. Controls the degree of smoothing.
Given to argument <code>span</code> of <code><a href="https://rdrr.io/r/stats/loess.html" class="external-link">stats::loess()</a></code>.</p></dd>


<dt id="arg-line-alpha">line_alpha<a class="anchor" aria-label="anchor" href="#arg-line-alpha"></a></dt>
<dd><p>Numeric. Affects alpha of main lines of the plot.</p></dd>


<dt id="arg-line-color">line_color<a class="anchor" aria-label="anchor" href="#arg-line-color"></a></dt>
<dd><p>Character. Affects color of the main lines of the plot.</p></dd>


<dt id="arg-line-size">line_size<a class="anchor" aria-label="anchor" href="#arg-line-size"></a></dt>
<dd><p>Numeric. Affects size of the main lines of the plot.</p></dd>


<dt id="arg-display-ribbon">display_ribbon<a class="anchor" aria-label="anchor" href="#arg-display-ribbon"></a></dt>
<dd><p>Logical value. If TRUE, a ribbon is displayed around
the main line of the plot visualizing uncertainty using standard deviation.</p></dd>


<dt id="arg-ribbon-alpha-ribbon-fill">ribbon_alpha, ribbon_fill<a class="anchor" aria-label="anchor" href="#arg-ribbon-alpha-ribbon-fill"></a></dt>
<dd><p>Parameters given to <code>ggplot2::geom_ribbion()</code>
that control the appearance of the ribbon around the main line of the plot.</p></dd>


<dt id="arg-vline-alpha-vline-color-vline-size-vline-type">vline_alpha, vline_color, vline_size, vline_type<a class="anchor" aria-label="anchor" href="#arg-vline-alpha-vline-color-vline-size-vline-type"></a></dt>
<dd><p>Parameters given to
<code><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html" class="external-link">ggplot2::geom_vline()</a></code> that control the appearance of vertical lines
of the plot.</p></dd>


<dt id="arg-summarize-with">summarize_with<a class="anchor" aria-label="anchor" href="#arg-summarize-with"></a></dt>
<dd><p>Character value. Name of the function with which to summarize. Either
<em>'mean'</em> or <em>'median'</em>.</p></dd>


<dt id="arg-nrow-ncol">nrow, ncol<a class="anchor" aria-label="anchor" href="#arg-nrow-ncol"></a></dt>
<dd><p>Numeric values or NULL. Used to arrange multiple plots.</p></dd>


<dt id="arg-breaks-y-labels-y-limits-y-expand-y">breaks_y, labels_y, limits_y, expand_y<a class="anchor" aria-label="anchor" href="#arg-breaks-y-labels-y-limits-y-expand-y"></a></dt>
<dd><p>Given to the corresponding arguments
of <code><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">ggplot2::scale_y_continuous()</a></code>.</p></dd>


<dt id="arg-verbose">verbose<a class="anchor" aria-label="anchor" href="#arg-verbose"></a></dt>
<dd><p>Logical. If <code>TRUE</code>, informative messages regarding
the computational progress will be printed.</p>
<p>(Warning messages will always be printed.)</p></dd>


<dt id="arg--">...<a class="anchor" aria-label="anchor" href="#arg--"></a></dt>
<dd><p>Used to absorb deprecated arguments or functions.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A ggplot.</p>
    </div>
    <div class="section level2">
    <h2 id="tutorials">Tutorials<a class="anchor" aria-label="anchor" href="#tutorials"></a></h2>


<p>Extensive tutorials for how to use this function can be found on our website
<a href="https://themilolab.github.io/SPATA2/" class="external-link">https://themilolab.github.io/SPATA2/</a> .</p>
    </div>

  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Jan Kueckelhaus, Dieter-Henrik Heiland.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.0.</p>
</div>

    </footer></div>





  </body></html>

