<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Plot DEA results via dot plots — plotDeaDotPlot • SPATA2</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Plot DEA results via dot plots — plotDeaDotPlot"><meta name="description" content="Visualizes results of DE analysis with
dot plots."><meta property="og:description" content="Visualizes results of DE analysis with
dot plots."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">SPATA2</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">3.0.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item"><a class="nav-link" href="../articles/installation.html">Installation</a></li>
<li class="nav-item"><a class="nav-link" href="../articles/spata-data.html">SPATAData</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><h6 class="dropdown-header" data-toc-skip>Initiating a SPATA2 object</h6></li>
    <li><a class="dropdown-item" href="../articles/initiation-and-preprocessing-customized.html">Initiation &amp; Preprocessing with individual data</a></li>
    <li><a class="dropdown-item" href="../articles/initiation-and-preprocessing-merfish.html">Initiation &amp; Preprocessing with MERFISH data</a></li>
    <li><a class="dropdown-item" href="../articles/initiation-and-preprocessing-visium.html">Initiation &amp; Preprocessing with Visium data</a></li>
    <li><a class="dropdown-item" href="../articles/initiation-and-preprocessing-visium-hd.html">Initiation &amp; Preprocessing with VisiumHD data</a></li>
    <li><h6 class="dropdown-header" data-toc-skip>Basic Manipulation of a SPATA2 object</h6></li>
    <li><a class="dropdown-item" href="../articles/adding-data.html">Adding Data to SPATA2 objects</a></li>
    <li><a class="dropdown-item" href="../articles/package-compatibility.html">Platform Compatibility</a></li>
    <li><a class="dropdown-item" href="../articles/molecular-variables.html">Molecular Variables</a></li>
    <li><a class="dropdown-item" href="../articles/image-handling.html">Image Handling</a></li>
    <li><a class="dropdown-item" href="../articles/spatial-segmentation.html">Spatial Segmentation</a></li>
    <li><h6 class="dropdown-header" data-toc-skip>Implemented Algorithms &amp; Concepts</h6></li>
    <li><a class="dropdown-item" href="../articles/cnv.html">Copy Number Variations</a></li>
    <li><a class="dropdown-item" href="../articles/clustering.html">Clustering</a></li>
    <li><a class="dropdown-item" href="../articles/dimensional-reduction.html">Dimensional Reduction</a></li>
    <li><a class="dropdown-item" href="../articles/dea.html">Differential Expression</a></li>
    <li><a class="dropdown-item" href="../articles/gsea.html">Gene Set Enrichment</a></li>
    <li><a class="dropdown-item" href="../articles/spatial-measures.html">Spatial Measures</a></li>
    <li><h6 class="dropdown-header" data-toc-skip>Spatial Gradient Screening</h6></li>
    <li><a class="dropdown-item" href="../articles/creating-spatial-annotations.html">Creating Spatial Annotations</a></li>
    <li><a class="dropdown-item" href="../articles/using-spatial-annotations.html">Using Spatial Annotations</a></li>
    <li><a class="dropdown-item" href="../articles/spatial-annotation-screening.html">Spatial Annotation Screening</a></li>
    <li><a class="dropdown-item" href="../articles/creating-spatial-trajectories.html">Creating Spatial Trajectories</a></li>
    <li><a class="dropdown-item" href="../articles/spatial-trajectory-screening.html">Spatial Trajectory Screening</a></li>
    <li><h6 class="dropdown-header" data-toc-skip>Visualization</h6></li>
    <li><a class="dropdown-item" href="../articles/plotting-surface.html">Surface Plotting</a></li>
  </ul></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Plot DEA results via dot plots</h1>

      <div class="d-none name"><code>plotDeaDotPlot.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Visualizes results of DE analysis with
dot plots.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">plotDeaDotPlot</span><span class="op">(</span></span>
<span>  <span class="va">object</span>,</span>
<span>  across <span class="op">=</span> <span class="fu"><a href="getDefaultGrouping.html">getDefaultGrouping</a></span><span class="op">(</span><span class="va">object</span><span class="op">)</span>,</span>
<span>  across_subset <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  relevel <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  method_de <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  by_group <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  max_adj_pval <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  min_lfc <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  n_highest_lfc <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  n_lowest_pval <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  genes <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  color_by <span class="op">=</span> <span class="st">"avg_log2FC"</span>,</span>
<span>  alpha_by <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  alpha_trans <span class="op">=</span> <span class="st">"identity"</span>,</span>
<span>  color_trans <span class="op">=</span> <span class="st">"identity"</span>,</span>
<span>  size_by <span class="op">=</span> <span class="st">"p_val_adj"</span>,</span>
<span>  size_trans <span class="op">=</span> <span class="st">"reverse"</span>,</span>
<span>  pt_alpha <span class="op">=</span> <span class="fl">0.9</span>,</span>
<span>  pt_size <span class="op">=</span> <span class="fl">2</span>,</span>
<span>  pt_color <span class="op">=</span> <span class="st">"blue4"</span>,</span>
<span>  pt_clrp <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  pt_clrsp <span class="op">=</span> <span class="st">"plasma"</span>,</span>
<span>  scales <span class="op">=</span> <span class="st">"free"</span>,</span>
<span>  nrow <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  ncol <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  transform_with <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  arrange_genes <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  reverse <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  reverse_within <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  assay_name <span class="op">=</span> <span class="fu"><a href="activateAssay.html">activeAssay</a></span><span class="op">(</span><span class="va">object</span><span class="op">)</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-object">object<a class="anchor" aria-label="anchor" href="#arg-object"></a></dt>
<dd><p>An object of class <code><a href="SPATA2.html">SPATA2</a></code> or, in case of S4 generics,
objects of classes for which a method has been defined.</p></dd>


<dt id="arg-across">across<a class="anchor" aria-label="anchor" href="#arg-across"></a></dt>
<dd><p>Character value or NULL. Specifies the grouping variable of interest.</p>
<p>Use <code><a href="getGroupingOptions.html">getGroupingOptions()</a></code> to obtain all variable names that group the
barcode spots of your object in a certain manner.</p></dd>


<dt id="arg-across-subset">across_subset<a class="anchor" aria-label="anchor" href="#arg-across-subset"></a></dt>
<dd><p>Character vector or NULL. Specifies the particular groups
of interest the grouping variable specified in argument <code>across</code> contains.</p>
<p>If set to NULL all of them are chosen. You can prefix groups you are NOT interested in
with a <em>'-'</em>. (Saves writing if there are more groups you are interested in
than groups you are not interested in.)</p>
<p>Use <code><a href="getGroupNames.html">getGroupNames()</a></code> to obtain all valid input options.</p></dd>


<dt id="arg-relevel">relevel<a class="anchor" aria-label="anchor" href="#arg-relevel"></a></dt>
<dd><p>Logical value. If set to TRUE the input order of <code>across_subset</code>
determines the order in which the groups of interest are displayed. Groups that
are not included are dropped which affects the colors with which they are displayed.</p></dd>


<dt id="arg-method-de">method_de<a class="anchor" aria-label="anchor" href="#arg-method-de"></a></dt>
<dd><p>Character value. Denotes the method to according to which the de-analysis is performed.
Given to argument <code>test.use</code> of the <code><a href="https://satijalab.org/seurat/reference/FindAllMarkers.html" class="external-link">Seurat::FindAllMarkers()</a></code>-function. Run <code>SPATA::dea_methods</code>
to obtain all valid input options.</p></dd>


<dt id="arg-by-group">by_group<a class="anchor" aria-label="anchor" href="#arg-by-group"></a></dt>
<dd><p>Logical value. If TRUE for every group in the grouping
variable a single dot plot is created. If FALSE one plot for all groups and all
gene sets is created.</p></dd>


<dt id="arg-genes">genes<a class="anchor" aria-label="anchor" href="#arg-genes"></a></dt>
<dd><p>Character vector or NULL. If character, vector of gene names
that determines which genes are included. If NULL, genes are taken according
to the threshold input for average log fold change and adjusted p-value.</p></dd>


<dt id="arg-color-by">color_by<a class="anchor" aria-label="anchor" href="#arg-color-by"></a></dt>
<dd><p>Character value. The variables by which
to color the data points.</p></dd>


<dt id="arg-pt-alpha">pt_alpha<a class="anchor" aria-label="anchor" href="#arg-pt-alpha"></a></dt>
<dd><p>Numeric value. Specifies the degree of transparency of all points.</p></dd>


<dt id="arg-pt-size">pt_size<a class="anchor" aria-label="anchor" href="#arg-pt-size"></a></dt>
<dd><p>Numeric value. Specifies the size of all points.</p></dd>


<dt id="arg-pt-clrp">pt_clrp<a class="anchor" aria-label="anchor" href="#arg-pt-clrp"></a></dt>
<dd><p>The color palette to be used if the specified variable displayed by
color is categorical/discrete. Run <code><a href="validActivationFunctions.html">validColorPalettes()</a></code> to see valid input.</p></dd>


<dt id="arg-pt-clrsp">pt_clrsp<a class="anchor" aria-label="anchor" href="#arg-pt-clrsp"></a></dt>
<dd><p>The color spectrum to be used if the specified variable displayed by
color is continuous. Run <code><a href="validActivationFunctions.html">validColorSpectra()</a></code> to see valid input.</p></dd>


<dt id="arg-nrow-ncol">nrow, ncol<a class="anchor" aria-label="anchor" href="#arg-nrow-ncol"></a></dt>
<dd><p>Numeric values or NULL. Used to arrange multiple plots.</p></dd>


<dt id="arg-transform-with">transform_with<a class="anchor" aria-label="anchor" href="#arg-transform-with"></a></dt>
<dd><p>List or <code>NULL.</code> If list, can be used to transform continuous variables before usage.
Names of the list slots refer to the variable. The content of the slot refers to the transforming functions.
E.g if the variable of interest is <em>GFAP</em> gene expression, the following would work:</p>
<ul><li><p>Single function: <code>transform_with = log10</code>,</p></li>
<li><p>Multiple functions: <code>transform_with = list(GFAP = list(log10, log2)</code></p></li>
</ul><p>In case of plotting:
Useful if you want to apply more than one transformation on variables mapped to
plotting aesthetics. Input for <code>transform_with</code> is applied before the
respective <code>&lt;aes&gt;_trans</code> argument.</p></dd>


<dt id="arg-assay-name">assay_name<a class="anchor" aria-label="anchor" href="#arg-assay-name"></a></dt>
<dd><p>Only relevant if the <code>SPATA2</code> object contains more than
one assay: Denotes the assay of interest and thus the
<a href="concept_molecular_modalities.html">molecular modality</a> to use. Defaults to the active assay
as set by <code><a href="activateAssay.html">activateAssay()</a></code>.</p></dd>


<dt id="arg--">...<a class="anchor" aria-label="anchor" href="#arg--"></a></dt>
<dd><p>Used to absorb deprecated arguments or functions.</p></dd>


<dt id="arg-x">x<a class="anchor" aria-label="anchor" href="#arg-x"></a></dt>
<dd><p>Character value. Specifies what is plotted on the x-axis.
If <em>p_val_adj</em> the scale is reversed. Ignored if <code>by_group</code> = FALSE.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A ggplot.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://themilolab.com/" class="external-link">SPATA2</a></span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"example_data"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">object</span> <span class="op">&lt;-</span> <span class="va">example_data</span><span class="op">$</span><span class="va">object_UKF269T_diet</span></span>
<span></span>
<span><span class="va">object</span> <span class="op">&lt;-</span> <span class="fu"><a href="runDEA.html">runDEA</a></span><span class="op">(</span><span class="va">object</span>, across <span class="op">=</span> <span class="st">"histology"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">plotDeaDotplot</span><span class="op">(</span><span class="va">object</span>, across <span class="op">=</span> <span class="st">"histology"</span><span class="op">)</span></span>
<span><span class="fu">plotDeaDotplot</span><span class="op">(</span><span class="va">object</span>, across <span class="op">=</span> <span class="st">"histology"</span>, across_subset <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"tumor"</span>, <span class="st">"transition"</span><span class="op">)</span><span class="op">)</span></span>
<span></span></code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Jan Kueckelhaus, Dieter-Henrik Heiland.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.0.</p>
</div>

    </footer></div>





  </body></html>

