<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Section dummy — section_dummy • SPATA2</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Section dummy — section_dummy"><meta name="description" content="Section dummy"><meta property="og:description" content="Section dummy"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">SPATA2</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">3.0.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item"><a class="nav-link" href="../articles/installation.html">Installation</a></li>
<li class="nav-item"><a class="nav-link" href="../articles/spata-data.html">SPATAData</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><h6 class="dropdown-header" data-toc-skip>Initiating a SPATA2 object</h6></li>
    <li><a class="dropdown-item" href="../articles/initiation-and-preprocessing-customized.html">Initiation &amp; Preprocessing with individual data</a></li>
    <li><a class="dropdown-item" href="../articles/initiation-and-preprocessing-merfish.html">Initiation &amp; Preprocessing with MERFISH data</a></li>
    <li><a class="dropdown-item" href="../articles/initiation-and-preprocessing-visium.html">Initiation &amp; Preprocessing with Visium data</a></li>
    <li><a class="dropdown-item" href="../articles/initiation-and-preprocessing-visium-hd.html">Initiation &amp; Preprocessing with VisiumHD data</a></li>
    <li><h6 class="dropdown-header" data-toc-skip>Basic Manipulation of a SPATA2 object</h6></li>
    <li><a class="dropdown-item" href="../articles/adding-data.html">Adding Data to SPATA2 objects</a></li>
    <li><a class="dropdown-item" href="../articles/package-compatibility.html">Platform Compatibility</a></li>
    <li><a class="dropdown-item" href="../articles/molecular-variables.html">Molecular Variables</a></li>
    <li><a class="dropdown-item" href="../articles/image-handling.html">Image Handling</a></li>
    <li><a class="dropdown-item" href="../articles/spatial-segmentation.html">Spatial Segmentation</a></li>
    <li><h6 class="dropdown-header" data-toc-skip>Implemented Algorithms &amp; Concepts</h6></li>
    <li><a class="dropdown-item" href="../articles/cnv.html">Copy Number Variations</a></li>
    <li><a class="dropdown-item" href="../articles/clustering.html">Clustering</a></li>
    <li><a class="dropdown-item" href="../articles/dimensional-reduction.html">Dimensional Reduction</a></li>
    <li><a class="dropdown-item" href="../articles/dea.html">Differential Expression</a></li>
    <li><a class="dropdown-item" href="../articles/gsea.html">Gene Set Enrichment</a></li>
    <li><a class="dropdown-item" href="../articles/spatial-measures.html">Spatial Measures</a></li>
    <li><h6 class="dropdown-header" data-toc-skip>Spatial Gradient Screening</h6></li>
    <li><a class="dropdown-item" href="../articles/creating-spatial-annotations.html">Creating Spatial Annotations</a></li>
    <li><a class="dropdown-item" href="../articles/using-spatial-annotations.html">Using Spatial Annotations</a></li>
    <li><a class="dropdown-item" href="../articles/spatial-annotation-screening.html">Spatial Annotation Screening</a></li>
    <li><a class="dropdown-item" href="../articles/creating-spatial-trajectories.html">Creating Spatial Trajectories</a></li>
    <li><a class="dropdown-item" href="../articles/spatial-trajectory-screening.html">Spatial Trajectory Screening</a></li>
    <li><h6 class="dropdown-header" data-toc-skip>Visualization</h6></li>
    <li><a class="dropdown-item" href="../articles/plotting-surface.html">Surface Plotting</a></li>
  </ul></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Section dummy</h1>

      <div class="d-none name"><code>section_dummy.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Section dummy</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">section_dummy</span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="area-measures">Area measures<a class="anchor" aria-label="anchor" href="#area-measures"></a></h2>



<p>The vignette on area measures in SPATA2 has been replaced.
Click <code><a href="concept_area_measure.html">here</a></code> to read it.</p>
    </div>
    <div class="section level2">
    <h2 id="distance-measures">Distance measures<a class="anchor" aria-label="anchor" href="#distance-measures"></a></h2>



<p>The vignette on distance measures in SPATA2 has been replaced. Click
<code><a href="concept_distance_measure.html">here</a></code> to read it.</p>
    </div>
    <div class="section level2">
    <h2 id="expansion-of-cropped-image-sections">Expansion of cropped image sections<a class="anchor" aria-label="anchor" href="#expansion-of-cropped-image-sections"></a></h2>



<p>The argument <code>expand</code> is a versatile way, to specify how a cropped
image section is extracted. If you want the cropped image as is, specify
<code>expand = 0</code>. Else, there are multiple options. In general, <code>expand</code> takes
three kinds of values, namely percentages, distances and distance exclamations.</p>
<ul><li><p>Percentage: A string suffixed with <em>%</em>. E.g. <code>expand = '50%'</code>
adds 50% of the distance from the center to the border of the image annotation
to the image frame.</p></li>
<li><p>Distance measures: In pixel or European units of length. E.g. <code>expand =  list(x = '1mm')</code>
expands the x-axis on both sides with 1mm. <code>expand = list(x = c('0.5mm', 1.5mm')</code>
expands the x-axis on the left side with 0.5mm and on the right side with 1.5mm.</p></li>
<li><p>Exclam distance measures: Distance measure with an exclamation mark
suffix. E.g. <code>expand = '1mm!'</code> centers the image and forces an axis length of
1 millimeter. (Example 5)</p></li>
</ul><p>Depending on how the values are specified different parts of the image can be
expanded.</p>
<p>Single values, like <code>expand = 50</code>, are recycled: Every end of each image axis
is expanded by 50 pixel. (Example 2)</p>
<p>Vectors of length two, like <code>expand = c('1mm', '2mm')</code>, are recycled: The beginning
of each axis is expanded by 1 millimeter. The end of each axis is expanded by
2mm. (Example 3)</p>
<p>Named lists can be more precise. <code>expand = list(x = c('1mm', '0.5mm'), y = c('0.25mm', '1mm'))</code>.
Applies the vectors to expand the corresponding axis. (Example 4)</p>
<p>Using exclam input the side of the axis must not be specified as the
axis is fixed as a whole. E.g <code>expand = list(x = '1mm!', y = '2mm!')</code> results
in the same output as <code>expand = list(x = c('1mm!', '1mm!'), y = c('2mm!', '2mm!')</code>.</p>
    </div>
    <div class="section level2">
    <h2 id="image-visualization-with-ggplot-">Image visualization with ggplot2<a class="anchor" aria-label="anchor" href="#image-visualization-with-ggplot-"></a></h2>



<p>When comparing the output of <code>ggplot() + ggpLayerImage()</code> with other image plotting functions,
you may notice that the image appears horizontally flipped when plotted using <code><a href="ggpLayerImage.html">ggpLayerImage()</a></code>.
This behavior is due to the use of a Cartesian coordinate system in <code>SPATA2</code>, where a pixel
with coordinates c(width = 1, height = 1) is plotted on the left side at the bottom.
In contrast, functions like <code><a href="https://rdrr.io/pkg/EBImage/man/display.html" class="external-link">EBImage::display()</a></code> or <code><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">graphics::plot()</a></code> use an <em>image space</em> coordinate system,
where pixel heights start from the top. Consequently, in <em>image space</em>, pixel c(width = 1, height = 1)
is displayed on the top resulting in mirror inverted visualization of the image.</p>
<p>We chose to use a Cartesian coordinate system in <code>SPATA2</code> because we believe it provides a more intuitive
framework for the spatial alignment of tissue, spatial annotations, spatial trajectories,
barcoded sots, single cells, etc. where coordinates in the corresponding data.frames are provided
in form of <em>x</em>- and <em>y</em>-variables. See <code><a href="getCoordsDf.html">getCoordsDf()</a></code>, <code><a href="getImgAnnOutlineDf.html">getImgAnnOutlineDf()</a></code>, <code><a href="getTissueOutlineDf.html">getTissueOutlineDf()</a></code> etc.</p>
<p>If you prefer to view your image in the regular orientation, you can use the <code><a href="flipAll.html">flipAll()</a></code> function on your object,
specifying <code>axis = "horizontal"</code>, to reverse this effect.</p>
    </div>
    <div class="section level2">
    <h2 id="selection-of-spatial-annotations">Selection of spatial annotations<a class="anchor" aria-label="anchor" href="#selection-of-spatial-annotations"></a></h2>



<p>Selection of spatial annotations via the arguments <code>ids</code>, <code>class</code>, <code>tags</code> and
<code>test</code> works in three steps:</p>
<p>First, if <code>ids</code> is a character it prefilters the annotations by ID and only
the specified ones are submitted to the next steps. If it is <code>NULL</code>, all
annotations are submitted to the next steps.</p>
<p>Secondd, if <code>class</code> is a character it filters the annotations remaining
after the first step by their class. If <code>NULL</code>, the step is skipped.</p>
<p>Third, if <code>tags</code> is a character it is used in combination with <code>test</code> to select
from the spatial annotations that remain after the second step based on the meta data
they are tagged with. There are multiple options:</p><ol><li><p>Argument <code>test</code> set to <em>'any'</em> or <em>1</em>: To be included, an image annotation
must be tagged with at least one of the input tags.</p></li>
<li><p>Argument <code>test</code> set to <em>'all'</em> or <em>2</em>: To be included, an image annotation
must be tagged with all of the input tags. Can contain tags that are not specified.</p></li>
<li><p>Argument <code>test</code> set to <em>'identical'</em> or <em>3</em>: To be included, an image annotation
must be tagged with all of the input tags. Can not be tagged with anything else.</p></li>
<li><p>Argument <code>test</code> set to <em>not_identical</em> or <em>4</em>: To be included, an image
annotation must <strong>not</strong> be tagged with the combination of input tags.</p></li>
<li><p>Argument <code>test</code> set to <em>'none'</em> or <em>5</em>: To be included, an image annotation
must <strong>not</strong> contain any of the input tags.</p></li>
</ol><p>If <code>tags</code> is <code>NULL</code>, the step is skipped. Therefore, if <code>ids</code>, <code>class</code> and <code>tags</code>
are all NULL, which is the default, all annotations are selected as all subsetting
steps are skipped. Eventually, the remaining spatial annotations are submitted to
whatever the respective function does.</p>
    </div>

  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Jan Kueckelhaus, Dieter-Henrik Heiland.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.0.</p>
</div>

    </footer></div>





  </body></html>

